// oxygen-cli/build/ox_modules/module-twilio.js
"use strict";

require("source-map-support/register");
var _OxygenError = _interopRequireDefault(require("../errors/OxygenError"));
var _util = _interopRequireDefault(require("../lib/util"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/*
 * Copyright (C) 2015-present CloudBeat Limited
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 */

/**
 * @name twilio
 * @description Provides methods for working with Twilio service.
 */

module.exports = function () {
  const utils = require('./utils');
  const OxError = this.OxError = _OxygenError.default;
  const errHelper = this.errHelper = require('../errors/helper');
  const helpers = this.helpers = {};
  var _client;
  module.isInitialized = function () {
    return _client !== undefined;
  };

  /**
   * @summary Set Twilio authentication details.
   * @function init
   * @param {String} accountSid - Account SID.
   * @param {String} authToken - Authentication token.
   */
  module.init = function (accountSid, authToken) {
    _client = require('twilio')(accountSid, authToken);
  };

  /**
   * @summary Retrieves last SMS message.
   * @function getLastSms
   * @param {Boolean} removeOnRead - Specifies whether to delete the message after reading it.
   * @param {Number} timeout - Timeout in milliseconds for waiting for the message to arrive.
   * @param {Number=} notOlderThan - Retrieve message only if it arrived not before the given time (in ms). Default is 4 minutes.
   * @param {String=} fromNumber - From number to filter message
   * @return {String} SMS text.
   */
  module.getLastSms = async function (removeOnRead, timeout, notOlderThan, fromNumber) {
    helpers.assertArgumentBool(removeOnRead, 'removeOnRead');
    helpers.assertArgumentNumberNonNegative(timeout, 'timeout');
    if (fromNumber) {
      helpers.assertArgumentString(fromNumber, 'fromNumber');
    }
    if (!notOlderThan) {
      notOlderThan = 4 * 60 * 1000;
    }
    var msg;
    var now = Date.now();
    var earliestMessageDate = new Date(now - notOlderThan);
    while (!msg && Date.now() - now < timeout) {
      await (() => {
        return new Promise((resolve, reject) => {
          _client.messages.list({
            dateSentAfter: earliestMessageDate
          }, function (err, messages) {
            var _msg;
            if (messages && typeof messages[Symbol.iterator] === 'function') {
              for (_msg of messages) {
                if (_msg.direction == 'inbound') {
                  if (fromNumber && _msg.from !== fromNumber) {
                    // ignore
                  } else {
                    var _msgDate = Date.parse(_msg.dateCreated);
                    // if message is newer than the previous one - save it
                    if (msg && Date.parse(msg.dateCreated) < _msgDate) {
                      msg = _msg;
                    } else if (!msg) {
                      msg = _msg;
                    }
                  }
                }
              }
            }
            resolve();
          });
        });
      })();
      await _util.default.sleep(800);
    }
    if (!msg) {
      throw new OxError(errHelper.errorCode.TWILIO_ERROR, "Couldn't get the SMS within " + timeout + 'ms.');
    }
    if (removeOnRead) {
      await (() => {
        return new Promise((resolve, reject) => {
          _client.messages(msg.sid).remove().then(() => {
            resolve();
          });
        });
      })();
    }
    return msg.body;
  };

  /**
   * @summary Retrieves last SMS message sent from API.
   * @function getLastSentApiSms
   * @param {Boolean} removeOnRead - Specifies whether to delete the message after reading it.
   * @param {Number} timeout - Timeout in milliseconds for waiting for the message to arrive.
   * @param {Number=} notOlderThan - Retrieve message only if it arrived not before the given time (in ms). Default is 4 minutes.
   * @return {String} SMS text.
   */
  module.getLastSentApiSms = async function (removeOnRead, timeout, notOlderThan) {
    helpers.assertArgumentBool(removeOnRead, 'removeOnRead');
    helpers.assertArgumentNumberNonNegative(timeout, 'timeout');
    if (!notOlderThan) {
      notOlderThan = 4 * 60 * 1000;
    }
    var msg;
    var now = Date.now();
    var earliestMessageDate = new Date(now - notOlderThan);
    while (!msg && Date.now() - now < timeout) {
      await (() => {
        return new Promise((resolve, reject) => {
          _client.messages.list({
            dateSentAfter: earliestMessageDate
          }, function (err, messages) {
            var _msg;
            if (messages && typeof messages[Symbol.iterator] === 'function') {
              for (_msg of messages) {
                if (_msg.direction == 'outbound-api') {
                  var _msgDate = Date.parse(_msg.dateCreated); // if message is newer than the previous one - save it

                  if (msg && Date.parse(msg.dateCreated) < _msgDate) {
                    msg = _msg;
                  } else if (!msg) {
                    msg = _msg;
                  }
                }
              }
            }
            resolve();
          });
        });
      })();
      await _util.default.sleep(800);
    }
    if (!msg) {
      throw new OxError(errHelper.errorCode.TWILIO_ERROR, "Couldn't get the SMS within " + timeout + 'ms.');
    }
    if (removeOnRead) {
      await (() => {
        return new Promise((resolve, reject) => {
          _client.messages(msg.sid).remove().then(() => {
            resolve();
          });
        });
      })();
    }
    return msg.body;
  };

  /**
   * @summary Send an SMS.
   * @function sendSms
   * @param {String} from - Phone number to send from.
   * @param {String} to - Phone number to send to.
   * @param {String} message - Message to send.
   * @return {String} Message SID.
   * @example <caption>[javascript] Usage example</caption>
   * twilio.init('Account Sid', 'Account Token');
   * twilio.sendSms('+1xxxxxxxxxx', '+972xxxxxxxxx', 'Hello World!');
   */
  module.sendSms = async function (from, to, message) {
    helpers.assertArgumentNonEmptyString(from, 'from');
    helpers.assertArgumentNonEmptyString(to, 'to');
    helpers.assertArgumentNonEmptyString(message, 'message');
    var response = null;
    await (() => {
      return new Promise((resolve, reject) => {
        _client.messages.create({
          body: message,
          from: from,
          to: to
        }).then(message => {
          response = message.sid;
          resolve();
        }).catch(err => {
          response = err;
          resolve();
        });
      });
    })();
    if (response.message) {
      var msg = response.message;
      if (response.moreInfo) {
        msg += ' For more info: ' + response.moreInfo;
      } else if (response.code) {
        msg = 'Unable to connect to Twilio: ' + msg;
      }
      throw new OxError(errHelper.errorCode.TWILIO_ERROR, msg);
    }
    return response;
  };
  helpers.assertArgument = (val, name) => utils.assertArgument.call(this, val, name);
  helpers.assertArgumentNonEmptyString = (val, name) => utils.assertArgumentNonEmptyString.call(this, val, name);
  helpers.assertArgumentString = (val, name) => utils.assertArgumentString.call(this, val, name);
  helpers.assertArgumentNumber = (val, name) => utils.assertArgumentNumber.call(this, val, name);
  helpers.assertArgumentNumberNonNegative = (val, name) => utils.assertArgumentNumberNonNegative.call(this, val, name);
  helpers.assertArgumentBool = (val, name) => utils.assertArgumentBool.call(this, val, name);
  helpers.assertArgumentBoolOptional = (val, name) => utils.assertArgumentBoolOptional.call(this, val, name);
  helpers.assertArgumentTimeout = (val, name) => utils.assertArgumentTimeout.call(this, val, name);
  return module;
};
