// oxygen-cli/build/runners/oxygen/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ParameterManager", {
  enumerable: true,
  get: function () {
    return _paramManager.default;
  }
});
exports.default = void 0;
require("source-map-support/register");
var _logger = _interopRequireDefault(require("../../lib/logger"));
var _events = require("events");
var _lodash = _interopRequireDefault(require("lodash"));
var _when = require("when");
var _path = _interopRequireDefault(require("path"));
var _suiteResult = _interopRequireDefault(require("../../model/suite-result"));
var _caseResult = _interopRequireDefault(require("../../model/case-result"));
var _testResult = _interopRequireDefault(require("../../model/test-result"));
var _status = _interopRequireDefault(require("../../model/status"));
var _util = _interopRequireDefault(require("../../lib/util"));
var _helper = _interopRequireDefault(require("../../errors/helper"));
var _OxygenError = _interopRequireDefault(require("../../errors/OxygenError"));
var _paramManager = _interopRequireDefault(require("../../lib/param-manager.js"));
var _WorkerProcess = _interopRequireDefault(require("../WorkerProcess"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const log = (0, _logger.default)('OxygenRunner');
const DEFAULT_ISSUER = 'user';
const MAX_RERUNS = 2;
const {
  v1
} = require('uuid');
const snooze = ms => new Promise(resolve => setTimeout(resolve, ms));
const getSuitesBreakpoints = suites => {
  const retval = [];
  if (suites && Array.isArray(suites) && suites.length > 0) {
    suites.map(suite => {
      if (suite && suite.cases && Array.isArray(suite.cases) && suite.cases.length > 0) {
        suite.cases.map(caze => {
          if (caze && caze.breakpoints && Array.isArray(caze.breakpoints) && caze.breakpoints.length > 0) {
            caze.breakpoints.map(breakpoint => {
              retval.push(breakpoint);
            });
          }
        });
      }
    });
  }
  return retval;
};
const removeBreakpointFromSuites = (suites, removeBreakpoint) => {
  if (suites && Array.isArray(suites) && suites.length > 0) {
    suites.map(suite => {
      if (suite && suite.cases && Array.isArray(suite.cases) && suite.cases.length > 0) {
        suite.cases.map(caze => {
          if (caze && caze.breakpoints && Array.isArray(caze.breakpoints) && caze.breakpoints.length > 0) {
            caze.breakpoints.map((breakpoint, idx) => {
              if (breakpoint.file === removeBreakpoint.file && breakpoint.line === removeBreakpoint.line) {
                caze.breakpoints.splice(idx, 1);
              }
            });
          }
        });
      }
    });
  }
};
class OxygenRunner extends _events.EventEmitter {
  constructor(options) {
    super();
    this._id = _util.default.generateUniqueId();
    this._isRunning = false;
    this._isInitializing = false;
    this._isDisposing = false;
    this._worker = null;
    this._debugMode = false;
    this._workerProcLastError = null;
    this._npmGRootExecution = true;
    if (options && typeof options.npmGRootExecution !== 'undefined') {
      this._npmGRootExecution = options.npmGRootExecution;
    }
    this._suite;
    this._envVars = {};
    this._vars = {};
    this._caps = {};
    this._modCaps = {};
    this._attributes = {};
    this._options = null;
    this._localTime = false;
    this._testKilled = false;
    this._scriptContentLineOffset = 3;
    this._currentCaseResultId = undefined;
    this._currentSuiteResultId = undefined;
    this._currentStepResultId = undefined;
    this._whenDisposed = (0, _when.defer)();
  }
  getId() {
    return this._id;
  }
  async init(options, caps = {}, reporter) {
    if (!options.suites) {
      throw new Error('Initialization failed - no test suites are defined. You must define "suites" property in Oxygen options.');
    }
    this._options = options;
    this._cwd = this._options.cwd || process.cwd();
    this._reporter = reporter;
    this._isInitialized = true;
    if (options) {
      let optEnv = {};
      if (options.env) {
        if (typeof options.env === 'string' && typeof options.envs === 'object' && options.envs[options.env]) {
          optEnv = options.envs[options.env];
        } else if (typeof options.env === 'object' && Object.keys(options.env)) {
          optEnv = options.env;
        }
      }
      if (options.envVars && Object.keys(options.envVars)) {
        optEnv = {
          ...optEnv,
          ...options.envVars
        };
      }
      this._env = {
        ...this._env,
        ...optEnv
      };
      this._attributes = options.testAttributes || {};
    }
    this._caps = {
      ...caps
    };
    this._suites = [...options.suites];
    this._debugMode = options.debugPortIde || options.debugPort ? true : false;
    this._debugPort = options.debugPort || options.debugPortIde || null;
    options.scriptContentLineOffset = this._scriptContentLineOffset;
    this._localTime = this._options && this._options.localTime || this._localTime;
    await this._startWorkerProcess();
    await this._worker_InitOxygen();
  }
  async dispose(status = null) {
    this._isDisposing = true;
    try {
      this._worker && (await this._worker.stopDebugger());
      if (this._worker && this._worker.isRunning) {
        await this._worker.stop(status);
      }
    } catch (e) {
      console.log('Error when disposing Runner', e);
      log.warn('Error when disposing Runner:', e);
    }
    if (status !== 'CANCELED') {
      this._resetGlobalVariables();
    }
  }
  async run() {
    if (this._isInitializing) {
      throw new Error("Initialization hasn't been completed yet, wait for 'initialized' event");
    }
    if (this._isRunning) {
      throw new Error("Previous test is still running, wait for 'test-ended' event");
    }
    this._isRunning = true;
    await this._reporter.onRunnerStart(this._id, this._options, this._caps);
    let error = null;
    let result = null;
    try {
      result = await this._runTest();
    } catch (e) {
      error = e;
    }
    if (!error && result && result.failure) {
      error = result.failure;
    }
    await this._reporter.onRunnerEnd(this._id, result, error);
    this._isRunning = false;
    if (error) {
      throw error;
    }
    return result;
  }
  async kill(status = null) {
    this._testKilled = true;
    if (this._worker && this._worker.isRunning) {
      await this._worker.kill(status);
    }
  }
  debugContinue() {
    if (this._debugMode && this._worker) {
      this._worker.debugger.continue();
    }
  }
  async updateBreakpoints(breakpoints, filePath) {
    try {
      if (this._worker && this._worker.debugger && this._worker.debugger.setBreakpointsActive) {
        await this._worker.debugger.setBreakpointsActive(false);
      } else {
        await snooze(500);
        return await this.updateBreakpoints(breakpoints, filePath);
      }
      if (this._debugMode && this._worker.debugger) {
        let promises = [];
        const tsBreakpoints = this._worker.debugger.getBreakpoints(filePath);
        const suitesBreakpoints = getSuitesBreakpoints(this._suites);
        for (var suiteBp of suitesBreakpoints) {
          if (!breakpoints.includes(suiteBp.line) && filePath === suiteBp.file) {
            removeBreakpointFromSuites(this._suites, suiteBp);
          }
        }
        for (var actualCurrentBp of tsBreakpoints) {
          if (!breakpoints.includes(actualCurrentBp)) {
            promises.push(this._worker.debugger.removeBreakpointByValue(filePath, actualCurrentBp));
          }
        }
        for (var userSetBp of breakpoints) {
          if (!tsBreakpoints.includes(userSetBp)) {
            promises.push(this._worker.debugger.setBreakpoint(filePath, userSetBp));
          }
        }
        let promiseAllPromise = await Promise.all(promises).then(value => {
          return value;
        });
        await this._worker.debugger.setBreakpointsActive(true);
        await snooze(500);
        log.debug('updateBreakpoints() done.');
        const breakpointsAferManipulations = this._worker.debugger.getBreakpoints(filePath);
        return {
          promiseAllPromise: promiseAllPromise,
          breakpointsAferManipulations: breakpointsAferManipulations
        };
      } else {
        return null;
      }
    } catch (e) {
      log.error('Debugger error', e);
    }
  }
  async replClose() {
    await this._worker._send({
      event: 'repl',
      name: 'repl_stop'
    });
  }
  async replSend(cmd) {
    await this._worker._send({
      event: 'repl',
      name: 'repl_eval',
      content: {
        cmd: cmd
      }
    });
  }
  async replStart() {
    await this._worker.invoke('replStart');
  }
  setBreakpoint(line) {}
  clearBreakpoint(line) {}
  async _runTest() {
    const result = new _testResult.default();
    result.startTime = _util.default.getTimeStamp();
    result.status = _status.default.PASSED;
    result.suites = [];
    let error = null;
    await this._worker_callBeforeTestHook();
    try {
      for (let suite of this._suites) {
        const suiteResult = await this._runSuite(suite);
        result.suites.push(suiteResult);
      }
    } catch (e) {
      error = e;
    }
    result.endTime = _util.default.getTimeStamp();
    result.duration = result.endTime - result.startTime;
    const hasFailedSuites = result.suites.some(suiteIterations => {
      if (suiteIterations && Array.isArray(suiteIterations) && suiteIterations.length > 0) {
        return suiteIterations.some(x => x.status === _status.default.FAILED);
      }
    });
    result.status = hasFailedSuites ? _status.default.FAILED : _status.default.PASSED;
    result.environment = {
      ...this._env
    };
    result.capabilities = {
      ...this._caps,
      ...this._modCaps
    };
    result.options = {
      ...this._options
    };
    if (error) {
      result.failure = _helper.default.getFailureFromError(error);
      result.status = _status.default.FAILED;
    }
    await this._worker_callAfterTestHook(result);
    return result;
  }
  async _runSuite(suite) {
    if (!suite) {
      log.error('suite is null in _runSuite()!!!');
    }
    if (!suite.name && !suite.path) {
      return [];
    }
    if (!suite.iterationCount) {
      suite.iterationCount = 1;
    }
    let showSuiteIterationsMessages = false;
    if (suite.iterationCount > 1) {
      showSuiteIterationsMessages = true;
    }
    const reRunOnFailure = this._options.reRunOnFailed || false;
    const maxReruns = !reRunOnFailure ? 0 : this._options.reRunCount || MAX_RERUNS;
    const reRunDelay = this._options.reRunDelay;
    const suiteIterations = [];
    for (let suiteIteration = 1; suiteIteration <= suite.iterationCount; suiteIteration++) {
      if (showSuiteIterationsMessages) {
        await this._reporter.onIterationStart(this._id, suiteIteration, 'Suite');
      }
      const suiteResult = new _suiteResult.default();
      suiteIterations.push(suiteResult);
      suiteResult.id = this._currentSuiteResultId = v1();
      suiteResult.name = suite.name || _util.default.getFileNameWithoutExt(suite.path);
      suiteResult.refId = suite.refId || suite.id || suite.name;
      suiteResult.startTime = _util.default.getTimeStamp();
      suiteResult.iterationNum = suiteIteration;
      suiteResult.status = _status.default.PASSED;
      await this._worker_callBeforeSuiteHook(suite);
      await this._reporter.onSuiteStart(this._id, suiteResult.id, suiteResult);
      for (let caze of suite.cases) {
        if (!caze.path) {
          continue;
        }
        if (!caze.name) {
          caze.name = _util.default.getFileNameWithoutExt(caze.path);
        }
        if (!caze.iterationCount) {
          caze.iterationCount = 1;
        }
        for (let caseIteration = 1; caseIteration <= caze.iterationCount; caseIteration++) {
          let showCaseIterationsMessages = false;
          if (caze && caze.iterationCount && caze.iterationCount > 1) {
            showCaseIterationsMessages = true;
          }
          if (showCaseIterationsMessages) {
            await this._reporter.onIterationStart(this._id, caseIteration, 'Case');
          }
          const caseResultId = this._currentCaseResultId = v1();
          await this._reporter.onCaseStart(this._id, suiteResult.id, suiteResult.refId, caseResultId, caze);
          let caseResult;
          for (let reRunCount = 0; reRunCount <= maxReruns; reRunCount++) {
            caseResult = await this._runCase(suite, caze, suiteIteration, caseIteration, reRunCount);
            if ((!caseResult || caseResult.status === _status.default.FAILED) && !reRunOnFailure) {
              break;
            } else if (caseResult && caseResult.status !== _status.default.FAILED) {
              break;
            }
            if (reRunDelay && reRunCount < maxReruns - 1) {
              await _util.default.sleep(reRunDelay);
            }
          }
          caseResult.id = caseResultId;
          await this._reporter.onCaseEnd(this._id, suiteResult.id, suiteResult.refId, caseResultId, caseResult);
          if (showCaseIterationsMessages) {
            await this._reporter.onIterationEnd(this._id, caseResult, 'Case');
          }
          suiteResult.cases.push(caseResult);
          if (caseResult.status === _status.default.FAILED) {
            suiteResult.status = _status.default.FAILED;
          } else if (caseResult.status === _status.default.WARNING) {
            suiteResult.status = _status.default.WARNING;
          }
        }
      }
      suiteResult.endTime = _util.default.getTimeStamp();
      suiteResult.duration = suiteResult.endTime - suiteResult.startTime;
      if (showSuiteIterationsMessages) {
        await this._reporter.onIterationEnd(this._id, suiteResult, 'Suite');
      }
      await this._worker_callAfterSuiteHook(suite, suiteResult);
      await this._reporter.onSuiteEnd(this._id, suiteResult.id, suiteResult);
    }
    return suiteIterations;
  }
  async _runCase(suite, caze, suiteIteration, caseIteration, reRunCount = 0) {
    const params = {};
    const reRun = reRunCount > 0;
    if (suite.paramManager) {
      if (reRun) {
        suite.paramManager.readPrev();
      }
      _lodash.default.extend(params, suite.paramManager.getValues());
    }
    if (caze.paramManager) {
      if (reRun) {
        caze.paramManager.readPrev();
      }
      _lodash.default.extend(params, caze.paramManager.getValues());
    }
    if (this._debugMode && this._worker && this._worker.debugger && caze.breakpoints) {
      for (let i = 0; i < caze.breakpoints.length; i++) {
        let line = caze.breakpoints[i].line;
        let file = caze.breakpoints[i].file;
        await this._worker.debugger.setBreakpoint(file, line);
      }
    }
    await this._worker_callBeforeCaseHook(caze);
    const caseResult = new _caseResult.default();
    caseResult.name = caze.name;
    caseResult.location = caze.path;
    caseResult.iterationNum = caseIteration;
    caseResult.reRunCount = reRunCount;
    try {
      await (!this._worker && this._worker_InitOxygen());
      if (!this._worker) {
        return;
      }
    } catch (e) {
      log.error('_worker_InitOxygen() thrown an error:', e);
      caseResult.startTime = caseResult.endTime = _util.default.getTimeStamp();
      caseResult.duration = 0;
      caseResult.failure = _helper.default.getFailureFromError(e);
      caseResult.status = _status.default.FAILED;
      if (suite.paramManager) {
        suite.paramManager.readNext();
      }
      if (caze.paramManager) {
        caze.paramManager.readNext();
      }
      return caseResult;
    }
    await this._worker_startSession();
    try {
      caseResult.startTime = _util.default.getTimeStamp();
      const {
        resultStore,
        context,
        moduleCaps,
        error
      } = await this._worker_Run(suite, caze, suiteIteration, caseIteration, params);
      this._processTestResults({
        resultStore,
        context,
        error,
        moduleCaps
      });
      caseResult.endTime = _util.default.getTimeStamp();
      caseResult.duration = caseResult.endTime - caseResult.startTime;
      caseResult.context = context;
      caseResult.steps = resultStore && resultStore.steps ? resultStore.steps : [];
      caseResult.logs = resultStore && resultStore.logs ? resultStore.logs : [];
      caseResult.har = resultStore && resultStore.har ? resultStore.har : null;
      caseResult.testAttributes = resultStore && resultStore.attributes ? resultStore.attributes : null;
      caseResult.attachments = resultStore && resultStore.attachments ? resultStore.attachments : null;
      var failedSteps = _lodash.default.find(caseResult.steps, {
        status: _status.default.FAILED
      });
      var warningSteps = _lodash.default.find(caseResult.steps, {
        status: _status.default.WARNING
      });
      caseResult.status = _lodash.default.isEmpty(failedSteps) && !error ? _status.default.PASSED : _status.default.FAILED;
      if (error) {
        caseResult.failure = error;
        caseResult.status = _status.default.FAILED;
        caseResult.steps = _util.default.makeTransactionFailedIfStepFailed(caseResult.steps);
      } else if (!_lodash.default.isEmpty(warningSteps)) {
        caseResult.status = _status.default.WARNING;
      }
      await this._worker_callAfterCaseHook(caze, caseResult);
    } catch (e) {
      log.error('_worker_Run() thrown an error:', e);
      caseResult.failure = _helper.default.getFailureFromError(e);
      caseResult.status = _status.default.FAILED;
    }
    const disposeOxygenModules = typeof this._options.autoDispose !== 'boolean' || this._options.autoDispose === true;
    await this._worker_endSession(caseResult.status, disposeOxygenModules);
    if (suite.paramManager) {
      suite.paramManager.readNext();
    }
    if (caze.paramManager) {
      caze.paramManager.readNext();
    }
    return caseResult;
  }
  async _worker_Run(suite, caze, suiteIteration, caseIteration, params) {
    if (!this._worker) {
      log.error('_worker is null but not suppose to!');
      if (this._whenTestCaseFinished) {
        this._whenTestCaseFinished.reject(new Error('_worker is null'));
      }
      return;
    }
    return await this._worker.run({
      scriptName: caze.name,
      scriptPath: caze.path,
      context: {
        params: params,
        env: this._env,
        caps: this._caps,
        vars: this._vars,
        attributes: this._attributes,
        test: {
          case: {
            name: caze.name,
            iteration: caseIteration
          },
          suite: {
            name: suite.name,
            iteration: suiteIteration
          }
        }
      },
      poFile: this._options.po || null
    });
  }
  async _worker_InitOxygen() {
    await (this._worker && this._worker.initOxygen(this._id, this._options, this._caps));
  }
  async _worker_startSession() {
    if (this._worker && this._worker.startSession) {
      await this._worker.startSession();
    }
  }
  async _worker_endSession(status = null, disposeModules = true) {
    if (this._worker && this._worker.endSession) {
      await this._worker.endSession(status, disposeModules);
    }
  }
  async _worker_callBeforeTestHook() {
    try {
      if (this && this._worker && this._worker.invokeTestHook) {
        await this._worker.invokeTestHook('beforeTest', [this._id, this._options, this._caps]);
      }
    } catch (e) {
      log.error('"beforeTest" hook failed:', e);
      this._reporter && (await this._reporter.onLogEntry(null, 'WARN', '"beforeTest" hook failed:' + e, DEFAULT_ISSUER));
    }
  }
  async _worker_callBeforeSuiteHook(suite) {
    try {
      if (this && this._worker && this._worker.invokeTestHook) {
        await this._worker.invokeTestHook('beforeSuite', [suite]);
      }
    } catch (e) {
      log.error('"beforeSuite" hook failed:', e);
      this._reporter && (await this._reporter.onLogEntry(null, 'WARN', '"beforeSuite" hook failed:' + e, DEFAULT_ISSUER));
    }
  }
  async _worker_callBeforeCaseHook(caze) {
    try {
      await (this._worker && this._worker.invokeTestHook('beforeCase', [caze]));
    } catch (e) {
      log.error('"beforeCase" hook failed:', e);
      this._reporter && (await this._reporter.onLogEntry(null, 'WARN', '"beforeCase" hook failed:' + e, DEFAULT_ISSUER));
    }
  }
  async _worker_callAfterTestHook(result) {
    try {
      await (this._worker && this._worker.invokeTestHook('afterTest', [this._id, result]));
    } catch (e) {
      log.error('"afterTest" hook failed:', e);
      this._reporter && (await this._reporter.onLogEntry(null, 'WARN', '"afterTest" hook failed:' + e, DEFAULT_ISSUER));
    }
  }
  async _worker_callAfterSuiteHook(suite, result) {
    try {
      await (this._worker && this._worker.invokeTestHook('afterSuite', [suite, result]));
    } catch (e) {
      log.error('"afterSuite" hook failed:', e);
      this._reporter && (await this._reporter.onLogEntry(null, 'WARN', '"afterSuite" hook failed:' + e, DEFAULT_ISSUER));
    }
  }
  async _worker_callAfterCaseHook(caze, result) {
    try {
      await (this._worker && this._worker.invokeTestHook('afterCase', [caze, result]));
    } catch (e) {
      log.error('"afterCase" hook failed:', e);
      this._reporter && (await this._reporter.onLogEntry(null, 'WARN', '"afterCase" hook failed:' + e, DEFAULT_ISSUER));
    }
  }
  async _startWorkerProcess() {
    const workerPath = _path.default.join(__dirname, 'worker.js');
    this._worker = new _WorkerProcess.default(this._id, workerPath, this._debugMode, this._debugPort, 'Oxygen', this._npmGRootExecution);
    await this._worker.start();
    this._hookWorkerEvents();
    await this._worker.startDebugger();
    this._hookWorkerDebuggerEvents();
  }
  _hookWorkerEvents() {
    const _this = this;
    this._worker.subscribe('command:before', this._handleBeforeCommand.bind(this));
    this._worker.subscribe('command:after', this._handleAfterCommand.bind(this));
    this._worker.on('error', payload => {
      const {
        error
      } = payload;
      log.error('Worker process error: ', error);
      if (this.exitDone) {
        this.emit('test-error', error);
      } else {
        _this._workerProcLastError = error;
      }
    });
    this._worker.on('exit', payload => {
      this.exitDone = true;
      const {
        exitCode
      } = payload;
      if (exitCode && exitCode !== 0) {
        let promise = null;
        if (_this._isDisposing) {
          promise = _this._whenDisposed;
        } else if (_this._isRunning) {
          promise = _this._whenTestCaseFinished;
        }
        let error = _this._workerProcLastError || null;
        if (exitCode == 134) {
          error = new _OxygenError.default(_helper.default.errorCode.SCRIPT_ERROR, 'Out of memory error. Make sure that you don\'t have any memory leaks in the test script.');
        } else if (!error) {
          error = new _OxygenError.default(`Worker process exited with code: ${exitCode}.`);
        }
        promise && promise.reject(error);
        _this._resetGlobalVariables();
      }
    });
    this._worker.on('message', async msg => {
      if (msg.event && msg.event === 'log') {
        if (msg.level === 'DEBUG') {
          log.debug(msg.msg);
        } else if (msg.level === 'INFO') {
          log.info(msg.msg);
        } else if (msg.level === 'ERROR') {
          log.error(msg.msg, msg.err);
        } else if (msg.level === 'WARN') {
          log.warn(msg.msg);
        }
        _this.emit('log', msg.time, msg.level, msg.msg, msg.src || DEFAULT_ISSUER);
        _this._reporter && (await _this._reporter.onLogEntry(msg.time, msg.level, msg.msg, msg.src || DEFAULT_ISSUER, {
          suiteId: this._currentSuiteResultId,
          caseId: this._currentCaseResultId,
          stepId: this._currentStepResultId
        }));
      } else if (msg.event && msg.event === 'line-update') {
        _this.emit('line-update', msg.line, msg.stack, msg.time);
      } else if (msg.event && msg.event === 'result-update') {
        if (msg.method === 'init') {
          _this.emit('init-done', msg);
        }
      } else if (msg.event && msg.event === 'workerError') {
        console.log('workerError:', msg);
        await this.dispose('failed');
        await this.kill('failed');
      } else if (msg.event && msg.event === 'repl') {
        _this.emit('repl', msg);
      }
    });
  }
  _hookWorkerDebuggerEvents() {
    if (!this._worker) {
      return;
    }
    this._worker.debugger && this._worker.debugger.on('debugger:break', breakpointData => {
      this.emit('breakpoint', breakpointData);
    });
    this._worker.debugger && this._worker.debugger.on('debugger:breakError', breakError => {
      this.emit('breakpointError', breakError);
    });
  }
  async _handleBeforeCommand(e) {
    this._currentStepResultId = e.id;
    this._reporter && (await this._reporter.onStepStart(this._id, this._currentSuiteResultId, this._currentCaseResultId, e));
    try {
      if (e && e.module && e.module === 'log') {} else {
        await (this._worker && this._worker.invokeTestHook('beforeCommand', [e]));
      }
    } catch (e) {
      log.error('"beforeCommand" hook failed:', e);
      this._reporter && (await this._reporter.onLogEntry(null, 'WARN', '"beforeCommand" hook failed:' + e, DEFAULT_ISSUER));
    }
  }
  async _handleAfterCommand(e) {
    this._reporter && (await this._reporter.onStepEnd(this._id, this._currentSuiteResultId, this._currentCaseResultId, e.result));
    try {
      if (e && e.module && e.module === 'log') {} else {
        await (this._worker && this._worker.invokeTestHook('afterCommand', [e.result]));
      }
    } catch (e) {
      log.error('"afterCommand" hook failed:', e);
      this._reporter && (await this._reporter.onLogEntry(null, 'WARN', '"afterCommand" hook failed:' + e, DEFAULT_ISSUER));
    }
    this._currentStepResultId = undefined;
  }
  _resetGlobalVariables() {
    this._modCaps = {};
    this._worker = null;
    this._isRunning = false;
    this._isDisposing = false;
    this._isInitializing = false;
    this._testKilled = false;
    this._whenDisposed = (0, _when.defer)();
  }
  _processTestResults({
    resultStore,
    moduleCaps,
    context = {},
    error = null
  }) {
    if (this._testKilled) {
      return null;
    }
    if (error) {
      this.emit('test-error', error);
    }
    this._vars = context.vars || this._vars;
    this._modCaps = {
      ...this._modCaps,
      ...(moduleCaps || {})
    };
    if (context.caps) delete context.caps;
    if (context.env) delete context.env;
    if (context.test) delete context.test;
  }
  _emitIterationEnd(ts, tc, suiteIterationNum, caseIterationNum, caseIterationResult) {
    this.emit('iteration-end', {
      status: caseIterationResult._status,
      failure: caseIterationResult.failure || null,
      case: {
        id: tc.id || null,
        name: tc.name,
        iteration: caseIterationNum,
        iterationCount: tc.iterationCount
      },
      suite: {
        id: ts.id || null,
        name: ts.name,
        iteration: suiteIterationNum,
        iterationCount: ts.iterationCount
      }
    });
  }
  _emitTestEnd(result) {
    this.emit('test-end', result);
  }
}
exports.default = OxygenRunner;
