// oxygen-cli/build/errors/helper.js
"use strict";

require("source-map-support/register");
var _OxygenError = _interopRequireDefault(require("./OxygenError"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const util = require('util');
const ERROR_CODES = {
  SCRIPT_ERROR: 'SCRIPT_ERROR',
  UNKNOWN_ERROR: 'UNKNOWN_ERROR',
  ASSERT_ERROR: 'ASSERT_ERROR',
  PDF_ERROR: 'PDF_ERROR',
  VERIFY_ERROR: 'VERIFY_ERROR',
  ELEMENT_NOT_FOUND: 'ELEMENT_NOT_FOUND',
  FRAME_NOT_FOUND: 'FRAME_NOT_FOUND',
  UNKNOWN_COMMAND_ERROR: 'UNKNOWN_COMMAND_ERROR',
  STALE_ELEMENT_REFERENCE: 'STALE_ELEMENT_REFERENCE',
  ELEMENT_NOT_VISIBLE: 'ELEMENT_NOT_VISIBLE',
  ELEMENT_NOT_INTERACTABLE: 'ELEMENT_NOT_INTERACTABLE',
  LOCATOR_MATCHES_MULTIPLE_ELEMENTS: 'LOCATOR_MATCHES_MULTIPLE_ELEMENTS',
  ELEMENT_STILL_EXISTS: 'ELEMENT_STILL_EXISTS',
  BROWSER_JS_EXECUTE_ERROR: 'BROWSER_JS_EXECUTE_ERROR',
  TIMEOUT: 'TIMEOUT',
  WINDOW_NOT_FOUND: 'WINDOW_NOT_FOUND',
  UNEXPECTED_ALERT_OPEN: 'UNEXPECTED_ALERT_OPEN',
  NO_ALERT_OPEN_ERROR: 'NO_ALERT_OPEN_ERROR',
  MAILINATOR_ERROR: 'MAILINATOR_ERROR',
  APPIUM_UNREACHABLE_ERROR: 'APPIUM_UNREACHABLE_ERROR',
  SELENIUM_UNREACHABLE_ERROR: 'SELENIUM_UNREACHABLE_ERROR',
  CHROMEDRIVER_ERROR: 'CHROMEDRIVER_ERROR',
  WEBDRIVER_ERROR: 'WEBDRIVER_ERROR',
  DB_CONNECTION_ERROR: 'DB_CONNECTION_ERROR',
  DB_QUERY_ERROR: 'DB_QUERY_ERROR',
  SOAP_ERROR: 'SOAP_ERROR',
  SHELL_ERROR: 'SHELL_ERROR',
  MONGODB_ERROR: 'MONGODB_ERROR',
  NOT_IMPLEMENTED_ERROR: 'NOT_IMPLEMENTED_ERROR',
  SERIAL_PORT_ERROR: 'SERIAL_PORT_ERROR',
  HTTP_ERROR: 'HTTP_ERROR',
  EMAIL_ERROR: 'EMAIL_ERROR',
  MODULE_NOT_INITIALIZED_ERROR: 'MODULE_NOT_INITIALIZED_ERROR',
  DEVICE_NOT_FOUND: 'DEVICE_NOT_FOUND',
  PARAMETERS_ERROR: 'PARAMETERS_ERROR',
  INVALID_CAPABILITIES: 'INVALID_CAPABILITIES',
  BROWSER_CONFIGURATION_ERROR: 'BROWSER_CONFIGURATION_ERROR',
  APPIUM_CONNECTION_ERROR: 'APPIUM_CONNECTION_ERROR',
  APPIUM_SESSION_ERROR: 'APPIUM_SESSION_ERROR',
  APPIUM_SESSION_TIMEOUT: 'APPIUM_SESSION_TIMEOUT',
  SELENIUM_SESSION_ERROR: 'SELENIUM_SESSION_ERROR',
  SELENIUM_SESSION_TIMEOUT: 'SELENIUM_SESSION_TIMEOUT',
  SELENIUM_CONNECTION_ERROR: 'SELENIUM_CONNECTION_ERROR',
  URL_OPEN_ERROR: 'URL_OPEN_ERROR',
  OPTION_NOT_FOUND: 'OPTION_NOT_FOUND',
  ATTRIBUTE_NOT_FOUND: 'ATTRIBUTE_NOT_FOUND',
  ELEMENT_STATE_ERROR: 'ELEMENT_STATE_ERROR',
  MOBILE_CONTEXT_ERROR: 'MOBILE_CONTEXT_ERROR',
  APPLICATION_NOT_FOUND_ERROR: 'APPLICATION_NOT_FOUND_ERROR',
  TWILIO_ERROR: 'TWILIO_ERROR',
  HOOK_ERROR: 'HOOK_ERROR',
  NOT_SUPPORTED: 'NOT_SUPPORTED',
  ASSERT_PASSED: 'ASSERT_PASSED',
  MODULE_NOT_FOUND: 'MODULE_NOT_FOUND',
  FILE_BROWSE_ERROR: 'FILE_BROWSE_ERROR',
  CSV_ERROR: 'CSV_ERROR',
  CRYPTO_ERROR: 'CRYPTO_ERROR',
  DNS_ERROR: 'DNS_ERROR',
  CIRCULAR_ERROR: 'CIRCULAR_ERROR',
  XML_ERROR: 'XML_ERROR',
  JAVASCRIPT_ERROR: 'JAVASCRIPT_ERROR'
};
const CHAI_ERROR_CODES = {
  AssertionError: ERROR_CODES.ASSERT_ERROR
};
const SCRIPT_ERROR_CODES = {
  TypeError: ERROR_CODES.SCRIPT_ERROR,
  SyntaxError: ERROR_CODES.SCRIPT_ERROR,
  ReferenceError: ERROR_CODES.SCRIPT_ERROR
};
module.exports = {
  getFailureFromError: function (err) {
    if (!err) {
      return null;
    }
    if (!(err instanceof _OxygenError.default)) {
      err = this.getOxygenError(err);
    }
    return {
      type: err.type,
      message: err.message,
      data: err.data,
      location: err.location,
      stacktrace: err.stacktrace || null
    };
  },
  getDbErrorMessage: function (err) {
    let message = '';
    if (err && err.message) {
      message += err.message;
    }
    if (err && err.odbcErrors && Array.isArray(err.odbcErrors) && err.odbcErrors.length > 0) {
      err.odbcErrors.map(item => {
        if (item && item.message) {
          message += ': ' + item.message;
        }
      });
    }
    return message;
  },
  getOxygenError: function (err, module, cmd, args) {
    if (err instanceof _OxygenError.default) {
      return err;
    }
    var errType = err.type || err.name || typeof err;
    if (err.message && err.message.includes('because element wasn\'t found')) {
      return new _OxygenError.default(ERROR_CODES.ELEMENT_NOT_FOUND, err.message, null, true, err);
    } else if (err.message && err.message.includes('Unsupported locator strategy')) {
      let matches = err.message.match(/invalid argument: (.*)/i);
      return new _OxygenError.default(ERROR_CODES.MOBILE_CONTEXT_ERROR, (matches.length === 2 ? matches[1] : err.message) + '. Make sure you are using the correct mobile context. See mob.setNativeContext and mob.setWebViewContext.');
    } else if (err.message && err.message.includes('Unable to locate an element with the xpath expression')) {
      let matches = err.message.match(/(The string '.*' is not a valid XPath expression.)/i);
      return new _OxygenError.default(ERROR_CODES.SCRIPT_ERROR, matches && matches.length === 2 ? matches[1] : err.message);
    } else if (err.message && err.message.includes('An invalid or illegal selector was specified')) {
      return new _OxygenError.default(ERROR_CODES.SCRIPT_ERROR, 'An invalid or illegal selector was specified');
    } else if (err.message && err.message.startsWith('Given css selector expression')) {
      let matches = err.message.match(/(Given css selector expression ".*" is invalid).*/i);
      return new _OxygenError.default(ERROR_CODES.SCRIPT_ERROR, matches && matches.length === 2 ? matches[1] : err.message);
    } else if (err.message && (err.message.includes('Unable to automate Chrome version') || err.message.includes('No Chromedriver found that can automate'))) {
      return new _OxygenError.default(ERROR_CODES.CHROMEDRIVER_ERROR, extractOriginalError(err.message));
    } else if (err.message && err.message.includes(ORIGINAL_ERROR_MESSAGE)) {
      return new _OxygenError.default(ERROR_CODES.UNKNOWN_ERROR, extractOriginalError(err.message), util.inspect(err));
    } else if (err.message && err.message.includes('Invalid URL')) {
      let matches = err.message.match(/( Invalid URL: .*$).*/i);
      return new _OxygenError.default(ERROR_CODES.SCRIPT_ERROR, matches && matches.length === 2 ? matches[1] : err.message);
    } else if (err.message && err.message.startsWith('element click intercepted')) {
      let matches = err.message.match(/element click intercepted: (.*)/i);
      return new _OxygenError.default(ERROR_CODES.ELEMENT_NOT_VISIBLE, matches && matches.length === 2 ? matches[1] : err.message);
    } else if (err.message && err.message.startsWith('No active session with ID')) {
      return new _OxygenError.default(ERROR_CODES.SELENIUM_SESSION_ERROR, 'Execution time exceeded the timeout set by Selenium');
    } else if (err.message === 'java.net.ConnectException: Connection refused: connect') {
      return new _OxygenError.default(ERROR_CODES.WEBDRIVER_ERROR, 'The underlying WebDriver seems to have crashed: ' + err.message);
    } else if (err.name && err.name.includes('stale element reference')) {
      return new _OxygenError.default(ERROR_CODES.STALE_ELEMENT_REFERENCE, err.name + ' ' + err.message);
    } else if (err.code && err.code === 'MODULE_NOT_FOUND') {
      return new _OxygenError.default(ERROR_CODES.MODULE_NOT_FOUND, err.message);
    } else if (err.message && err.message.includes('no such window') && err.message.includes('target window already closed') && err.message.includes('web view not found')) {
      return new _OxygenError.default(ERROR_CODES.WINDOW_NOT_FOUND, err.message);
    } else if (err.message && err.message.includes('you trying to pass into the execute method wasn\'t found')) {
      return new _OxygenError.default(ERROR_CODES.ELEMENT_NOT_FOUND, err.message);
    } else if (err.message && err.message.includes('element not interactable')) {
      return new _OxygenError.default(ERROR_CODES.ELEMENT_NOT_INTERACTABLE, err.message);
    } else if (err.message && err.message.includes('still not clickable after')) {
      return new _OxygenError.default(ERROR_CODES.ELEMENT_NOT_INTERACTABLE, err.message);
    } else if (err.name === 'invalid selector') {
      return new _OxygenError.default(ERROR_CODES.SCRIPT_ERROR, err.message);
    } else if (err.message && err.message.includes('Timeout awaiting \'request\' for')) {
      const errMessage = err.message.replace('Timeout awaiting \'request\' for', 'Couldn\'t get a response from Selenium/WebDriver within');
      return new _OxygenError.default(module === 'web' ? ERROR_CODES.SELENIUM_SESSION_TIMEOUT : ERROR_CODES.APPIUM_SESSION_TIMEOUT, errMessage);
    } else if (err && err.message && err.message.includes('socket hang up')) {
      return new _OxygenError.default(module === 'web' ? ERROR_CODES.SELENIUM_SESSION_TIMEOUT : ERROR_CODES.APPIUM_SESSION_TIMEOUT, err.message);
    } else if (err.message && err.message.includes('javascript error')) {
      return new _OxygenError.default(ERROR_CODES.JAVASCRIPT_ERROR, err.message);
    } else if (err.message && err.message.includes('An element could not be located on the page using the given search parameters')) {
      return new _OxygenError.default(ERROR_CODES.ELEMENT_NOT_FOUND, err.message);
    }
    var oxErrorCode = CHAI_ERROR_CODES[errType];
    if (oxErrorCode) {
      if (oxErrorCode === ERROR_CODES.ASSERT && (module === 'verify' || cmd.indexOf('verify') === 0)) {
        return new _OxygenError.default(ERROR_CODES.VERIFY, err.message, null, false);
      }
      return new _OxygenError.default(oxErrorCode, err.message, null, true, err);
    }
    oxErrorCode = SCRIPT_ERROR_CODES[errType];
    if (oxErrorCode) {
      if (oxErrorCode === ERROR_CODES.ASSERT && (module === 'verify' || cmd.indexOf('verify') === 0)) {
        return new _OxygenError.default(ERROR_CODES.VERIFY, err.message, null, false);
      }
      return new _OxygenError.default(oxErrorCode, err.message || null, null, true, err);
    }
    console.log('=== Error Details ===');
    console.log('Type: ' + err.type + ' Name: ' + err.name + ' Code: ' + err.code + ' Msg: ' + err.message);
    console.log(util.inspect(err));
    const errMessage = err.message ? `${errType}: ${err.message}` : `${errType}`;
    return new _OxygenError.default(ERROR_CODES.UNKNOWN_ERROR, errMessage, util.inspect(err), true, err);
  },
  getSeleniumInitError: function (err) {
    if (err.message) {
      var ieZoomErrorMsg = err.message.match(/(Unexpected error launching Internet Explorer\. Browser zoom level was set to \d+%\. It should be set to \d+%)/gm);
      if (ieZoomErrorMsg) {
        return new _OxygenError.default(ERROR_CODES.BROWSER_CONFIGURATION_ERROR, ieZoomErrorMsg.toString());
      }
      if (err.message.includes('Unable to create new service: ChromeDriverService')) {
        return new _OxygenError.default(ERROR_CODES.CHROMEDRIVER_ERROR, err.message, null, true, err);
      }
      if (err.message.includes('Unable to create new service:')) {
        return new _OxygenError.default(ERROR_CODES.SELENIUM_UNREACHABLE_ERROR, err.message, null, true, err);
      }
      var chromeDriverMsg = err.message.match(/(This version of ChromeDriver only supports Chrome version [0-9]*)/gm);
      if (chromeDriverMsg) {
        return new _OxygenError.default(ERROR_CODES.CHROMEDRIVER_ERROR, chromeDriverMsg.toString());
      }
    }
    if (err.message && err.message.includes('cannot find Chrome binary')) {
      return new _OxygenError.default(ERROR_CODES.CHROMEDRIVER_ERROR, 'Cannot find Chrome binary');
    } else if (err.code === 'ECONNREFUSED' || err.code === 'ECONNRESET' || err.code === 'ENOTFOUND' || err.message === 'Failed to create session.\nsocket hang up') {
      return new _OxygenError.default(ERROR_CODES.SELENIUM_UNREACHABLE_ERROR, "Couldn't connect to Selenium server");
    } else if (err.message === 'All minutes for this organization has been exausted' || err.message === '401 Unauthorized' || err.message && err.message.startsWith('Failed to create session.')) {
      return new _OxygenError.default(ERROR_CODES.SELENIUM_CONNECTION_ERROR, err.message);
    } else if (err.message.includes('Unable to create new service:')) {
      return new _OxygenError.default(ERROR_CODES.SELENIUM_CONNECTION_ERROR, err.message + '\n\nThis is probably due to missing ChromeDriver/IEDriverServer/GeckoDriver binary.\n');
    }
    console.log('=== Error Details ===');
    console.log('Type: ' + err.type + ' Name: ' + err.name + ' Code: ' + err.code + ' Msg: ' + err.message);
    console.log(util.inspect(err));
    return new _OxygenError.default(ERROR_CODES.UNKNOWN_ERROR, err.type + ': ' + err.message, util.inspect(err));
  },
  getAppiumInitError: function (err) {
    if (err.message && err.message.includes('cannot find Chrome binary')) {
      return new _OxygenError.default(ERROR_CODES.CHROMEDRIVER_ERROR, 'Cannot find Chrome binary');
    } else if (err.code === 'ECONNREFUSED' || err.code === 'ECONNRESET' || err.code === 'ENOTFOUND' || err.message === 'Failed to create session.\nsocket hang up') {
      return new _OxygenError.default(ERROR_CODES.APPIUM_UNREACHABLE_ERROR, "Couldn't connect to Appium server");
    } else if (err.message && err.message.includes('Could not find a connected Android device')) {
      return new _OxygenError.default(ERROR_CODES.DEVICE_NOT_FOUND, 'Could not find a connected Android device');
    } else if (err.message && err.message.includes('Unable to automate Chrome version')) {
      return new _OxygenError.default(ERROR_CODES.CHROMEDRIVER_ERROR, extractOriginalError(err.message));
    } else if (err.message && err.message.includes('No Chromedriver found that can automate')) {
      return new _OxygenError.default(ERROR_CODES.CHROMEDRIVER_ERROR, extractOriginalError(err.message));
    } else if (err.message && err.message.includes('Unable to find an active device or emulator with')) {
      return new _OxygenError.default(ERROR_CODES.DEVICE_NOT_FOUND, extractOriginalError(err.message));
    } else if (err.message && err.message.includes('is not installed on device')) {
      return new _OxygenError.default(ERROR_CODES.APPLICATION_NOT_FOUND_ERROR, extractOriginalError(err.message));
    } else if (err.message === 'All minutes for this organization has been exausted' || err.message === '401 Unauthorized') {
      return new _OxygenError.default(ERROR_CODES.APPIUM_CONNECTION_ERROR, err.message);
    } else if (err.message && (err.message.includes('A new session could not be created.') || err.message.includes('Failed to create session.'))) {
      return new _OxygenError.default(ERROR_CODES.APPIUM_SESSION_ERROR, extractOriginalError(err.message), null, true);
    }
    console.log('=== Error Details ===');
    console.log('Type: ' + err.type + ' Name: ' + err.name + ' Code: ' + err.code + ' Msg: ' + err.message);
    console.log(util.inspect(err));
    return new _OxygenError.default(ERROR_CODES.UNKNOWN_ERROR, extractOriginalError(err.message), util.inspect(err));
  },
  getAssertError: function (expected, actual, isFatal = true) {
    actual = actual.toString().replace(/\n/g, '\\n');
    return new _OxygenError.default(ERROR_CODES.ASSERT_ERROR, `Expected: "${expected}". Got: "${actual}"`, undefined, isFatal);
  },
  getVerifyError: function (expected, actual) {
    actual = actual.toString().replace(/\n/g, '\\n');
    return new _OxygenError.default(ERROR_CODES.VERIFY_ERROR, `Expected: "${expected}". Got: "${actual}"`, undefined, false);
  },
  errorCode: ERROR_CODES
};
const ORIGINAL_ERROR_MESSAGE = 'Original error: ';
const UNKNOWN_ERROR_MESSAGE = 'unknown error: ';
function extractOriginalError(msg) {
  if (!msg) {
    return '';
  }
  var errorIndex1 = msg.indexOf(ORIGINAL_ERROR_MESSAGE);
  if (errorIndex1 === -1) {
    return msg;
  }
  var errorIndex2 = msg.indexOf(ORIGINAL_ERROR_MESSAGE, errorIndex1 + ORIGINAL_ERROR_MESSAGE.length);
  msg = msg.substring((errorIndex2 > -1 ? errorIndex2 : errorIndex1) + ORIGINAL_ERROR_MESSAGE.length);
  var unknownErrorIndex = msg.indexOf(UNKNOWN_ERROR_MESSAGE);
  if (unknownErrorIndex > -1) {
    return msg.substring(unknownErrorIndex + UNKNOWN_ERROR_MESSAGE.length);
  }
  return msg;
}
