// oxygen-cli/build/errors/OxygenError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _stackTrace = _interopRequireDefault(require("stack-trace"));
var stackTraceParser = _interopRequireWildcard(require("stacktrace-parser"));
var _fs = _interopRequireDefault(require("fs"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const STACKTRACE_FILTERS = ['\\node_modules\\', '/node_modules/', '/oxygen-node/', '\\oxygen-node\\', '(module.js', '(internal/module.js', 'at <anonymous>', 'internal/', 'internal\\'];
class OxygenError extends Error {
  constructor(type, message, data, isFatal, orgErr = null) {
    super(message || undefined);
    this.type = type || this.type || null;
    this.orgErr = orgErr;
    if (orgErr && typeof orgErr === 'string' && type !== orgErr) {
      this.subtype = orgErr;
    } else if (orgErr && typeof orgErr === 'object' && orgErr.type) {
      this.subtype = orgErr.type;
    }
    this.message = message || null;
    this.data = data || null;
    this.screenshot = null;
    this.isFatal = typeof isFatal === 'undefined' ? true : isFatal;
    if (orgErr && orgErr.stack) {
      this.stack = orgErr.stack;
    }
    if (type || message) {
      this.captureStackTrace();
      this.filterStackTrace();
      this.generateLocation();
    }
  }
  filterStackTrace() {
    if (!this.stack) {
      this.captureStackTrace();
    }
    this.stack = this.stack.split('\n').filter(this._stackTraceFilterFn).join('\n');
  }
  _stackTraceFilterFn(value) {
    return !STACKTRACE_FILTERS.some(filter => value.includes(filter));
  }
  generateLocation() {
    let anotherFile;
    let anotherLineNumber;
    let anotherColumn;
    if (this.orgErr && this.orgErr.stack) {
      const anotherStack = stackTraceParser.parse(this.orgErr.stack);
      if (anotherStack && Array.isArray(anotherStack) && anotherStack.length > 0) {
        for (let i = 0; i < anotherStack.length; i++) {
          if (_fs.default.existsSync(anotherStack[i]['file']) && anotherStack[i]['lineNumber'] && !anotherFile && !anotherLineNumber) {
            let file = anotherStack[i];
            if (anotherStack[i] && anotherStack[1 + i] && anotherStack[i]['file'] === anotherStack[1 + i]['file']) {
              file = anotherStack[1 + i];
            }
            anotherFile = file['file'];
            anotherLineNumber = file['lineNumber'];
            anotherColumn = file['column'] || 1;
          }
        }
      }
    }
    let stackTrace = _stackTrace.default.parse(this) || [];
    stackTrace = stackTrace.filter(item => {
      const exist = _fs.default.existsSync(item.fileName);
      return exist;
    });
    if (stackTrace.length > 0) {
      const call = stackTrace[0];
      if (call && call.fileName === 'vm.js' && this.orgErr && this.orgErr.stack) {
        let location = this.orgErr.stack.split('\n')[0];
        const locationSplit = location.split(':');
        if (locationSplit && locationSplit[0]) {
          if (parseInt(locationSplit[locationSplit.length - 1]) && parseInt(locationSplit[locationSplit.length - 2])) {} else if (parseInt(locationSplit[locationSplit.length - 1])) {
            location += ':0';
          } else {
            location += ':0:0';
          }
        }
        this.location = location;
        this.stacktrace = [location, ...stackTrace.map(call => `${this.patchFilePathOnWindows(call.getFileName())}:${call.getLineNumber()}:${call.getColumnNumber()}`)];
      } else if (anotherFile && anotherLineNumber && anotherColumn) {
        this.location = `${this.patchFilePathOnWindows(anotherFile)}:${anotherLineNumber}:${anotherColumn}`;
        this.stacktrace = [this.location];
      } else {
        this.location = `${this.patchFilePathOnWindows(call.getFileName())}:${call.getLineNumber()}:${call.getColumnNumber()}`;
        this.stacktrace = stackTrace.map(call => `${this.patchFilePathOnWindows(call.getFileName())}:${call.getLineNumber()}:${call.getColumnNumber()}`);
      }
    } else if (anotherFile && anotherLineNumber && anotherColumn) {
      this.location = `${this.patchFilePathOnWindows(anotherFile)}:${anotherLineNumber}:${anotherColumn}`;
      this.stacktrace = [this.location];
    } else {
      this.location = null;
    }
  }
  patchFilePathOnWindows(filePath) {
    if (filePath && process.platform === 'win32' && typeof filePath === 'string' && filePath.length > 0) {
      return filePath.replace(/\//g, '\\');
    }
    return filePath;
  }
  captureStackTrace() {
    if (this.stack) {
      return;
    } else {
      try {
        var err = new Error();
        Error.captureStackTrace(err, OxygenError);
        this.stack = err.stack;
      } catch (e) {
        console.error(e.message);
      }
    }
  }
}
exports.default = OxygenError;
