// oxygen-cli/build/core/OxygenCore.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _glob = _interopRequireDefault(require("glob"));
var _path = _interopRequireDefault(require("path"));
var _fs = _interopRequireDefault(require("fs"));
var _deasync = _interopRequireDefault(require("deasync"));
var _future = _interopRequireDefault(require("fibers/future"));
var _fibers = _interopRequireDefault(require("fibers"));
var _os = require("os");
var _stepResult = _interopRequireDefault(require("../model/step-result"));
var _OxygenEvents = _interopRequireDefault(require("./OxygenEvents"));
var _util = _interopRequireDefault(require("../lib/util"));
var coreUtils = _interopRequireWildcard(require("./utils"));
var _OxygenError = _interopRequireDefault(require("../errors/OxygenError"));
var _helper = _interopRequireDefault(require("../errors/helper"));
var _status = _interopRequireDefault(require("../model/status.js"));
var Modules = _interopRequireWildcard(require("../ox_modules/index"));
var _logger2 = _interopRequireWildcard(require("../lib/logger"));
var _OxygenSubModule = require("./OxygenSubModule");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const {
  v1
} = require('uuid');
Object.defineProperty(global, '__stack', {
  get: function () {
    var orig = Error.prepareStackTrace;
    Error.prepareStackTrace = function (_, stack) {
      return stack;
    };
    var err = new Error();
    Error.captureStackTrace(err);
    var stack = err.stack;
    Error.prepareStackTrace = orig;
    return stack;
  }
});
const DEFAULT_TIMEOUT = 30000;
const DEFAULT_OPTS = {
  backtrace: false,
  compiler: [],
  failAmbiguousDefinitions: false,
  failFast: false,
  ignoreUndefinedDefinitions: false,
  name: [],
  profile: [],
  require: [],
  order: 'defined',
  snippetSyntax: undefined,
  snippets: true,
  source: true,
  strict: false,
  tagExpression: '',
  tagsInTitle: false,
  timeout: DEFAULT_TIMEOUT
};
const MODULE_NAME_MATCH_REGEX = /^module-(.+?)\.js$/;
const SERVICE_NAME_MATCH_REGEX = /^service-(.+?)\.js$/;
const DO_NOT_WRAP_METHODS = ['driver', 'getDriver'];
const DEFAULT_CTX = {
  params: {},
  vars: {},
  env: {},
  caps: {},
  attributes: {}
};
const DEFAULT_RESULT_STORE = {
  steps: [],
  logs: [],
  attachments: [],
  har: null,
  attributes: {}
};
class Oxygen extends _OxygenEvents.default {
  constructor() {
    super();
    this.isInitialized = false;
    this.resultStore = {
      ...DEFAULT_RESULT_STORE
    };
    this.ctx = {
      ...DEFAULT_CTX
    };
    this.modules = {};
    this.services = {};
    this.repo = {};
    this.capabilities = null;
    this.opts = null;
    this.oxBaseDir = _path.default.join(__dirname, '../');
    this.logger = this._wrapLogger((0, _logger2.default)('Oxygen'));
    this._waitStepResultList = [];
  }
  async init(options, caps, ctx = {}, results = {}) {
    this.opts = Object.assign(DEFAULT_OPTS, options);
    this.cwd = this.opts.cwd || process.cwd();
    this.ctx = Object.assign(DEFAULT_CTX, ctx || {});
    this.ctx.caps = {
      ...(ctx.caps || {}),
      ...caps
    };
    this.resultStore = Object.assign(DEFAULT_RESULT_STORE, results || {});
    this.capabilities = this.ctx.caps = caps;
    if (ctx.attributes && (!results.attributes || !Object.keys(results.attributes).length)) {
      this.resultStore.attributes = {
        ...ctx.attributes
      };
    }
    this.makeOxGlobal();
    this._loadServices();
    this._loadModules();
    if (typeof options.globalScope === 'undefined' || options.globalScope === true) {
      this.makeContextAndModulesGlobal();
    }
    this.isInitialized = true;
  }
  async disposeModules(status = null) {
    try {
      await this._disposeModules(status);
    } catch (e) {
      console.error('Failed to dispose modules: ', e);
      throw e;
    }
  }
  async dispose(status = null) {
    try {
      await this._disposeModules(status);
      await this._disposeServices();
      this.isInitialized = false;
    } catch (e) {
      this.isInitialized = false;
      console.error('Failed to dispose: ', e);
      throw e;
    }
  }
  get adjustScriptLine() {
    const isInDebugMode = _util.default.isInDebugMode();
    const result = isInDebugMode ? -1 : 0;
    return result;
  }
  get context() {
    return this.ctx;
  }
  set context(ctx) {
    this.ctx = ctx;
    if (global.ox) {
      global.ox.ctx = ctx;
      if (typeof this.opts.globalScope === 'undefined' || this.opts.globalScope === true) {
        global.params = ctx.params || {};
        global.env = ctx.env || {};
        global.attributes = ctx.attributes || {};
      }
    }
  }
  get repository() {
    return this.repo;
  }
  set repository(repo) {
    this.repo = repo;
    if (global.ox) {
      global.ox.repo = repo;
    }
  }
  get results() {
    for (const transStep of this.resultStore.steps) {
      if (transStep.name.includes('.transaction')) {
        const transName = transStep.transaction;
        transStep.duration = 0;
        for (const step of this.resultStore.steps) {
          if (step.transaction === transName && step !== transStep) {
            transStep.duration += step.duration;
            if (step.endTime > transStep.endTime) {
              transStep.endTime = step.endTime;
            }
            if (step.status === _status.default.WARNING) {
              transStep.status = _status.default.WARNING;
            }
          }
        }
      }
    }
    return this.resultStore;
  }
  getModulesCapabilities() {
    if (!this.modules) {
      return null;
    }
    const modCaps = {};
    for (let moduleName in this.modules) {
      const module = this.modules[moduleName];
      if (module.name && module._getCapabilities && typeof module._getCapabilities === 'function') {
        const caps = module._getCapabilities();
        if (caps && typeof caps === 'object' && Object.keys(caps).length > 0) {
          modCaps[module.name] = caps;
        }
      }
    }
    return modCaps;
  }
  resetResults() {
    this.resultStore.steps = [];
    this.resultStore.logs = [];
    this.resultStore.attachments = [];
    this.har = null;
  }
  async onBeforeCase(context) {
    for (let moduleName in this.modules) {
      const module = this.modules[moduleName];
      if (!module) {
        continue;
      }
      try {
        module.onBeforeCase && (await module.onBeforeCase(context));
        module._iterationStart && (await module._iterationStart());
      } catch (e) {
        this.logger.error(`Failed to call "onBeforeCase" method of ${moduleName} module.`, e);
      }
    }
  }
  async onAfterCase(error = null) {
    for (let moduleName in this.modules) {
      const module = this.modules[moduleName];
      if (!module) {
        continue;
      }
      try {
        module.onAfterCase && (await module.onAfterCase(error));
        module._iterationEnd && (await module._iterationEnd(error));
      } catch (e) {
        this.logger.error(`Failed to call "onAfterCase" method of ${moduleName} module.`, e);
      }
    }
  }
  makeOxGlobal() {
    if (!global.ox) {
      global.ox = {
        modules: this.modules,
        ctx: this.ctx,
        options: this.opts,
        caps: this.capabilities,
        resultStore: this.resultStore,
        addAttribute: this.addAttribute.bind(this),
        addAttachment: this.addAttachment.bind(this)
      };
      global.vars = this.ctx.vars;
    }
  }
  makeContextAndModulesGlobal() {
    if (global.ox) {
      for (let moduleName in this.modules) {
        if (!global[moduleName]) {
          global[moduleName] = this.modules[moduleName];
        }
      }
      global.params = global.ox.ctx.params;
      global.env = global.ox.ctx.env;
      global.ctx = global.ox.ctx;
      global.attributes = global.ox.ctx.attributes || {};
    }
  }
  addAttribute(name, value) {
    if (!name || !value) {
      throw new _OxygenError.default(_helper.default.errorCode.SCRIPT_ERROR, "'name' and 'value' arguments are required for ox.addAttribute");
    }
    if (this.resultStore && this.resultStore.attributes) {
      this.resultStore.attributes[name] = value;
    }
  }
  addAttachment(filePath) {
    if (!this.resultStore) {
      return;
    }
    if (!_fs.default.existsSync(filePath)) {
      console.err(`File ${filePath} doesn't exist`);
      return;
    }
    if (!this.resultStore.attachments) {
      this.resultStore.attachments = [];
    }
    this.resultStore.attachments.push(coreUtils.newFileAttachment(filePath));
  }
  loadPageObjectFile(poPath) {
    if (!poPath) {
      return;
    }
    try {
      const po = require(poPath);
      this.repository = po;
      if (typeof this.opts.globalScope === 'undefined' || this.opts.globalScope === true) {
        global.po = po;
      }
    } catch (e) {
      if (typeof this.opts.globalScope === 'undefined' || this.opts.globalScope === true) {
        global.po = {};
      }
    }
  }
  _wrapLogger(_logger) {
    const loggerWrap = {
      info: (...args) => this._log(_logger, 'info', args),
      debug: (...args) => this._log(_logger, 'debug', args),
      error: (...args) => this._log(_logger, 'error', args),
      warn: (...args) => this._log(_logger, 'warn', args),
      userInfo: (...args) => this._log(_logger, 'info', args, _logger2.ISSUERS.USER),
      userDebug: (...args) => this._log(_logger, 'debug', args, _logger2.ISSUERS.USER),
      userError: (...args) => this._log(_logger, 'error', args, _logger2.ISSUERS.USER),
      userWarn: (...args) => this._log(_logger, 'warn', args, _logger2.ISSUERS.USER)
    };
    return loggerWrap;
  }
  _log(_logger, level, args, src = _logger2.DEFAULT_LOGGER_ISSUER) {
    if (!_logger[level]) {
      return;
    }
    _logger[level].apply(_logger, args);
    const message = _util.default.stringify(args, 2);
    const time = _util.default.getTimeStamp();
    if (this.resultStore && this.resultStore.logs) {
      this.resultStore.logs.push({
        time: time,
        level: level.toUpperCase(),
        msg: message,
        src: src
      });
    }
    args = message ? args.shift() : args;
    this.emitLog(time, level, message, args, src);
  }
  _loadServices() {
    const oxServicesDirPath = _path.default.resolve(this.oxBaseDir, './ox_services');
    const serviceFiles = _glob.default.sync('service-*.js', {
      cwd: oxServicesDirPath
    });
    this.logger.debug('Loading services...');
    for (var i = 0; i < serviceFiles.length; i++) {
      const serviceFileName = serviceFiles[i];
      const serviceFilePath = _path.default.join(oxServicesDirPath, serviceFileName);
      const result = serviceFileName.match(SERVICE_NAME_MATCH_REGEX);
      const serviceName = result[1];
      if (Array.isArray(this.opts.services) && !this.opts.services.includes(serviceName)) {
        continue;
      }
      try {
        this.logger.debug('Loading service: ' + serviceName);
        const service = this._loadService(serviceName, serviceFilePath);
        service.init();
        this.services[serviceName] = service;
      } catch (e) {
        this.logger.error('Error initializing service "' + serviceName + '": ' + e.message + _os.EOL + (e.stacktrace ? e.stacktrace : ''));
      }
    }
  }
  _loadService(serviceName, servicePath) {
    let ServiceClass = require(servicePath);
    if (ServiceClass.default) {
      ServiceClass = ServiceClass.default;
    }
    const serviceLogger = this._wrapLogger((0, _logger2.default)(`Service:${serviceName}`));
    return new ServiceClass(this.opts, this.ctx, this.resultStore, serviceLogger);
  }
  _loadModules() {
    this.logger.debug('Loading internal modules...');
    this._loadInternalModules();
    this.logger.debug('Loading external modules...');
    this._loadExternalModules();
  }
  _loadExternalModules() {
    let modulesFolderPath = './modules';
    if (this.opts.modules_ext && typeof this.opts.modules_ext === 'string') {
      modulesFolderPath = this.opts.modules_ext;
    }
    const oxModulesDirPath = _path.default.resolve(this.cwd, modulesFolderPath);
    if (!_fs.default.existsSync(oxModulesDirPath)) {
      return false;
    }
    let moduleFiles = [];
    if (Array.isArray(this.opts.modules) && this.opts.modules.length > 0) {
      this.opts.modules.forEach(moduleName => moduleFiles.push(`module-${moduleName}.js`));
    } else {
      moduleFiles = _glob.default.sync('module-*.js', {
        cwd: oxModulesDirPath
      });
    }
    for (let moduleFileName of moduleFiles) {
      const moduleName = moduleFileName.match(MODULE_NAME_MATCH_REGEX)[1];
      try {
        const startTime = new Date();
        const moduleLogger = this._wrapLogger((0, _logger2.default)(`Module:${moduleName}`));
        const mod = coreUtils.loadModuleFromFile(moduleName, moduleFileName, moduleLogger, oxModulesDirPath, this);
        this._callServicesOnModuleLoaded(mod);
        this.modules[moduleName] = global.ox.modules[moduleName] = this._wrapModule(moduleName, mod);
        const endTime = new Date();
        const duration = (endTime - startTime) / 1000;
        this.logger.debug('Loading module: ' + moduleName + ' [ ' + duration + ' sec ]');
      } catch (e) {
        this.logger.error('Error initializing module "' + moduleName + '": ' + e.message + _os.EOL + (e.stacktrace ? e.stacktrace : ''));
        if (moduleName === 'web' || moduleName === 'mob') {
          break;
        }
      }
    }
  }
  _loadInternalModules() {
    const excludeUndefinedMods = this.opts.modules && this.opts.modules.length > 0;
    for (let moduleName of Object.keys(Modules)) {
      if (excludeUndefinedMods && !this.opts.modules.includes(moduleName)) {
        continue;
      }
      const ModuleClass = Modules[moduleName];
      const oxModulesDirPath = _util.default.getOxModulesDir();
      try {
        const startTime = new Date();
        const moduleLogger = this._wrapLogger((0, _logger2.default)(`Module:${moduleName}`));
        const mod = coreUtils.loadModuleFromClass(moduleName, ModuleClass, moduleLogger, oxModulesDirPath, this);
        this._callServicesOnModuleLoaded(mod);
        this.modules[moduleName] = global.ox.modules[moduleName] = this._wrapModule(moduleName, mod);
        const endTime = new Date();
        const duration = (endTime - startTime) / 1000;
        this.logger.debug('Loading module: ' + moduleName + ' [ ' + duration + ' sec ]');
      } catch (e) {
        this.logger.error('Error initializing module "' + moduleName + '": ' + e.message + _os.EOL + (e.stacktrace ? e.stacktrace : ''));
        if (moduleName === 'web' || moduleName === 'mob') {
          break;
        }
      }
    }
  }
  _wrapModule(name, module) {
    if (!module) {
      return undefined;
    }
    const wrapper = {
      name: name
    };
    const _this = this;
    let moduleMethods = Object.keys(module);
    if (!moduleMethods.some(x => x === 'exports')) {
      moduleMethods = this._getAllPropertyNames(module);
    }
    for (let methodName of moduleMethods) {
      const method = module[methodName];
      if (!(method instanceof Function) && !(method instanceof _OxygenSubModule.OxygenSubModule) && typeof method !== 'function' || methodName === 'exports' || methodName === '_compile') {
        continue;
      }
      if (method instanceof _OxygenSubModule.OxygenSubModule) {
        wrapper[methodName] = this._wrapModule(`${name}.${methodName}`, method);
      } else if (DO_NOT_WRAP_METHODS.includes(methodName)) {
        wrapper[methodName] = module[methodName].bind(module);
      } else if (methodName.indexOf('_') === 0 || methodName.indexOf('on') === 0) {
        wrapper[methodName] = function () {
          var args = Array.prototype.slice.call(arguments);
          _this.logger.debug('Executing: ' + _util.default.getMethodSignature(name, methodName, args));
          try {
            return module[methodName].apply(module, args);
          } catch (e) {
            throw _helper.default.getOxygenError(e, name, methodName, args);
          }
        };
      } else {
        if (methodName === 'getCapabilities') {
          wrapper['_getCapabilities'] = method.bind(module);
        }
        wrapper[methodName] = (...args) => {
          try {
            return _this._commandWrapper(methodName, args, module, name);
          } catch (e) {
            if (e instanceof _OxygenError.default) {
              throw e;
            }
            throw _helper.default.getOxygenError(e, name, methodName, args);
          }
        };
      }
    }
    return wrapper;
  }
  _commandWrapper(cmdName, cmdArgs, module, moduleName) {
    if (!module || !module[cmdName]) {
      return undefined;
    }
    let retval = null;
    let error = null;
    const cmdFn = module[cmdName];
    const publicMethod = !cmdName.startsWith('_') && cmdName !== 'dispose';
    if (this.opts.delay && publicMethod && cmdName !== 'init' && cmdName !== 'transaction') {
      _deasync.default.sleep(this.opts.delay * 1000);
    }
    if (!module.isInitialized && publicMethod && cmdName !== 'init' && cmdName !== 'transaction') {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, 'Missing ' + moduleName + '.init()');
    }
    let decryptedArgs = this._getDecryptedArgsForApply(cmdArgs, moduleName);
    decryptedArgs = this._populateParametersValue(decryptedArgs);
    cmdArgs = this._populateParametersValue(cmdArgs);
    const startTime = _util.default.getTimeStamp();
    let cmdLocation = null;
    const stepResultId = v1();
    if (publicMethod) {
      cmdLocation = this._getCommandLocation();
      this.emitBeforeCommand(stepResultId, cmdName, moduleName, cmdFn, cmdArgs, this.ctx, cmdLocation, startTime);
    }
    this.logger.debug('Executing: ' + _util.default.getMethodSignature(moduleName, cmdName, cmdArgs));
    try {
      if (cmdName === 'dispose') {
        this._wrapAsync(this._callServicesOnModuleWillDispose).apply(this, [module]);
      }
      const retvalPromise = this._wrapAsync(module[cmdName]).apply(module, decryptedArgs);
      if (retvalPromise && retvalPromise.then) {
        let promiseDone = false;
        retvalPromise.then(value => {
          retval = value;
          promiseDone = true;
        }, e => {
          error = e;
          promiseDone = true;
        });
        _deasync.default.loopWhile(() => !promiseDone);
      } else {
        retval = retvalPromise;
      }
      if (cmdName === 'init') {
        this._wrapAsync(this._callServicesOnModuleInitialized).apply(this, [module]);
      }
    } catch (e) {
      if (e && e.message && typeof e.message === 'string' && e.message.includes('invalid session id')) {
        return;
      } else {
        if (module && typeof module.isInitialized === 'boolean' && !module.isInitialized && cmdName !== 'init') {
          return;
        }
        error = _helper.default.getOxygenError(e, moduleName, cmdName, cmdArgs);
      }
    }
    const endTime = _util.default.getTimeStamp();
    let stepResult;
    let done = false;
    if (publicMethod) {
      const waitId = +new Date();
      this._waitStepResultList.push(waitId);
      stepResult = this._getStepResult(module, moduleName, cmdName, cmdArgs, cmdLocation, startTime, endTime, retval, error);
      stepResult.id = stepResultId;
      const index = this._waitStepResultList.indexOf(waitId);
      this._waitStepResultList.splice(index, 1);
      this.resultStore.steps.push(stepResult);
      this.emitAfterCommand(cmdName, moduleName, cmdFn, cmdArgs, this.ctx, cmdLocation, endTime, stepResult);
      done = true;
    }
    if (error && error.isFatal && !this.opts.continueOnError) {
      if (!error.location && cmdLocation) {
        error.location = cmdLocation;
      }
      if (stepResult && stepResult.failure && stepResult.failure.location) {
        error.location = stepResult.failure.location;
      }
      throw error;
    }
    if (!publicMethod) {
      done = true;
    }
    _deasync.default.loopWhile(() => !done && !error);
    return retval;
  }
  _wrapAsync(fn, context) {
    return function (...args) {
      var self = context || this;
      if (!_fibers.default.current) {
        const retval = fn.apply(self, args);
        let done = false;
        let error = null;
        let finalVal = null;
        if (retval && retval.then) {
          Promise.resolve(retval).then(val => {
            finalVal = val;
            done = true;
          }).catch(e => {
            error = e;
            done = true;
          });
        } else {
          finalVal = retval;
          done = true;
        }
        try {
          _deasync.default.loopWhile(() => !done && !error);
        } catch (e) {
          if (e && e.message && typeof e.message === 'string' && e.message.includes('readyState')) {
            return undefined;
          }
          this.logger.error('deasync.loopWhile() failed:', e);
          return undefined;
        }
        if (!error) {
          return finalVal;
        }
        throw error;
      }
      let error = null;
      let done = false;
      let retval = null;
      try {
        const future = new _future.default();
        var result = fn.apply(self, args);
        if (result && typeof result.then === 'function') {
          result.then(val => future.return(val), err => future.throw(err));
          return future.wait();
        }
        return result;
      } catch (e) {
        error = e;
      }
      _deasync.default.loopWhile(() => !done && !error);
      if (!error) {
        return retval;
      }
      throw error;
    };
  }
  _getCommandLocation() {
    const stack = __stack;
    let caller = null;
    for (let call of stack) {
      if (__filename === call.getFileName()) {
        continue;
      }
      caller = call;
      break;
    }
    if (caller) {
      return `${caller.getFileName()}:${caller.getLineNumber() + this.adjustScriptLine}:${caller.getColumnNumber()}`;
    }
    return null;
  }
  _getStepResult(module, moduleName, methodName, args, location, startTime, endTime, retval, err) {
    var step = new _stepResult.default();
    const displayName = module.getStepDisplayName && module.getStepDisplayName(methodName, args, retval, err) || _util.default.getMethodSignature(moduleName, methodName, args);
    step.name = displayName;
    step.transaction = global._lastTransactionName || null;
    step.location = location;
    step.extra = {
      ...module.extra
    };
    if (module.extra && module.extra.http) {
      step.type = 'http';
    }
    if (err && err.type && err.type === _helper.default.errorCode.ASSERT_PASSED) {
      step.status = _status.default.PASSED;
    } else {
      if (err) {
        if (err.isFatal) {
          step.status = _status.default.FAILED;
        } else {
          step.status = _status.default.WARNING;
        }
      } else if (moduleName === 'log' && methodName === 'warn') {
        step.status = _status.default.WARNING;
      } else {
        step.status = _status.default.PASSED;
      }
    }
    step.action = typeof module._isAction === 'function' ? module._isAction(methodName) : false;
    step.startTime = startTime;
    step.endTime = endTime;
    step.duration = endTime - startTime;
    if (typeof module._getStats === 'function') {
      step.stats = module._getStats(methodName);
    } else {
      step.stats = {};
    }
    if (err) {
      if (err.type === _helper.default.errorCode.ASSERT_PASSED) {} else {
        step.failure = _helper.default.getFailureFromError(err);
        step.failure.location = location;
        if (typeof module._takeScreenshotSilent === 'function' && !this.opts.disableScreenshot) {
          try {
            step.screenshot = module._takeScreenshotSilent(methodName);
          } catch (e) {
            _deasync.default.sleep(1000);
            try {
              step.screenshot = module._takeScreenshotSilent(methodName);
            } catch (e) {}
          }
        }
      }
    }
    return step;
  }
  async _disposeServices() {
    if (!this.services || typeof this.services !== 'object') {
      return false;
    }
    for (let key in this.services) {
      const service = this.services[key];
      if (service.dispose) {
        try {
          await service.dispose();
        } catch (e) {
          this.logger.error(`Failed to dispose service '${key}':`, e);
        }
      }
    }
    return true;
  }
  async _disposeModules(status = null) {
    if (!this.modules || typeof this.modules !== 'object') {
      return false;
    }
    for (let key in this.modules) {
      const mod = this.modules[key];
      if (mod.dispose) {
        try {
          const disposeResult = mod.dispose(status);
          if (disposeResult && typeof disposeResult.then === 'function') {
            await disposeResult();
          }
        } catch (e) {
          this.logger.error(`Failed to dispose module '${key}': `, e);
        }
      }
    }
    return true;
  }
  _callServicesOnModuleLoaded(module) {
    for (let serviceName in this.services) {
      const service = this.services[serviceName];
      if (!service) {
        continue;
      }
      try {
        service.onModuleLoaded(module);
      } catch (e) {
        this.logger.error(`Failed to call "onModuleLoaded" method of ${serviceName} service.`, e);
      }
    }
  }
  async _callServicesOnModuleInitialized(module) {
    for (let serviceName in this.services) {
      const service = this.services[serviceName];
      if (!service) {
        continue;
      }
      try {
        await service.onModuleInitialized(module);
      } catch (e) {
        this.logger.error(`Failed to call "onModuleInitialized" method of ${serviceName} service.`, e);
      }
    }
  }
  async _callServicesOnModuleWillDispose(module) {
    if (!this || !this.services) {
      return;
    }
    for (let serviceName in this.services) {
      const service = this.services[serviceName];
      if (!service) {
        continue;
      }
      try {
        if (service.onModuleWillDispose) {
          service.onModuleWillDispose(module);
        }
      } catch (e) {
        this.logger.error(`Failed to call "onModuleWillDispose" method of ${serviceName} service.`, e);
      }
    }
  }
  _populateParametersValue(args) {
    if (!args || !Array.isArray(args) || args.length == 0) {
      return args;
    }
    const _newArgs = [];
    for (let arg of args) {
      _newArgs.push(this._replaceParameterInArgument(arg));
    }
    return _newArgs;
  }
  _replaceParameterInArgument(arg) {
    if (arg && typeof arg.getDecryptResult === 'function') {
      return 'ENCRYPTED';
    }
    if (!arg || typeof arg !== 'string') {
      return arg;
    }
    for (let paramName in this.ctx.params) {
      if (this.ctx.params.hasOwnProperty(paramName)) {
        const paramValue = this.ctx.params[paramName];
        arg = arg.replace(new RegExp('\\${' + paramName + '}', 'g'), paramValue);
      }
    }
    for (let envName in this.ctx.env) {
      if (this.ctx.env.hasOwnProperty(envName)) {
        const envValue = this.ctx.env[envName];
        arg = arg.replace(new RegExp('\\${' + envName + '}', 'g'), envValue);
      }
    }
    return arg;
  }
  _getAllPropertyNames(obj) {
    const props = [];
    do {
      Object.getOwnPropertyNames(obj).forEach(function (prop) {
        if (props.indexOf(prop) === -1) {
          props.push(prop);
        }
      });
    } while (obj = Object.getPrototypeOf(obj));
    return props;
  }
  _checkStepResult(resolve, reject) {
    if (this._waitStepResultList.length === 0) {
      resolve();
    } else {
      this._checkStepResult(resolve, reject);
    }
  }
  _waitStepResult() {
    return new Promise((resolve, reject) => {
      this._checkStepResult(resolve, reject);
    });
  }
  _getDecryptedArgsForApply(args, moduleName) {
    if (!args || !Array.isArray(args) || args.length == 0) {
      return args;
    }
    const _newArgs = [];
    for (let arg of args) {
      _newArgs.push(this._decryptParameterForApplyInArgument(arg, moduleName));
    }
    return _newArgs;
  }
  _decryptParameterForApplyInArgument(arg, moduleName) {
    if (arg && typeof arg.getDecryptResult === 'function') {
      if (moduleName === 'log') {
        return 'ENCRYPTED';
      } else {
        return arg.getDecryptResult();
      }
    }
    return arg;
  }
}
exports.default = Oxygen;
