// oxygen-cli/build/core/OxygenEvents.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _events = require("events");
var _util = _interopRequireDefault(require("../lib/util"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class OxygenEvents extends _events.EventEmitter {
  constructor() {
    super();
  }
  emitBeforeCommand(stepResultId, cmdName, moduleName, cmdFn, cmdArgs, ctx, location, startTime) {
    this.emit('command:before', {
      id: stepResultId,
      name: cmdName,
      module: moduleName,
      args: cmdArgs,
      signature: _util.default.getMethodSignature(moduleName, cmdName, cmdArgs),
      ctx: ctx,
      time: startTime,
      location: location
    });
  }
  emitAfterCommand(cmdName, moduleName, cmdFn, cmdArgs, ctx, location, endTime, result) {
    this.emit('command:after', {
      id: result.id,
      name: cmdName,
      module: moduleName,
      args: cmdArgs,
      signature: _util.default.getMethodSignature(moduleName, cmdName, cmdArgs),
      ctx: ctx,
      time: endTime,
      location: location,
      result: result
    });
  }
  emitLog(time, level, msg, args, src) {
    this.emit('log', {
      time: time,
      level: level,
      message: msg,
      args: args,
      src: src
    });
  }
}
exports.default = OxygenEvents;
