// oxygen-cli/build/core/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadModuleFromClass = loadModuleFromClass;
exports.loadModulesFromFile = loadModulesFromFile;
exports.newFileAttachment = newFileAttachment;
require("source-map-support/register");
var _fs = _interopRequireDefault(require("fs"));
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const {
  v1
} = require('uuid');
function loadModuleFromClass(moduleName, moduleClass, moduleLogger, oxModulesDirPath, args) {
  if (moduleClass.default) {
    moduleClass = moduleClass.default;
  }
  const cmdDirPath = oxModulesDirPath ? _path.default.join(oxModulesDirPath, 'module-' + moduleName, 'commands') : null;
  addCommandsToPrototype(cmdDirPath, moduleName, moduleClass);
  const module = new moduleClass(args.opts, args.ctx, args.resultStore, moduleLogger, args.modules, args.services);
  if (!module.name) {
    module.name = moduleName;
  }
  applyThisInHelpers(module);
  return module;
}
function loadModulesFromFile(moduleName, moduleFileName, moduleLogger, oxModulesDirPath, args) {
  const moduleClass = require(_path.default.join(oxModulesDirPath, moduleFileName));
  const cmdDirPath = oxModulesDirPath ? _path.default.join(oxModulesDirPath, 'module-' + moduleName, 'commands') : null;
  addCommandsToPrototype(cmdDirPath, moduleName, moduleClass);
  const module = new moduleClass(args.opts, args.ctx, args.resultStore, moduleLogger, args.modules, args.services);
  if (!module.name) {
    module.name = moduleName;
  }
  applyThisInHelpers(module);
  return module;
}
function newFileAttachment(filePath) {
  const fileName = _path.default.basename(filePath);
  return {
    id: v1(),
    fileName,
    filePath,
    type: 'other'
  };
}
function applyThisInHelpers(module) {
  if (module.helpers || module._this && module._this.helpers) {
    const helpers = module.helpers || module._this.helpers;
    for (var funcName in helpers) {
      if (typeof helpers[funcName] === 'function') {
        helpers[funcName] = helpers[funcName].bind(module._this || module);
      }
    }
  }
}
function addCommandsToPrototype(cmdDirPath, moduleName, moduleClass) {
  if (cmdDirPath && _fs.default.existsSync(cmdDirPath)) {
    try {
      const commands = require(cmdDirPath);
      for (const [cmdName, cmdFunc] of Object.entries(commands)) {
        moduleClass.prototype[cmdName] = cmdFunc;
      }
    } catch (e) {
      throw new Error(`Cannot load external commands for module "${moduleName}": ` + e.message);
    }
  }
}
