// oxygen-cli/build/runners/cucumber/CucumberEventListener.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compareScenarioLineWithSourceLine = compareScenarioLineWithSourceLine;
exports.default = void 0;
exports.getStepFromFeature = getStepFromFeature;
require("source-map-support/register");
var _events = require("events");
class CucumberEventListener extends _events.EventEmitter {
  constructor(eventBroadcaster) {
    super();
    this.features = {};
    this.gherkinDocEvents = [];
    this.acceptedPickles = [];
    this.currentPickle = null;
    this.currentFeature = null;
    this.currentFeatureUri = null;
    this.testCasePreparedEvents = [];
    eventBroadcaster.on('gherkin-document', this.onGherkinDocument.bind(this)).on('pickle-accepted', this.onPickleAccepted.bind(this)).on('test-case-prepared', this.onTestCasePrepared.bind(this)).on('test-case-started', this.onTestCaseStarted.bind(this)).on('test-step-started', this.onTestStepStarted.bind(this)).on('test-step-finished', this.onTestStepFinished.bind(this)).on('test-case-finished', this.onTestCaseFinished.bind(this)).on('test-run-started', this.onTestRunStarted.bind(this)).on('test-run-finished', this.onTestRunFinished.bind(this));
  }
  onGherkinDocument(gherkinDocEvent) {
    this.gherkinDocEvents.push(gherkinDocEvent);
    const uri = gherkinDocEvent.uri;
    const doc = gherkinDocEvent.document;
    const feature = doc.feature;
    this.emit('feature:ready', uri, feature);
  }
  onPickleAccepted(pickleEvent) {
    this.acceptedPickles.push(pickleEvent);
  }
  onTestCaseStarted(testCaseStartedEvent) {
    const pickleEvent = this.acceptedPickles.shift();
    const uri = pickleEvent.uri;
    const doc = this.gherkinDocEvents.find(gde => gde.uri === uri).document;
    const feature = doc.feature;
    const scenario = pickleEvent.pickle;
    this.currentPickle = scenario;
    if (this.currentFeatureUri && this.currentFeatureUri !== uri) {
      this.emit('feature:after', this.currentFeatureUri, this.currentFeature);
      this.currentFeatureUri = uri;
      this.currentFeature = feature;
      this.emit('feature:before', this.currentFeatureUri, this.currentFeature);
    } else if (!this.currentFeature) {
      this.currentFeature = feature;
      this.currentFeatureUri = uri;
      this.emit('feature:before', this.currentFeatureUri, this.currentFeature);
    }
    this.emit('scenario:before', uri, feature, scenario);
  }
  onTestStepStarted(testStepStartedEvent) {
    const sourceLocation = testStepStartedEvent.testCase.sourceLocation;
    const uri = sourceLocation.uri;
    const doc = this.gherkinDocEvents.find(gde => gde.uri === uri).document;
    const feature = doc.feature;
    const scenario = feature.children.find(child => compareScenarioLineWithSourceLine(child, sourceLocation));
    const step = getStepFromFeature(feature, this.currentPickle, testStepStartedEvent.index, sourceLocation);
    this.emit('step:before', uri, feature, scenario, step, sourceLocation);
  }
  onTestCasePrepared(testCasePreparedEvent) {
    this.testCasePreparedEvents.push(testCasePreparedEvent);
    const sourceLocation = testCasePreparedEvent.sourceLocation;
    const uri = sourceLocation.uri;
    const doc = this.gherkinDocEvents.find(gde => gde.uri === uri).document;
    const scenario = doc.feature.children.find(child => compareScenarioLineWithSourceLine(child, sourceLocation));
    const scenarioHasHooks = scenario.steps.filter(step => step.type === 'Hook').length > 0;
    if (scenarioHasHooks) {
      return;
    }
    const allSteps = testCasePreparedEvent.steps;
    allSteps.forEach((step, idx) => {
      if (!step.sourceLocation) {
        step.sourceLocation = {
          line: step.actionLocation.line,
          column: 0,
          uri: step.actionLocation.uri
        };
        const hook = {
          type: 'Hook',
          location: step.sourceLocation,
          keyword: 'Hook',
          text: ''
        };
        scenario.steps.splice(idx, 0, hook);
      }
    });
  }
  onTestStepFinished(testStepFinishedEvent) {
    const sourceLocation = testStepFinishedEvent.testCase.sourceLocation;
    const uri = sourceLocation.uri;
    const doc = this.gherkinDocEvents.find(gde => gde.uri === uri).document;
    const feature = doc.feature;
    const scenario = feature.children.find(child => compareScenarioLineWithSourceLine(child, sourceLocation));
    const step = getStepFromFeature(feature, this.currentPickle, testStepFinishedEvent.index, sourceLocation);
    const result = testStepFinishedEvent.result;
    this.emit('step:after', uri, feature, scenario, step, result, sourceLocation);
  }
  onTestCaseFinished(testCaseFinishedEvent) {
    const sourceLocation = testCaseFinishedEvent.sourceLocation;
    const uri = sourceLocation.uri;
    const doc = this.gherkinDocEvents.find(gde => gde.uri === uri).document;
    const feature = doc.feature;
    const scenario = feature.children.find(child => compareScenarioLineWithSourceLine(child, sourceLocation));
    this.emit('scenario:after', uri, feature, scenario, sourceLocation);
    this.currentPickle = null;
  }
  onTestRunStarted(testRunStartedEvent) {
    this.emit('test:start');
  }
  onTestRunFinished(testRunFinishedEvent) {
    if (this.currentFeature) {
      this.emit('feature:after', this.currentFeatureUri, this.currentFeature);
      this.currentFeature = null;
    }
    this.emit('test:end', testRunFinishedEvent.result);
  }
}
exports.default = CucumberEventListener;
function attachEventLogger(eventBroadcaster) {
  const EVENTS = ['source', 'attachment', 'gherkin-document', 'pickle', 'pickle-accepted', 'pickle-rejected', 'test-run-started', 'test-case-prepared', 'test-case-started', 'test-step-started', 'test-step-attachment', 'test-step-finished', 'test-case-finished', 'test-run-finished'];
  EVENTS.forEach(e => {
    eventBroadcaster.on(e, x => {
      console.log('\n-----' + e + ' -----\n' + JSON.stringify(x, null, 2));
    });
  });
}
function compareScenarioLineWithSourceLine(scenario, sourceLocation) {
  if (scenario.type.indexOf('ScenarioOutline') > -1) {
    return scenario.examples.some(example => example.tableBody.some(tableEntry => tableEntry.location.line === sourceLocation.line));
  } else {
    return scenario.location.line === sourceLocation.line;
  }
}
function getStepFromFeature(feature, pickle, stepIndex, sourceLocation) {
  let combinedSteps = [];
  feature.children.forEach(child => {
    if (child.type.indexOf('Scenario') > -1 && !compareScenarioLineWithSourceLine(child, sourceLocation)) {
      return;
    }
    combinedSteps = combinedSteps.concat(child.steps);
  });
  const targetStep = combinedSteps[stepIndex];
  if (targetStep.type === 'Step') {
    const stepLine = targetStep.location.line;
    const pickleStep = pickle.steps.find(s => s.locations.some(loc => loc.line === stepLine));
    if (pickleStep) {
      return {
        ...targetStep,
        text: pickleStep.text
      };
    }
  }
  return targetStep;
}
