// oxygen-cli/build/runners/cucumber/CucumberWorker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _fibers = _interopRequireDefault(require("fibers"));
var Cucumber = _interopRequireWildcard(require("cucumber"));
var _isGlob = _interopRequireDefault(require("is-glob"));
var _glob = _interopRequireDefault(require("glob"));
var _sync = require("@wdio/sync");
var _utils = require("@wdio/utils");
var _events = require("events");
var _CucumberEventListener = _interopRequireDefault(require("./CucumberEventListener"));
var _CucumberReporter = _interopRequireDefault(require("./CucumberReporter"));
var _OxygenCore = _interopRequireDefault(require("../../core/OxygenCore"));
var _util = _interopRequireDefault(require("../../lib/util"));
var _status = _interopRequireDefault(require("../../model/status"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
require = require('esm')(module);
var td = require('testdouble');
const DEFAULT_TIMEOUT = 30000;
const DEFAULT_OPTS = {
  backtrace: false,
  compiler: [],
  failAmbiguousDefinitions: false,
  failFast: false,
  ignoreUndefinedDefinitions: false,
  name: [],
  profile: [],
  require: [],
  order: 'defined',
  snippetSyntax: undefined,
  snippets: true,
  source: true,
  strict: false,
  tagExpression: '',
  tagsInTitle: false,
  timeout: DEFAULT_TIMEOUT
};
global.browser = {};
class CucumberWorker {
  constructor(reporter) {
    this.rid = null;
    this.isInitialized = false;
    this.reporter = reporter;
    this.cucumberEventListener = null;
    this.cucumberReporter = null;
    this.oxygen = null;
    this.beforeCommandHandler = this.beforeCommandHandler.bind(this);
    this.afterCommandHandler = this.afterCommandHandler.bind(this);
    this.onBeforeFeature = this.onBeforeFeature.bind(this);
    this.onAfterFeature = this.onAfterFeature.bind(this);
    this.onTestEnd = this.onTestEnd.bind(this);
  }
  async init(rid, config, caps) {
    this.rid = rid;
    this.config = config;
    this.cwd = this.config.cwd || process.cwd();
    this.specs = this.resolveSpecFiles(config.specs || []);
    this.cucumberOpts = Object.assign(DEFAULT_OPTS, config.cucumberOpts);
    this.testHooks = _util.default.loadTestHooks(config);
    this.capabilities = caps;
    await this.initializeOxygenCore();
    this.isInitialized = true;
  }
  async dispose(status = null) {
    this.isInitialized = false;
    if (this.oxygen) {
      try {
        await this.oxygen.dispose(status);
      } catch (e) {
        this._logger.error('Failed to dispose Oxygen', null, e);
      } finally {
        this.oxygen = null;
      }
    }
  }
  async _runFnInFiberContext(fn, args) {
    return new Promise((resolve, reject) => (0, _fibers.default)(() => {
      try {
        const result = fn.apply(this, args || []);
        return resolve(result);
      } catch (err) {
        return reject(err);
      }
    }).run());
  }
  async run({
    scriptPath,
    context,
    poFile = null
  }) {
    Cucumber.supportCodeLibraryBuilder.reset(this.cwd);
    if (context && this.oxygen) {
      this.oxygen.context = context;
      this.oxygen.loadPageObjectFile(poFile);
    } else {
      if (!this.oxygen) {
        console.warn('this.oxygen is undefined');
      } else {
        console.warn('context in runOpts is undefined');
      }
    }
    this.registerCompilers();
    this.loadRequireFiles();
    this.wrapSteps();
    Cucumber.setDefaultTimeout(this.cucumberOpts.timeout);
    const supportCodeLibrary = Cucumber.supportCodeLibraryBuilder.finalize();
    const eventBroadcaster = new _events.EventEmitter();
    this.hookInCucumberEvents(eventBroadcaster);
    this.cucumberReporter = new _CucumberReporter.default(this.rid, this.config, this.cucumberEventListener, this.oxygen, this.reporter, this.testHooks);
    const pickleFilter = new Cucumber.PickleFilter({
      featurePaths: this.specs,
      names: this.cucumberOpts.name,
      tagExpression: this.cucumberOpts.tagExpression
    });
    const testCases = await Cucumber.getTestCasesFromFilesystem({
      cwd: this.cwd,
      eventBroadcaster,
      featurePaths: this.specs.map(spec => spec.replace(/(:\d+)*$/g, '')),
      order: this.cucumberOpts.order,
      pickleFilter
    });
    const runtime = new Cucumber.Runtime({
      eventBroadcaster,
      options: this.cucumberOpts,
      supportCodeLibrary,
      testCases
    });
    let hookError = await this._runFnInFiberContext(this.testHooks['beforeTest'], [this.rid, this.config, this.capabilities]);
    if (hookError) {
      throw hookError;
    }
    let error = null;
    try {
      await runtime.start();
    } catch (e) {
      error = e;
    }
    let testResultStatus = _status.default.PASSED;
    if (this.cucumberReporter && this.cucumberReporter.suites && Object.keys(this.cucumberReporter.suites)) {
      const suites = this.cucumberReporter.suites;
      Object.keys(suites).forEach(function (key) {
        const suite = suites[key];
        if (suite.status === _status.default.FAILED) {
          testResultStatus = _status.default.FAILED;
        }
      });
    }
    this.cucumberReporter.status = testResultStatus;
    hookError = await this._runFnInFiberContext(this.testHooks['afterTest'], [this.rid, this.cucumberReporter, error]);
    if (hookError) {
      throw hookError;
    }
    return testResultStatus;
  }
  async initializeOxygenCore() {
    if (!this.oxygen) {
      this.oxygen = new _OxygenCore.default();
      await this.oxygen.init(this.config, this.capabilities);
    }
  }
  async disposeOxygenCore(status = null) {
    if (this.oxygen) {
      try {
        await this.oxygen.dispose(status);
      } catch (e) {
        console.error('Failed to dispose Oxygen modules:', e);
      }
      this.oxygen = null;
    }
  }
  registerCompilers() {
    this.cucumberOpts.compiler.forEach(compiler => {
      if (compiler instanceof Array) {
        let parts = compiler[0].split(':');
        require(parts[1])(compiler[1]);
      } else {
        let parts = compiler.split(':');
        require(parts[1]);
      }
    });
  }
  requiredFiles() {
    return this.cucumberOpts.require.reduce((files, requiredFile) => {
      const absolutePath = _util.default.resolvePath(requiredFile, this.cwd);
      if ((0, _isGlob.default)(absolutePath)) {
        return files.concat(_glob.default.sync(absolutePath));
      } else {
        return files.concat([absolutePath]);
      }
    }, []);
  }
  loadRequireFiles() {
    td.replace('cucumber', Cucumber);
    this.requiredFiles().forEach(codePath => {
      delete require.cache[require.resolve(codePath)];
      require(codePath);
    });
  }
  resolveSpecFiles(specs) {
    if (!Array.isArray(specs)) {
      return [];
    }
    return specs.reduce((files, specFile) => {
      const absolutePath = _util.default.resolvePath(specFile, this.cwd);
      if ((0, _isGlob.default)(absolutePath)) {
        return files.concat(_glob.default.sync(absolutePath));
      } else {
        return files.concat(absolutePath);
      }
    }, []);
  }
  beforeCommandHandler() {
    if (this.config.beforeCommand && typeof this.config.beforeCommand === 'function') {
      this.config.beforeCommand();
    }
  }
  afterCommandHandler() {
    if (this.config.afterCommand && typeof this.config.afterCommand === 'function') {
      this.config.afterCommand();
    }
  }
  wrapSteps() {
    const wrapStep = this.wrapStep.bind(this);
    const config = this.config;
    const id = this.id;
    Cucumber.setDefinitionFunctionWrapper((fn, options = {}) => {
      if (fn.name.startsWith('wdioHook')) {
        return fn;
      }
      const isStep = !fn.name.startsWith('userHook');
      const retryTest = isStep && isFinite(options.retry) ? parseInt(options.retry, 10) : 0;
      return wrapStep(fn, retryTest, isStep, config, id);
    });
  }
  wrapStep(code, retryTest = 0, isStep, config, id) {
    const executeFn = (0, _utils.isFunctionAsync)(code) || !_utils.hasWdioSyncSupport ? _sync.executeAsync : _sync.executeSync;
    const wrapWithHooks = this.wrapWithHooks.bind(this);
    return function (...args) {
      return executeFn.call(this, wrapWithHooks(code), retryTest, args);
    };
  }
  wrapWithHooks(code) {
    const userFn = async function (...args) {
      let result;
      let error;
      try {
        result = await (0, _utils.runFnInFiberContext)(code.bind(this, ...args))();
      } catch (err) {
        error = err;
      }
      if (error) {
        throw error;
      }
      return result;
    };
    return userFn;
  }
  hookInCucumberEvents(eventBroadcaster) {
    this.cucumberEventListener = new _CucumberEventListener.default(eventBroadcaster);
  }
  onBeforeFeature() {}
  onAfterFeature() {}
  onTestEnd() {}
}
exports.default = CucumberWorker;
