// oxygen-cli/build/runners/cucumber/CucumberReporter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _suiteResult = _interopRequireDefault(require("../../model/suite-result"));
var _caseResult = _interopRequireDefault(require("../../model/case-result"));
var _stepResult = _interopRequireDefault(require("../../model/step-result"));
var _status = _interopRequireDefault(require("../../model/status"));
var _util = _interopRequireDefault(require("../../lib/util"));
var _helper = _interopRequireDefault(require("../../errors/helper"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class CucumberReporter {
  constructor(runnerId, options, cucumberEventListener, oxygenEventListener, reportDispatcher, testHooks) {
    this.runnerId = runnerId;
    this.suites = {};
    this.currentStep = null;
    this.currentCase = null;
    this.cucumberEventListener = cucumberEventListener;
    this.oxygenEventListener = oxygenEventListener;
    this.reportDispatcher = reportDispatcher;
    this.options = options;
    this.onBeforeFeature = this.onBeforeFeature.bind(this);
    this.onAfterFeature = this.onAfterFeature.bind(this);
    this.onBeforeScenario = this.onBeforeScenario.bind(this);
    this.onAfterScenario = this.onAfterScenario.bind(this);
    this.onBeforeStep = this.onBeforeStep.bind(this);
    this.onAfterStep = this.onAfterStep.bind(this);
    this.onTestEnd = this.onTestEnd.bind(this);
    this.onBeforeOxygenCommand = this.onBeforeOxygenCommand.bind(this);
    this.onAfterOxygenCommand = this.onAfterOxygenCommand.bind(this);
    this.onOxygenLog = this.onOxygenLog.bind(this);
    this.testHooks = testHooks;
    this.hookEvents();
  }
  hookEvents() {
    this.cucumberEventListener.on('feature:before', this.onBeforeFeature);
    this.cucumberEventListener.on('feature:after', this.onAfterFeature);
    this.cucumberEventListener.on('scenario:before', this.onBeforeScenario);
    this.cucumberEventListener.on('scenario:after', this.onAfterScenario);
    this.cucumberEventListener.on('step:before', this.onBeforeStep);
    this.cucumberEventListener.on('step:after', this.onAfterStep);
    this.cucumberEventListener.on('test:end', this.onTestEnd);
    this.oxygenEventListener.on('command:before', this.onBeforeOxygenCommand);
    this.oxygenEventListener.on('command:after', this.onAfterOxygenCommand);
    this.oxygenEventListener.on('log', this.onOxygenLog);
  }
  onBeforeOxygenCommand(e) {}
  onAfterOxygenCommand(e) {
    if (this.currentStep && e.result) {
      if (!this.currentStep.steps) {
        this.currentStep.steps = [];
      }
      this.currentStep.steps.push(e.result);
    }
  }
  onOxygenLog(log) {
    if (this.currentCase) {
      if (!this.currentCase.logs) {
        this.currentCase.logs = [];
      }
      this.currentCase.logs.push({
        time: log.time,
        level: log.level.toUpperCase(),
        msg: log.message,
        src: log.src
      });
    }
    if (this.reportDispatcher.onLogEntry && typeof this.reportDispatcher.onLogEntry === 'function') {
      this.reportDispatcher.onLogEntry(log.time, log.level, log.message, log.src);
    }
  }
  onBeforeFeature(uri, feature) {
    const location = `${uri}:${feature.location.line}`;
    const suiteResult = this.suites[location] = new _suiteResult.default();
    suiteResult.name = feature.name;
    suiteResult.tags = this.simplifyCucumberTags(feature.tags);
    suiteResult.location = location;
    suiteResult.startTime = _util.default.getTimeStamp();
    if (typeof this.testHooks.beforeSuite === 'function') {
      this.testHooks.beforeSuite(this.runnerId, suiteResult);
    }
    if (this.reportDispatcher.onSuiteStart && typeof this.reportDispatcher.onSuiteStart === 'function') {
      this.reportDispatcher.onSuiteStart(this.runnerId, uri, suiteResult);
    }
  }
  onAfterFeature(uri, feature) {
    const location = `${uri}:${feature.location.line}`;
    const suiteResult = this.suites[location];
    if (!suiteResult) {
      return;
    }
    suiteResult.endTime = _util.default.getTimeStamp();
    suiteResult.duration = suiteResult.endTime - suiteResult.startTime;
    suiteResult.status = this.determineSuiteStatus(suiteResult);
    if (typeof this.testHooks.afterSuite === 'function') {
      this.testHooks.afterSuite(this.runnerId, suiteResult);
    }
    if (this.reportDispatcher.onSuiteEnd && typeof this.reportDispatcher.onSuiteEnd === 'function') {
      this.reportDispatcher.onSuiteEnd(this.runnerId, uri, suiteResult);
    }
  }
  async onBeforeScenario(uri, feature, scenario) {
    const suiteId = `${uri}:${feature.location.line}`;
    const caseLocation = scenario.locations.length > 0 ? scenario.locations[0] : {
      line: 1
    };
    const caseId = `${uri}:${caseLocation.line}`;
    const suiteResult = this.suites[suiteId];
    const cases = suiteResult.cases;
    const caseResult = this.currentCase = new _caseResult.default();
    caseResult.name = scenario.name;
    caseResult.tags = this.simplifyCucumberTags(scenario.tags);
    caseResult.location = caseId;
    caseResult.startTime = _util.default.getTimeStamp();
    cases.push(caseResult);
    if (typeof this.oxygenEventListener.onBeforeCase === 'function') {
      await this.oxygenEventListener.onBeforeCase();
    }
    if (typeof this.testHooks.beforeCase === 'function') {
      this.testHooks.beforeCase(this.runnerId, scenario);
    }
    if (this.reportDispatcher.onCaseStart && typeof this.reportDispatcher.onCaseStart === 'function') {
      this.reportDispatcher.onCaseStart(this.runnerId, uri, caseId, scenario);
    }
  }
  async onAfterScenario(uri, feature, scenario, sourceLocation) {
    const suiteId = `${uri}:${feature.location.line}`;
    const caseId = `${uri}:${sourceLocation.line}`;
    const suiteResult = this.suites[suiteId];
    const cases = suiteResult.cases;
    const caseResult = cases.find(x => x.location === caseId);
    if (!caseResult) {
      return;
    }
    caseResult.endTime = _util.default.getTimeStamp();
    caseResult.duration = caseResult.endTime - caseResult.startTime;
    caseResult.status = this.determineCaseStatus(caseResult);
    caseResult.logs = this.oxygenEventListener.resultStore && this.oxygenEventListener.resultStore.logs ? this.oxygenEventListener.resultStore.logs : [];
    caseResult.har = this.oxygenEventListener.resultStore && this.oxygenEventListener.resultStore.har ? this.oxygenEventListener.resultStore.har : null;
    if (typeof this.oxygenEventListener.onAfterCase === 'function') {
      await this.oxygenEventListener.onAfterCase();
    }
    if (typeof this.testHooks.afterCase === 'function') {
      this.testHooks.afterCase(this.runnerId, caseResult);
    }
    if (this.reportDispatcher.onCaseEnd && typeof this.reportDispatcher.onCaseEnd === 'function') {
      this.reportDispatcher.onCaseEnd(this.runnerId, uri, caseId, caseResult);
    }
  }
  onBeforeStep(uri, feature, scenario, step, sourceLocation) {
    const suiteId = `${uri}:${feature.location.line}`;
    const caseId = `${uri}:${sourceLocation.line}`;
    const suiteResult = this.suites[suiteId];
    const cases = suiteResult.cases;
    const caseResult = cases.find(x => x.location === caseId);
    const stepResult = this.currentStep = new _stepResult.default();
    caseResult.steps.push(stepResult);
    stepResult.name = step.text;
    stepResult.startTime = _util.default.getTimeStamp();
    stepResult.location = `${uri}:${step.location.line}`;
    if (typeof this.testHooks.beforeStep === 'function') {
      this.testHooks.beforeStep(this.runnerId, step);
    }
    if (this.reportDispatcher.onStepStart && typeof this.reportDispatcher.onStepStart === 'function') {
      this.reportDispatcher.onStepStart(this.runnerId, stepResult);
    }
  }
  onAfterStep(uri, feature, scenario, step, result, sourceLocation) {
    const suiteId = `${uri}:${feature.location.line}`;
    const caseId = `${uri}:${sourceLocation.line}`;
    const suiteResult = this.suites[suiteId];
    const cases = suiteResult.cases;
    const caseResult = cases.find(x => x.location === caseId);
    const stepLocation = `${uri}:${step.location.line}`;
    const stepResult = caseResult.steps.find(x => x.location === stepLocation);
    stepResult.endTime = _util.default.getTimeStamp();
    stepResult.duration = stepResult.endTime - stepResult.startTime;
    stepResult.status = result.status;
    if (result.exception) {
      stepResult.failure = _helper.default.getFailureFromError(result.exception);
      if (this.currentStep && this.currentStep.location) {
        stepResult.failure.cucumberLocation = this.currentStep.location;
      }
    }
    if (typeof this.testHooks.afterStep === 'function') {
      this.testHooks.afterStep(this.runnerId, stepResult, result.exception);
    }
    if (this.reportDispatcher.onStepEnd && typeof this.reportDispatcher.onStepEnd === 'function') {
      this.reportDispatcher.onStepEnd(this.runnerId, stepResult);
    }
  }
  onTestEnd(result) {}
  determineCaseStatus(caseResult) {
    const hasFailedStep = caseResult.steps.some(x => x.status === _status.default.FAILED);
    if (hasFailedStep) {
      return _status.default.FAILED;
    }
    return _status.default.PASSED;
  }
  determineSuiteStatus(suiteResult) {
    const hasFailedCase = suiteResult.cases.some(x => x.status === _status.default.FAILED);
    if (hasFailedCase) {
      return _status.default.FAILED;
    }
    return _status.default.PASSED;
  }
  simplifyCucumberTags(tags) {
    if (!Array.isArray(tags) || tags.length == 0) {
      return tags;
    }
    const simpleTags = [];
    for (let tag of tags) {
      simpleTags.push(tag.name);
    }
    return simpleTags;
  }
}
exports.default = CucumberReporter;
