// oxygen-cli/build/runners/cucumber/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _glob = _interopRequireDefault(require("glob"));
var _isGlob = _interopRequireDefault(require("is-glob"));
var _path = _interopRequireDefault(require("path"));
var _events = require("events");
var _util = _interopRequireDefault(require("../../lib/util"));
var _WorkerProcess = _interopRequireDefault(require("../WorkerProcess"));
var _logger = _interopRequireDefault(require("../../lib/logger"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const log = (0, _logger.default)('CucumberRunner');
const DEFAULT_ISSUER = 'user';
const DEFAULT_TIMEOUT = 30000;
const DEFAULT_OPTS = {
  backtrace: false,
  compiler: [],
  failAmbiguousDefinitions: false,
  failFast: false,
  ignoreUndefinedDefinitions: false,
  name: [],
  profile: [],
  require: [],
  order: 'defined',
  snippetSyntax: undefined,
  snippets: true,
  source: true,
  strict: false,
  tagExpression: '',
  tagsInTitle: false,
  timeout: DEFAULT_TIMEOUT
};
class CucumberRunner extends _events.EventEmitter {
  constructor(config) {
    super();
    this._debugMode = config && config.debugPortIde ? true : false;
    this._debugPort = null;
    this._npmGRootExecution = true;
    if (config && typeof config.npmGRootExecution !== 'undefined') {
      this._npmGRootExecution = config.npmGRootExecution;
    }
    this.isInitialized = false;
    this.id = _util.default.generateUniqueId();
    const workerPath = _path.default.join(__dirname, 'worker.js');
    this.worker = new _WorkerProcess.default(this.id, workerPath, this._debugMode, this._debugPort, 'Cucumber', this._npmGRootExecution);
    this.handleWorkerEvents();
  }
  async init(config, caps, reporter) {
    this.config = config;
    this.cwd = this.config.cwd || process.cwd();
    this.specs = this.resolveSpecFiles(config.specs || []);
    this.capabilities = caps;
    this.reporter = reporter;
    this.isInitialized = true;
    this.cucumberOpts = Object.assign(DEFAULT_OPTS, config.cucumberOpts);
    this.worker.start();
    await this.worker.init(this.id, config, caps);
  }
  async dispose(status = null) {
    await this.worker.dispose(status);
    this.isInitialized = false;
  }
  async run() {
    try {
      this.reporter.onRunnerStart(this.id, this.config, this.capabilities);
      await this.worker.run({
        context: {
          caps: this.capabilities
        },
        poFile: this.config.po || null
      });
      this.reporter.onRunnerEnd(this.id, null);
    } catch (e) {
      console.log('Fatal error in Cucumber runner:', e);
      this.reporter.onRunnerEnd(this.id, null, e);
    }
    return await this.reporter.waitForResult(this.id);
  }
  async kill(status = null) {
    this._testKilled = true;
  }
  resolveSpecFiles(specs) {
    if (!Array.isArray(specs)) {
      return [];
    }
    return specs.reduce((files, specFile) => {
      const absolutePath = _util.default.resolvePath(specFile, this.cwd);
      if ((0, _isGlob.default)(absolutePath)) {
        return files.concat(_glob.default.sync(absolutePath));
      } else {
        return files.concat([absolutePath]);
      }
    }, []);
  }
  handleWorkerEvents() {
    if (!this.worker) {
      return;
    }
    this.worker.on('message', async msg => {
      if (msg.event && msg.event === 'log') {
        if (msg.level === 'DEBUG') {
          log.debug(msg.msg);
        } else if (msg.level === 'INFO') {
          log.info(msg.msg);
        } else if (msg.level === 'ERROR') {
          log.error(msg.msg, msg.err);
        } else if (msg.level === 'WARN') {
          log.warn(msg.msg);
        }
        this.emit('log', msg.time, msg.level, msg.msg, msg.src || DEFAULT_ISSUER);
        this.reporter && this.reporter.onLogEntry(msg.time, msg.level, msg.msg, msg.src || DEFAULT_ISSUER);
      }
    });
    this.worker.on('reporter', this.callReporter.bind(this));
  }
  callReporter({
    method,
    args
  }) {
    if (this.reporter && this.reporter[method] && typeof this.reporter[method] === 'function') {
      this.reporter[method].apply(this.reporter, args);
    }
  }
}
exports.default = CucumberRunner;
