// oxygen-cli/build/ox_services/service-devtools/submodule-network.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _OxygenSubModule = require("../../core/OxygenSubModule");
var _chromeHar = require("chrome-har");
var _helper = _interopRequireDefault(require("../../errors/helper"));
var _OxygenError = _interopRequireDefault(require("../../errors/OxygenError"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const observe = ['Page.domContentEventFired', 'Page.frameAttached', 'Page.loadEventFired', 'Page.frameRequestedNavigation', 'Page.frameStartedLoading', 'Page.navigatedWithinDocument', 'Network.dataReceived', 'Network.loadingFailed', 'Network.loadingFinished', 'Network.requestServedFromCache', 'Network.requestWillBeSent', 'Network.responseReceived', 'Network.requestWillBeSentExtraInfo', 'Network.resourceChangedPriority', 'Network.responseReceivedExtraInfo'];
class NetworkSubModule extends _OxygenSubModule.OxygenSubModule {
  constructor(name, parent) {
    super(name, parent);
    this._devTools = null;
    this._collectData = false;
    this._events = [];
  }
  init(devTools) {
    if (!devTools || !this._parent || !this._parent.getDriver || typeof this._parent.getDriver !== 'function' || !this._parent.getDriver()) {
      super.init();
      return false;
    }
    this._devTools = devTools;
    this._driver = this._parent.getDriver();
    observe.forEach(method => {
      this._driver.on(method, params => {
        if (this._collectData) {
          this._events.push({
            method,
            params
          });
        }
      });
    });
    super.init();
    return true;
  }
  async dispose() {
    try {
      if (this._driver) {
        observe.forEach(async method => {
          await this._driver.removeListener(method, () => {});
        });
        await this._driver.emit('Network.disable');
        await this._driver.emit('Network.close');
      }
    } catch (e) {}
    this._devTools = null;
    this._driver = null;
    this._events = [];
    super.dispose();
  }
  start() {
    this._collectData = true;
    this._events = [];
  }
  stop() {
    this._collectData = false;
  }
  async getRequests() {
    return await this._getEntries();
  }
  async _getEntries() {
    try {
      const har = (0, _chromeHar.harFromMessages)(this._events);
      if (har && har.log && har.log.entries && har.log.entries.map) {
        har.log.entries.reverse();
        return await Promise.all(har.log.entries.map(async item => {
          let responseBody = {};
          try {
            responseBody = await this._driver.cdp('Network', 'getResponseBody', {
              requestId: item._requestId
            });
          } catch (e) {}
          if (responseBody && responseBody.body) {
            try {
              responseBody.body = JSON.parse(responseBody.body);
            } catch (e) {}
          }
          let requestPostData = {};
          try {
            requestPostData = await this._driver.cdp('Network', 'getRequestPostData', {
              requestId: item._requestId
            });
          } catch (e) {}
          return {
            ...item,
            url: item.request.url,
            requestId: item._requestId,
            ...item.response,
            responseBody,
            requestPostData
          };
        }));
      } else {
        return [];
      }
    } catch (e) {
      return [];
    }
  }
  async waitForUrl(pattern, timeout = 60 * 1000) {
    if (!this._driver || !this._isInitialized || !this._parent) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`network` or `web` module is not initialized.');
    }
    if (!this._collectData) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`network.start()` must be executed prior to using `network` commands.');
    }
    this._parent.helpers.assertArgument(pattern, 'pattern');
    this._parent.helpers.assertArgumentTimeout(timeout, 'timeout');
    const start = Date.now();
    while (Date.now() - start < timeout) {
      const networkRequests = await this._getEntries();
      for (let req of networkRequests) {
        if (pattern.constructor.name === 'RegExp' && pattern.test(req.url) || pattern === req.url) {
          return req;
        }
      }
      await this._driver.pause(500);
    }
    throw new _OxygenError.default(_helper.default.errorCode.TIMEOUT, `No request matching the URL "${pattern}" was found.`);
  }
  async waitForNotUrl(pattern, timeout = 60 * 1000) {
    if (!this._driver || !this._isInitialized || !this._parent) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`network` or `web` module is not initialized.');
    }
    if (!this._collectData) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`network.start()` must be executed prior to using `network` commands.');
    }
    this._parent.helpers.assertArgument(pattern, 'pattern');
    this._parent.helpers.assertArgumentTimeout(timeout, 'timeout');
    const start = Date.now();
    while (Date.now() - start < timeout) {
      const networkRequests = await this._getEntries();
      for (let req of networkRequests) {
        if (pattern.constructor.name === 'RegExp' && pattern.test(req.url) || pattern === req.url) {
          throw new _OxygenError.default(_helper.default.errorCode.TIMEOUT, `A request matching the URL "${pattern}" was found.`);
        }
      }
      await this._driver.pause(500);
    }
  }
  async waitFor(matcher, timeout = 60 * 1000) {
    if (!this._driver || !this._isInitialized || !this._parent) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`network` module is not initialized.');
    }
    if (!this._collectData) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`web.network.start()` must be executed prior to using `network` commands.');
    }
    this._parent.helpers.assertArgument(matcher, 'matcher');
    this._parent.helpers.assertArgumentTimeout(timeout, 'timeout');
    const start = Date.now();
    while (Date.now() - start < timeout) {
      const networkRequests = await this._getEntries();
      for (var req of networkRequests) {
        if (matcher(req)) {
          return req;
        }
      }
      await this._driver.pause(500);
    }
    throw new _OxygenError.default(_helper.default.errorCode.TIMEOUT, 'No request found using the provided matcher.');
  }
  async assertUrl(url, timeout = 60 * 1000) {
    if (!this._driver || !this._isInitialized || !this._parent) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`network` or `web` module is not initialized.');
    }
    if (!this._collectData) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`web.network.start()` must be executed prior to using `network` commands.');
    }
    this._parent.helpers.assertArgument(url, 'url');
    this._parent.helpers.assertArgumentTimeout(timeout, 'timeout');
    const start = Date.now();
    while (Date.now() - start < timeout) {
      const networkRequests = await this._getEntries();
      for (let req of networkRequests) {
        if (url.constructor.name === 'RegExp' && url.test(req.url) || url === req.url) {
          return req;
        }
      }
      await this._driver.pause(500);
    }
    throw new _OxygenError.default(_helper.default.errorCode.ASSERT_ERROR, `No request matching the URL "${url}" was found.`);
  }
  async assertStatusCode(url, statusCode, failureMessage = null, timeout = 60 * 1000) {
    if (!this._driver || !this._isInitialized || !this._parent) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`network` or `web` module is not initialized.');
    }
    if (!this._collectData) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`web.network.start()` must be executed prior to using `network` commands.');
    }
    this._parent.helpers.assertArgument(url, 'url');
    this._parent.helpers.assertArgument(statusCode, 'statusCode');
    this._parent.helpers.assertArgumentTimeout(timeout, 'timeout');
    const start = Date.now();
    let matchedReq = null;
    while (Date.now() - start < timeout && !matchedReq) {
      const networkRequests = await this._getEntries();
      for (let req of networkRequests) {
        if (url.constructor.name === 'RegExp' && url.test(req.url) || url === req.url) {
          matchedReq = req;
          break;
        }
      }
      !matchedReq && (await this._driver.pause(500));
    }
    if (!matchedReq) {
      throw new _OxygenError.default(_helper.default.errorCode.ASSERT_ERROR, `No request matching the URL "${url}" was found.`);
    }
    if (statusCode !== matchedReq.status) {
      const message = failureMessage || `The expected status code "${statusCode}" does not match "${matchedReq.status}".`;
      throw new _OxygenError.default(_helper.default.errorCode.ASSERT_ERROR, message);
    }
  }
  async assertResponseContent(url, content, failureMessage = null, timeout = 60 * 1000) {
    if (!this._driver || !this._isInitialized || !this._parent) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`network` or `web` module is not initialized.');
    }
    if (!this._collectData) {
      throw new _OxygenError.default(_helper.default.errorCode.MODULE_NOT_INITIALIZED_ERROR, '`web.network.start()` must be executed prior to using `network` commands.');
    }
    this._parent.helpers.assertArgument(url, 'url');
    this._parent.helpers.assertArgument(content, 'content');
    this._parent.helpers.assertArgumentTimeout(timeout, 'timeout');
    const start = Date.now();
    let matchedReq = null;
    while (Date.now() - start < timeout && !matchedReq) {
      const networkRequests = await this._getEntries();
      for (let req of networkRequests) {
        if (url.constructor.name === 'RegExp' && url.test(req.url) || url === req.url) {
          matchedReq = req;
          break;
        }
      }
      !matchedReq && (await this._driver.pause(500));
    }
    if (!matchedReq) {
      throw new _OxygenError.default(_helper.default.errorCode.ASSERT_ERROR, `No request matching the URL "${url}" was found.`);
    }
    const resp = await this._driver.cdp('Network', 'getResponseBody', {
      requestId: matchedReq.requestId
    });
    if (resp && resp.body && !resp.base64Encoded) {
      if (content.constructor.name === 'RegExp' && (content.test(resp.body) || resp.body.indexOf(content) > -1)) {
        return true;
      } else {
        const message = failureMessage || `Response "${url}" body content does not match "${content}.`;
        throw new _OxygenError.default(_helper.default.errorCode.ASSERT_ERROR, message);
      }
    }
  }
  async blockUrls(urls) {
    await this._driver.cdp('Network', 'setBlockedURLs', {
      urls
    });
  }
}
exports.default = NetworkSubModule;
