// oxygen-cli/build/runners/WorkerProcess.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _child_process = require("child_process");
var _events = require("events");
var _when = require("when");
var _logger = _interopRequireDefault(require("../lib/logger"));
var _debugger = _interopRequireDefault(require("../lib/debugger"));
var _util = _interopRequireDefault(require("../lib/util"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const log = (0, _logger.default)('WorkerProcess');
const snooze = ms => new Promise(resolve => setTimeout(resolve, ms));
class WorkerProcess extends _events.EventEmitter {
  constructor(pid, workerPath, debugMode, debugPort, name = null, npmGRootExecution = true) {
    super();
    this._name = name;
    this._isRunning = false;
    this._isInitialized = false;
    this._stoppedByUser = true;
    this._pid = pid;
    this._childProc = null;
    this._debugMode = debugMode;
    this._debugPort = debugPort;
    this._debugger = null;
    this._workerPath = workerPath;
    this._npmGRootExecution = npmGRootExecution;
    this._breakpointErrors = [];
    this._calls = {};
    this._eventHandlers = {};
  }
  async start() {
    const env = Object.assign(process.env, {
      OX_WORKER: true,
      DEBUG: !isNaN(this._debugPort),
      NO_UPDATE_NOTIFIER: true
    });
    log.info(`Starting worker process ${this._pid}.`);
    let forkOpts = {
      cwd: process.cwd(),
      env,
      execArgv: [],
      detached: false
    };
    if (this._debugPort) {
      forkOpts.execArgv = Object.assign(forkOpts.execArgv, ['--inspect-brk=' + this._debugPort]);
    }
    forkOpts.execArgv.push('--trace-warnings');
    if (this._npmGRootExecution) {
      try {
        let globalNpmModulesPath;
        if (process.platform === 'darwin') {
          await (() => {
            return new Promise((resolve, reject) => {
              try {
                (0, _child_process.exec)('npm root -g', {
                  cwd: process.cwd()
                }, function (err, stdout) {
                  if (err) {
                    console.log('darwin npm root error:', err);
                    resolve();
                  }
                  if (stdout && stdout.toString) {
                    globalNpmModulesPath = stdout.toString().trim();
                    resolve();
                  }
                });
              } catch (e) {
                reject(e);
              }
            });
          })();
        } else {
          const execResult = (0, _child_process.execSync)('npm root -g');
          if (execResult && execResult.toString) {
            globalNpmModulesPath = execResult.toString().trim();
          }
        }
        if (globalNpmModulesPath && forkOpts && forkOpts.env) {
          forkOpts.env.NODE_PATH = globalNpmModulesPath;
        }
      } catch (e) {
        console.log('npm root error:', e);
      }
    }
    this._childProc = (0, _child_process.fork)(this._workerPath, forkOpts);
    this._hookChildProcEvents();
    this._isRunning = true;
    return this._childProc;
  }
  async startDebugger() {
    if (this._debugMode) {
      await this._initializeDebugger();
    }
  }
  async stopDebugger() {
    if (this._debugger) {
      await this._debugger.close();
      this._debugger.removeAllListeners();
      this._debugger = null;
    }
  }
  async stop(status = null) {
    if (this._isInitialized && this._childProc) {
      if (this._debugger && this._debugger._paused) {
        if ('CANCELED' === status) {
          await this._debugger.disable();
        } else {
          await this._debugger.resumeTerminate();
        }
      }
      await this.invoke('dispose', status);
      await snooze(100);
    } else if (this._childProc) {
      await snooze(100);
    }
    if (this._debugger) {
      await this._debugger.close();
    }
    this._send({
      type: 'exit',
      status: status
    });
    this.kill();
    this._reset();
  }
  kill() {
    if (this._childProc) {
      this._childProc.kill('SIGINT');
    }
  }
  async initOxygen(rid, options, caps) {
    if (!this._isInitialized) {
      const beforeTime = new Date().getTime();
      await this.invoke('init', rid, options, caps);
      const afterTime = new Date().getTime();
      const duration = afterTime - beforeTime;
      this._isInitialized = true;
      let start = '';
      if (this._name) {
        start = this._name + ' ';
      }
      log.info(start + 'Worker initialized in ' + duration + ' ms');
    }
  }
  async startSession() {
    let start = '';
    if (this._name) {
      start = this._name + ' ';
    }
    log.info(start + 'Session started');
  }
  async endSession(status = null, disposeModules = true) {
    if (this._childProc) {
      const beforeTime = new Date().getTime();
      await this.invoke('endSession', status, disposeModules);
      const afterTime = new Date().getTime();
      const duration = afterTime - beforeTime;
      let start = '';
      if (this._name) {
        start = this._name + ' ';
      }
      log.info(start + 'Session ended in ' + duration + ' ms');
    }
  }
  async dispose(status = null) {
    if (this._debugger) {
      await this._debugger.close();
    }
    if (this._isInitialized && this._childProc) {
      this._isInitialized = false;
      await this.invoke('dispose', status);
      this._send({
        type: 'exit',
        status: status
      });
      this.kill();
    } else if (this._childProc) {
      this._isInitialized = false;
      this._send({
        type: 'exit',
        status: status
      });
      this.kill();
    }
  }
  get debugger() {
    return this._debugger;
  }
  get isRunning() {
    return this._isRunning;
  }
  get isInitialized() {
    return this._isInitialized;
  }
  async run(runOpts) {
    return await this.invoke('run', runOpts);
  }
  subscribe(eventName, handler) {
    if (!this._eventHandlers[eventName] && typeof handler === 'function') {
      this._eventHandlers[eventName] = handler;
      this._send({
        type: 'subscribe',
        event: eventName
      });
    }
  }
  async invoke(method) {
    let args = Array.prototype.slice.call(arguments);
    if (args.length > 0) {
      args.shift();
    }
    if (this._childProc) {
      const callId = _util.default.generateUniqueId();
      this._send({
        type: 'invoke',
        method: method,
        callId: callId,
        args: args
      });
      this._calls[callId] = (0, _when.defer)();
      return this._calls[callId].promise;
    }
  }
  async invokeTestHook(hookName, hookArgs) {
    const callId = _util.default.generateUniqueId();
    this._send({
      type: 'invoke:hook',
      method: hookName,
      callId: callId,
      args: hookArgs
    });
    this._calls[callId] = (0, _when.defer)();
    return this._calls[callId].promise;
  }
  _send(message) {
    if (!this._childProc) {
      return false;
    }
    this._childProc.send(message);
    return true;
  }
  _reset() {
    this._childProc = null;
    this._debugger = null;
    this._isRunning = false;
    this._debugMode = null;
    this._debugPort = null;
    this._whenOxygenDisposed = null;
    this._whenOxygenInitialized = null;
    this._whenModulesDisposed = null;
    this._isInitialized = false;
  }
  _hookChildProcEvents() {
    if (!this._childProc) {
      return;
    }
    this._childProc.on('message', this._handleChildMessage.bind(this));
    this._childProc.on('error', this._handleChildError.bind(this));
    this._childProc.on('exit', this._handleChildExit.bind(this));
    this._childProc.on('disconnect', this._handleChildDisconnect.bind(this));
    this._childProc.on('uncaughtException', this._handleChildUncaughtException.bind(this));
    this._childProc.on('SIGINT', this._handleChildSigInt.bind(this));
  }
  _handleChildMessage(msg) {
    if (msg.event) {
      switch (msg.event) {
        case 'invoke:result':
          if (msg.callId && Object.prototype.hasOwnProperty.call(this._calls, msg.callId)) {
            const promise = this._calls[msg.callId];
            if (msg.error) {
              promise.reject(msg.error);
            } else {
              promise.resolve(msg.retval);
            }
            delete this._calls[msg.callId];
          }
          break;
        case 'reporter':
          if (msg.method) {
            this.emit('reporter', {
              pid: this._pid,
              method: msg.method,
              args: msg.args
            });
          }
          break;
        case 'event':
          this._handleWorkerEvent(msg.name, msg.args);
          break;
      }
    }
    this.emit('message', Object.assign(msg, {
      pid: this._pid
    }));
  }
  async _handleWorkerEvent(name, args) {
    if (name && this._eventHandlers && this._eventHandlers[name]) {
      try {
        await this._eventHandlers[name].apply(undefined, args);
      } catch (e) {
        log.error('Error thrown by worker event handler "${name}:', e);
      }
    }
  }
  _handleChildError(error) {
    log.debug(`Worker ${this._pid} thrown an error: ${error}.`);
    this.emit('error', Object.assign({
      error: error
    }, {
      pid: this._pid
    }));
  }
  _handleChildExit(exitCode, signal) {
    log.debug(`Worker ${this._pid} finished with exit code ${exitCode} and signal ${signal}.`);
    this._reset();
    this.emit('exit', {
      pid: this._pid,
      exitCode,
      signal
    });
  }
  _handleChildDisconnect() {
    log.debug(`Worker ${this._pid} disconnected.`);
  }
  _handleChildUncaughtException(error) {
    log.debug(`Worker ${this._pid} thrown an uncaught error: ${error}.`);
    this._reset();
    this.emit('error', Object.assign({
      error: error
    }, {
      pid: this._pid
    }));
    this.emit('exit', {
      pid: this._pid,
      exitCode: 1
    });
  }
  _handleChildSigInt() {
    log.debug(`Worker ${this._pid} received SIGINT signal.`);
  }
  async _initializeDebugger() {
    this._debugger = new _debugger.default(this._pid);
    let whenDebuggerReady = (0, _when.defer)();
    const _this = this;
    this._debugger.on('ready', function (err) {
      if (_this._debugger && _this._debugger.continue) {
        _this._debugger.continue();
      }
      whenDebuggerReady.resolve();
    });
    this._debugger.on('error', function (err) {
      log.error('Debugger error :', err);
      this.emit('debugger:error', Object.assign(err, {
        pid: this._pid
      }));
      if (!whenDebuggerReady.isResolved) {
        whenDebuggerReady.reject(err.message);
      }
    });
    this._debugger.on('break', function (breakpointData) {
      this.emit('debugger:break', breakpointData);
    });
    this._debugger.on('breakError', async function (breakError) {
      if (breakError && breakError.message) {
        if (_this._breakpointErrors.includes(breakError.message)) {} else {
          _this._breakpointErrors.push(breakError.message);
          this.emit('debugger:breakError', breakError);
        }
      }
    });
    try {
      log.info(`Connecting to debugger on port ${this._debugPort}...`);
      await this._debugger.connect(this._debugPort, '127.0.0.1');
    } catch (e) {
      log.error('Cannot connect to the debugger: ', e);
      const message = 'Cannot connect to the debugger: ' + e.message;
      whenDebuggerReady.reject(message);
    }
    return whenDebuggerReady.promise;
  }
}
exports.default = WorkerProcess;
