// oxygen-cli/build/lib/cli-util.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OXYGEN_PAGE_OBJECT_FILE_NAME = exports.OXYGEN_ENV_FILE_NAME = exports.OXYGEN_CONFIG_FILE_NAME = exports.DEFAULT_REPORTER = void 0;
exports.generateTestOptions = generateTestOptions;
exports.getCommandLineOptions = getCommandLineOptions;
exports.getConfigurations = getConfigurations;
exports.getEnvironments = getEnvironments;
exports.getPageObjectFilePath = getPageObjectFilePath;
exports.loadEnvironmentVariables = loadEnvironmentVariables;
exports.loadSuiteDefinitionsFromFolder = loadSuiteDefinitionsFromFolder;
exports.loadSuites = loadSuites;
exports.processTargetPath = processTargetPath;
require("source-map-support/register");
var _path = _interopRequireDefault(require("path"));
var _fs = _interopRequireDefault(require("fs"));
var _util = _interopRequireDefault(require("./util"));
var _async = require("async");
var _moduleRequire = _interopRequireDefault(require("./moduleRequire"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const OXYGEN_CONFIG_FILE_NAME = exports.OXYGEN_CONFIG_FILE_NAME = 'oxygen.conf';
const OXYGEN_ENV_FILE_NAME = exports.OXYGEN_ENV_FILE_NAME = 'oxygen.env';
const OXYGEN_PAGE_OBJECT_FILE_NAME = exports.OXYGEN_PAGE_OBJECT_FILE_NAME = 'oxygen.po';
const DEFAULT_REPORTER = exports.DEFAULT_REPORTER = 'html';
async function generateTestOptions(config, argv) {
  const options = {
    ...config
  };
  options.env = loadEnvironmentVariables(config, argv);
  options.po = getPageObjectFilePath(config, argv);
  options.suites = await loadSuites(config, argv);
  return options;
}
async function loadSuites(config, argv) {
  if (config.framework && typeof config.framework === 'string' && config.framework.toLowerCase() !== 'oxygen') {
    return;
  }
  const {
    target
  } = config;
  const isConfigFile = target && (target.name && target.name.indexOf(OXYGEN_CONFIG_FILE_NAME) === 0 || target.extension === '' && target.configPath && target.configPath.indexOf(OXYGEN_CONFIG_FILE_NAME) > -1);
  let suites = [];
  if (!isConfigFile) {
    if (target && target.extension === '.js') {
      suites.push(await _util.default.generateTestSuiteFromJSFile(target.path, config.parameters.file, config.parameters.mode, false, config.iterations || 1));
    } else if (target && target.extension === '.json') {
      suites.push(await _util.default.generateTestSuiteFromJsonFile(target.path, config.parameters.file, config.parameters.mode, config));
    }
  } else {
    let suiteDefs = [];
    if (config.suites && Array.isArray(config.suites)) {
      suiteDefs = config.suites;
    }
    const suitesFolder = _path.default.join(target.cwd, 'suites');
    if (_fs.default.existsSync(suitesFolder)) {
      suiteDefs = [...suiteDefs, ...loadSuiteDefinitionsFromFolder(suitesFolder)];
    }
    suites = await Promise.all(suiteDefs.map(async suiteDef => await _util.default.generateTestSuiteFromJson(suiteDef, config)));
  }
  if (argv.suites && typeof argv.suites === 'string') {
    const selectedSuiteNames = argv.suites.split(',');
    suites = suites.filter(x => selectedSuiteNames.includes(x.name));
  }
  return suites;
}
function loadSuiteDefinitionsFromFolder(folderPath) {
  const files = _fs.default.readdirSync(folderPath);
  let suiteDefs = [];
  (0, _async.forEach)(files, file => {
    if (_path.default.extname(file) === '.json') {
      const fullPath = _path.default.join(folderPath, file);
      suiteDefs.push((0, _moduleRequire.default)(fullPath));
    }
  });
  return suiteDefs;
}
function getPageObjectFilePath(config, argv = {}) {
  const target = config.target || {};
  const poFileName = argv.po || `${OXYGEN_PAGE_OBJECT_FILE_NAME}.js`;
  const cwd = target.cwd || process.cwd();
  let poFilePath = _path.default.resolve(cwd, poFileName);
  poFilePath = _fs.default.existsSync(poFilePath) ? poFilePath : null;
  if (poFilePath) {
    try {
      (0, _moduleRequire.default)(poFilePath);
      return poFilePath;
    } catch (e) {
      const err = new Error(`Unable to load file: ${poFilePath}. Reason: ${e.message} ${e.stack}`);
      throw err;
    }
  } else {
    return null;
  }
}
function loadEnvironmentVariables(config, argv) {
  const target = config.target || {};
  const envName = argv.env || config.environment || 'default';
  const cwd = target.cwd || process.cwd();
  const defaultEnvFile = _path.default.join(cwd, `${OXYGEN_ENV_FILE_NAME}.js`);
  if (_fs.default.existsSync(defaultEnvFile)) {
    const env = (0, _moduleRequire.default)(defaultEnvFile);
    if (env && typeof env === 'object' && Object.prototype.hasOwnProperty.call(env, envName)) {
      return env[envName];
    }
  }
  const dedicatedEnvFileJs = _path.default.join(cwd, 'env', `${envName}.js`);
  const dedicatedEnvFileJson = _path.default.join(cwd, 'env', `${envName}.json`);
  if (_fs.default.existsSync(dedicatedEnvFileJs)) {
    return (0, _moduleRequire.default)(dedicatedEnvFileJs);
  } else if (_fs.default.existsSync(dedicatedEnvFileJson)) {
    return (0, _moduleRequire.default)(dedicatedEnvFileJson);
  }
  return {};
}
function getEnvironments(target) {
  let targetCwd = null;
  if (target && target.cwd) {
    targetCwd = target.cwd;
  }
  const cwd = targetCwd || process.cwd();
  const defaultEnvFile = _path.default.join(cwd, `${OXYGEN_ENV_FILE_NAME}.js`);
  if (_fs.default.existsSync(defaultEnvFile)) {
    try {
      return (0, _moduleRequire.default)(defaultEnvFile);
    } catch (e) {
      const err = new Error(`Unable to load file: ${defaultEnvFile}. Reason: ${e.message} ${e.stack}`);
      throw err;
    }
  }
  return {};
}
function getConfigurations(target, argv) {
  let targetCwd = null;
  if (target && target.cwd) {
    targetCwd = target.cwd;
  }
  const DEFAULT_OPTS = {
    cwd: target ? targetCwd || process.cwd() : process.cwd(),
    target: target,
    browserName: 'chrome',
    seleniumUrl: 'http://localhost:4444/wd/hub',
    appiumUrl: 'http://localhost:4723/wd/hub',
    reopenSession: false,
    reRunOnFailed: false,
    iterations: 1,
    debugPort: null,
    delay: null,
    collectDeviceLogs: false,
    collectAppiumLogs: false,
    collectBrowserLogs: false,
    reporting: {
      reporters: [DEFAULT_REPORTER]
    },
    parameters: {
      file: null,
      mode: 'seq'
    }
  };
  const cmdOpts = getCommandLineOptions(argv);
  let projConfigOpts = {};
  if (target && target.name === OXYGEN_CONFIG_FILE_NAME && (target.extension === '.js' || target.extension === '.json')) {
    try {
      projConfigOpts = (0, _moduleRequire.default)(target.path);
    } catch (e) {
      const err = new Error(`Unable to load file: ${target.path}. Reason: ${e.message} ${e.stack}`);
      throw err;
    }
  } else if (target && target.configPath) {
    try {
      projConfigOpts = (0, _moduleRequire.default)(target.configPath);
    } catch (e) {
      const err = new Error(`Unable to load file: ${target.configPath}. Reason: ${e.message} ${e.stack}`);
      throw err;
    }
  }
  const envs = getEnvironments(target);
  if (projConfigOpts.envs) {
    projConfigOpts.envs = {
      ...projConfigOpts.envs,
      ...envs
    };
  } else {
    projConfigOpts = {
      ...projConfigOpts,
      envs: envs
    };
  }
  const isCmdOptsReportFormat = cmdOpts.reporting && cmdOpts.reporting.reporters;
  const isConfigReportFormat = projConfigOpts.reporting && projConfigOpts.reporting.reporters;
  if (!isCmdOptsReportFormat && !isConfigReportFormat) {
    if (!projConfigOpts.reporting) {
      projConfigOpts.reporting = {};
    }
    projConfigOpts.reporting.reporters = [DEFAULT_REPORTER];
  }
  let name = cmdOpts.name || projConfigOpts.name || null;
  if (!name && target) {
    name = target.name !== OXYGEN_CONFIG_FILE_NAME ? target.name : target.baseName;
  }
  return {
    ...DEFAULT_OPTS,
    ...projConfigOpts,
    ...cmdOpts,
    name: name
  };
}
function getCommandLineOptions(argv) {
  const opts = {
    name: argv.name || null,
    browserName: argv.b || argv.browser || null,
    seleniumUrl: argv.s || argv.server || null,
    appiumUrl: argv.s || argv.server || null,
    reopenSession: argv.reopen ? argv.reopen === 'true' : null,
    iterations: argv.i ? parseInt(argv.i) : argv.iter ? parseInt(argv.iter) : null,
    debugPort: argv.dbgport || null,
    wsPort: argv.wsport || null,
    delay: argv.d || argv.delay || null,
    autoStartWebDriver: argv.autowd && argv.autowd === 'true' || false
  };
  if (argv.rf && typeof argv.rf === 'string' && argv.rf.length > 0) {
    const reportFormats = argv.rf.split(',');
    opts.reporting = {
      reporters: reportFormats
    };
    if (argv.ro && typeof argv.ro === 'string' && argv.ro.length > 0) {
      opts.reporting.outputDir = argv.ro;
    }
  }
  if (argv.p || argv.param) {
    opts.parameters = {
      file: argv.p || argv.param || null,
      mode: argv.pm || 'seq'
    };
  }
  if (argv.specs && typeof argv.specs === 'string' && argv.specs.length > 0) {
    const specs = argv.specs.split(',');
    opts.specs = specs;
  }
  if (argv.modules && typeof argv.modules === 'string' && argv.modules.length > 0) {
    const modules = argv.modules.split(',');
    opts.modules = modules;
  }
  return deleteNullProperties(opts);
}
function deleteNullProperties(obj) {
  if (typeof obj !== 'object') {
    return obj;
  }
  const clone = {
    ...obj
  };
  const keys = Object.keys(obj);
  (0, _async.forEach)(keys, key => {
    if (Object.prototype.hasOwnProperty.call(obj, key) && (obj[key] == null || obj[key] == undefined)) {
      delete clone[key];
    }
  });
  return clone;
}
function processTargetPath(targetPath, userCwd) {
  let cwd = userCwd || process.cwd();
  if (typeof targetPath === 'undefined') {
    targetPath = cwd;
  } else {
    targetPath = _util.default.resolvePath(targetPath, cwd);
  }
  if (!_fs.default.existsSync(targetPath)) {
    return null;
  }
  const stats = _fs.default.lstatSync(targetPath);
  const isDirector = stats.isDirectory();
  let configFilePath;
  if (isDirector) {
    cwd = userCwd || targetPath;
    configFilePath = _path.default.join(targetPath, OXYGEN_CONFIG_FILE_NAME + '.js');
    if (!_fs.default.existsSync(configFilePath)) {
      configFilePath = _path.default.join(targetPath, OXYGEN_CONFIG_FILE_NAME + '.json');
      if (!_fs.default.existsSync(configFilePath)) {
        return null;
      }
    }
  } else if (targetPath.endsWith(OXYGEN_CONFIG_FILE_NAME + '.js') || targetPath.endsWith(OXYGEN_CONFIG_FILE_NAME + '.json')) {
    configFilePath = targetPath;
    cwd = targetPath = _path.default.dirname(targetPath);
  } else {
    cwd = userCwd || _path.default.dirname(targetPath);
    configFilePath = _path.default.join(cwd, OXYGEN_CONFIG_FILE_NAME + '.js');
    if (!_fs.default.existsSync(configFilePath)) {
      configFilePath = _path.default.join(cwd, OXYGEN_CONFIG_FILE_NAME + '.json');
      if (!_fs.default.existsSync(configFilePath)) {
        configFilePath = undefined;
      }
    }
  }
  if (!targetPath) {
    return null;
  }
  return {
    path: targetPath,
    configPath: configFilePath,
    cwd: cwd || _path.default.dirname(targetPath),
    name: _util.default.getFileNameWithoutExt(targetPath),
    fullName: _path.default.basename(targetPath),
    baseName: _path.default.basename(_path.default.dirname(targetPath)),
    extension: _path.default.extname(targetPath)
  };
}
