// oxygen-cli/build/lib/parallel-launcher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var Runners = _interopRequireWildcard(require("../runners"));
var _parallelLimit = _interopRequireDefault(require("async/parallelLimit"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
const Duration = require('duration');
const hash = require('object-hash');
const {
  v1
} = require('uuid');
class ParallelLauncher {
  constructor(config, reporter) {
    this._config = config;
    this.reporter = reporter;
    this._queue = null;
    this.runners = [];
    this.createdRunnersStat = [];
  }
  async run(capsSet) {
    const collection = [];
    if (!this._config.parallel) {
      throw new Error('Cannot start the parallel testing - "parallel" settings are missing.');
    }
    if (!this._config.parallel.mode) {
      this._config.parallel.mode = 'case';
    }
    const workersCount = !isNaN(this._config.parallel.workers) ? this._config.parallel.workers : 1;
    const mode = this._config.parallel.mode === 'iteration' ? 'iteration' : this._config.parallel.mode === 'suite' ? 'suite' : 'case';
    const suites = this._config.suites;
    const resultName = this._config.name || '';
    if (!suites || !Array.isArray(suites)) {
      throw new Error('Cannot start the parallel testing - no suites are defined.');
    }
    const testCaps = capsSet ? Array.isArray(capsSet) ? capsSet[0] : capsSet : null;
    const resultHash = testCaps ? hash(testCaps) : '';
    const resultKey = `${resultName}-${resultHash}`;
    let suiteIndex = 0;
    suites.forEach(suiteDef => {
      const suiteKey = suiteDef.key || suiteDef.id || suiteDef.name;
      const suiteRefId = suiteDef.key || suiteDef.id || v1();
      if (mode === 'iteration') {
        for (let i = 1; i <= suiteDef.iterationCount; i++) {
          let mockedParamManager = undefined;
          if (suiteDef.paramManager) {
            mockedParamManager = new ParamManagerMock(suiteDef.paramManager.getValues());
            suiteDef.paramManager.readNext();
          }
          const workerId = `${suiteKey}-${suiteIndex}/${i}`;
          const suiteCopy = {
            ...suiteDef,
            refId: suiteRefId,
            iterationCount: 1,
            paramManager: mockedParamManager
          };
          const testConfig = {
            ...this._config,
            suites: [suiteCopy],
            _groupResult: {
              resultKey
            }
          };
          const workerConfig = {
            workerId,
            testConfig,
            testCaps,
            startupDelay: 200
          };
          collection.push(callback => this._launchWorkerFunc(workerConfig, callback));
        }
      } else if (mode === 'case' && Array.isArray(suiteDef.cases)) {
        let caseIndex = 0;
        suiteDef.cases.forEach(caseDef => {
          const caseKey = caseDef.key || caseDef.id || caseDef.name;
          const workerId = `${caseKey}-${suiteIndex}/${caseIndex}`;
          const suiteCopy = {
            ...suiteDef,
            refId: suiteRefId,
            cases: [caseDef]
          };
          const testConfig = {
            ...this._config,
            suites: [suiteCopy],
            _groupResult: {
              resultKey,
              suiteKey
            }
          };
          const workerConfig = {
            workerId,
            testConfig,
            testCaps,
            startupDelay: 0
          };
          collection.push(callback => this._launchWorkerFunc(workerConfig, callback));
          caseIndex++;
        });
      } else {
        const suiteKey = suiteDef.key || suiteDef.id || suiteDef.name;
        const workerId = `${suiteKey}-${suiteIndex}`;
        const testConfig = {
          ...this._config,
          suites: [{
            ...suiteDef,
            refId: suiteRefId
          }],
          _groupResult: {
            resultKey
          }
        };
        const workerConfig = {
          workerId,
          testConfig,
          testCaps,
          startupDelay: 0
        };
        collection.push(callback => this._launchWorkerFunc(workerConfig, callback));
      }
      suiteIndex++;
    });
    const start = new Date();
    await (0, _parallelLimit.default)(collection, workersCount);
    let total = 0;
    console.log('Created Runners Stat', this.createdRunnersStat);
    const runnersIds = Object.keys(this.createdRunnersStat);
    console.log('Created Runners count', runnersIds.length);
    runnersIds.map(item => {
      const cazes = this.createdRunnersStat[item].length;
      total += cazes;
      console.log('Runner', item, ' run ', cazes, ' cases');
    });
    console.log('Parallel count', workersCount);
    console.log('Total: ', total);
    const end = new Date();
    const duration = new Duration(start, end);
    console.log('Duration: ', duration.toString(1));
  }
  _launchWorkerFunc(workerConfig, callback) {
    const cb = error => {
      if (error) {
        callback(error, null);
      }
    };
    this._launchTest(workerConfig, cb).then(result => {
      callback(null, result);
    }).catch(error => {
      callback(error, null);
    });
  }
  _instantiateRunner() {
    if (this._config.framework && typeof this._config.framework === 'string') {
      if (Object.prototype.hasOwnProperty.call(Runners, this._config.framework)) {
        return new Runners[this._config.framework]();
      }
      return null;
    }
    return new Runners.oxygen();
  }
  async _launchTest({
    workerId,
    testConfig,
    testCaps,
    startupDelay
  }, callback) {
    if (!callback) {
      return;
    }
    await this._sleep(startupDelay);
    let runner = this._instantiateRunner();
    this.createdRunnersStat[runner._id] = [workerId];
    if (!runner) {
      const framework = this._config.framework;
      callback(new Error(`Cannot find runner for the specified framework: ${framework}.`));
      return;
    }
    try {
      await runner.init(testConfig, testCaps, this.reporter);
      const result = await runner.run();
      await runner.dispose(result.status || null);
      callback();
    } catch (e) {
      console.error('Failed to launch the test:', e);
      if (this._queue) {
        this._queue.kill();
      }
      if (e.error) {
        var errMsg = '';
        var err = e.error;
        if (err.type) errMsg += err.type + ' - ';
        if (err.message) errMsg += err.message;else errMsg = err.toString();
        callback(new Error(errMsg));
      } else {
        callback(e);
      }
    }
    this.runners.push(runner);
  }
  _sleep(duration) {
    return new Promise(resolve => setTimeout(resolve, duration));
  }
}
exports.default = ParallelLauncher;
class ParamManagerMock {
  constructor(values) {
    this.values = values;
  }
  init() {}
  getMode() {}
  readNext() {}
  readPrev() {}
  getValues() {
    return this.values;
  }
  get rows() {
    return 1;
  }
}
