// oxygen-cli/build/reporter/FileReporterBase.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _fs = _interopRequireDefault(require("fs"));
var _path = _interopRequireDefault(require("path"));
var _moment = _interopRequireDefault(require("moment"));
var _ReporterBase = _interopRequireDefault(require("./ReporterBase"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const REPORTS_FOLDER_NAME = 'reports';
const REPORT_FILE_NAME = 'report';
class FileReporterBase extends _ReporterBase.default {
  constructor(options) {
    super(options);
  }
  createFolderStructureAndFilePath(fileExtension) {
    if (!fileExtension || typeof fileExtension !== 'string' || fileExtension.length == 0) {
      throw new Error('"fileExtension" argument must be specified');
    }
    if (!fileExtension.startsWith('.')) {
      fileExtension = `.${fileExtension}`;
    }
    if (!this.options || !this.options.target || !this.options.reporting) {
      throw new Error('FileReporterBase is not properly initialized');
    }
    let resultsBaseFolder = null;
    if (this.options.cwd && this.options.reporting.outputDir) {
      if (_path.default.isAbsolute(this.options.reporting.outputDir)) {
        resultsBaseFolder = this.options.reporting.outputDir;
      } else {
        resultsBaseFolder = _path.default.resolve(this.options.cwd, this.options.reporting.outputDir);
      }
    } else if (this.options.target) {
      resultsBaseFolder = _path.default.join(this.options.target.cwd, REPORTS_FOLDER_NAME);
    } else {
      throw new Error('Error constructing reports path. Either srcFile or outputFolder is required.');
    }
    this.createFolderIfNotExists(resultsBaseFolder);
    let resultFolderPath = resultsBaseFolder;
    if (!this.options.reporting.outputDir) {
      const subFolderName = (0, _moment.default)().format('YYYY-MM-DD_HHmmss');
      resultFolderPath = _path.default.join(resultsBaseFolder, subFolderName);
      this.createFolderIfNotExists(resultFolderPath);
    }
    return _path.default.join(resultFolderPath, `${REPORT_FILE_NAME}${fileExtension}`);
  }
  createFolderIfNotExists(folderPath) {
    try {
      _fs.default.mkdirSync(folderPath);
    } catch (e) {
      if (e.code != 'EEXIST') throw e;
    }
    return folderPath;
  }
  replaceScreenshotsWithFiles(results, folderPath) {
    if (!Array.isArray(results)) {
      throw new Error('Invalid argument "results" - must be an array.');
    }
    if (!folderPath) {
      throw new Error('"folderPath" argument cannot be null or empty.');
    }
    const stepsWithScreenshot = [];
    for (let result of results) {
      for (let suite of result.suites) {
        for (let caze of suite.cases) {
          this._populateStepsWithScreenshots(caze.steps, stepsWithScreenshot);
        }
      }
    }
    const screenshotFilePrefix = 'screenshot-';
    const screenshotFileSuffix = '.png';
    for (let i = 0; i < stepsWithScreenshot.length; i++) {
      let filename = screenshotFilePrefix + i + screenshotFileSuffix;
      let filepath = _path.default.join(folderPath, filename);
      let step = stepsWithScreenshot[i];
      _fs.default.writeFileSync(filepath, step.screenshot, 'base64');
      step.screenshotFile = filename;
      step.screenshot = null;
    }
  }
  formatTotalDuration(results) {
    if (!Array.isArray(results)) {
      throw new Error('Invalid argument "results" - must be an array.');
    }
    for (let result of results) {
      for (let suite of result.suites) {
        for (let caze of suite.cases) {
          for (let step of caze.steps) {
            step.duration = this.formatDuration(step.duration);
          }
        }
      }
    }
  }
  _populateStepsWithScreenshots(steps, stepsWithScreenshot) {
    for (let step of steps) {
      if (step.screenshot) {
        stepsWithScreenshot.push(step);
      }
      if (step.steps) {
        this._populateStepsWithScreenshots(step.steps, stepsWithScreenshot);
      }
    }
  }
  formatDuration(miliseconds) {
    var seconds = miliseconds / 1000;
    var hh = Math.floor(seconds / 3600);
    var min = Math.floor((seconds - hh * 3600) / 60);
    var sec = seconds - hh * 3600 - min * 60;
    let result = '';
    if (hh) {
      result += `${hh.toFixed(0)} hr `;
    }
    if (min) {
      result += `${min.toFixed(0)} min `;
    }
    if (sec) {
      result += `${sec.toFixed(0)} sec`;
    }
    return result;
  }
}
exports.default = FileReporterBase;
