// oxygen-cli/build/reporter/ReportAggregator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _events = require("events");
var _path = _interopRequireDefault(require("path"));
var _testResult = _interopRequireDefault(require("../model/test-result"));
var _util = _interopRequireDefault(require("../lib/util"));
var _when = require("when");
var _reporterJson = _interopRequireDefault(require("../ox_reporters/reporter-json"));
var _reporterJunit = _interopRequireDefault(require("../ox_reporters/reporter-junit"));
var _reporterHtml = _interopRequireDefault(require("../ox_reporters/reporter-html"));
var _reporterExcel = _interopRequireDefault(require("../ox_reporters/reporter-excel"));
var _reporterPdf = _interopRequireDefault(require("../ox_reporters/reporter-pdf"));
var _reporterXml = _interopRequireDefault(require("../ox_reporters/reporter-xml"));
var _helper = _interopRequireDefault(require("../errors/helper"));
var _status = _interopRequireDefault(require("../model/status"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
const Reporters = {
  json: _reporterJson.default,
  junit: _reporterJunit.default,
  html: _reporterHtml.default,
  excel: _reporterExcel.default,
  pdf: _reporterPdf.default,
  xml: _reporterXml.default
};
const DEFAULT_TEST_NAME = 'Oxygen Test';
const DEFAULT_REPORTERS = [];
class ReportAggregator extends _events.EventEmitter {
  constructor(options) {
    super();
    this.results = [];
    this.runnerEndPromises = {};
    this.options = options;
  }
  getExitCode() {
    let exitCode = 0;
    if (this.results && Array.isArray(this.results) && this.results.length > 0) {
      const testFailded = this.results.find(item => item.status === 'failed');
      if (testFailded) {
        exitCode = -1;
      }
    } else {
      exitCode = -1;
    }
    return exitCode;
  }
  async init() {
    if (!this.options._oxide) {
      const {
        default: ReportPortalReporter
      } = await Promise.resolve().then(() => _interopRequireWildcard(require('../ox_reporters/reporter-rp')));
      Reporters.rp = ReportPortalReporter;
    }
    this.reporters = [];
    const generalReportingOpts = this.options.reporting || {};
    for (let reporter of generalReportingOpts.reporters || DEFAULT_REPORTERS) {
      if (typeof reporter !== 'string' && !Object.prototype.hasOwnProperty.call(reporter, 'name')) {
        continue;
      }
      const reporterName = typeof reporter === 'string' ? reporter : reporter.name;
      const reporterOpts = typeof reporter === 'object' ? reporter : undefined;
      if (Object.prototype.hasOwnProperty.call(Reporters, reporterName)) {
        if (reporterOpts && typeof reporterOpts.enabled === 'boolean' && !reporterOpts.enabled) {
          continue;
        }
        const reporter = new Reporters[reporterName](this.options, reporterOpts, this);
        if (reporter.init) {
          try {
            await reporter.init();
          } catch (err) {
            console.log(`Failed to initialize "${reporterName}" reporter: ${err.message}`);
          }
        }
        this.reporters.push(reporter);
      }
    }
  }
  getResults() {
    return this.results;
  }
  async generateReports() {
    if (!Array.isArray(this.reporters) || this.reporters.length == 0) {
      return false;
    }
    const groupedResults = this.groupResults();
    for (let reporter of this.reporters) {
      try {
        const reportPath = await reporter.generate(groupedResults);
        if (reportPath) {
          console.log(`Your report is ready: ${reportPath}`);
        }
      } catch (e) {
        console.error(`Report generation failed: ${e.message}`);
      }
    }
    return true;
  }
  async waitForResult(rid) {
    if (!rid || !this.runnerEndPromises[rid]) {
      return null;
    }
    return this.runnerEndPromises[rid];
  }
  async waitForResults() {
    if (this.runnerEndPromises && this.runnerEndPromises.length) {
      return Promise.all(this.runnerEndPromises);
    }
  }
  async onLaunchStart(options) {
    const eventArgs = {
      options
    };
    await this._invokeReportersHook('onLaunchStart', eventArgs);
  }
  async onLaunchEnd() {
    const results = this.groupResults();
    const eventArgs = {
      results
    };
    this.emit('launch:end', results);
    await this._invokeReportersHook('onLaunchEnd', eventArgs);
  }
  async onRunnerStart(rid, opts, caps) {
    if (!rid) {
      throw new Error('"rid" cannot be empty.');
    }
    const testResult = new _testResult.default();
    testResult.rid = rid;
    testResult.name = opts.name || DEFAULT_TEST_NAME;
    testResult.startTime = _util.default.getTimeStamp();
    testResult.capabilities = caps;
    testResult.environment = opts.envVars || {};
    testResult.options = opts;
    this.results.push(testResult);
    this.runnerEndPromises[rid] = (0, _when.defer)();
    console.log(`Test ${rid} has started...`);
    const eventArgs = {
      rid,
      opts,
      caps
    };
    this.emit('runner:start', eventArgs);
    await this._invokeReportersHook('onRunnerStart', eventArgs);
  }
  async onRunnerEnd(rid, finalResult, fatalError) {
    const testResult = this.results.find(x => x.rid === rid);
    if (testResult) {
      testResult.endTime = _util.default.getTimeStamp();
      testResult.duration = testResult.endTime - testResult.startTime;
      testResult.status = this._determineTestStatusBySuites(testResult, fatalError);
      if (testResult.status === _status.default.FAILED) {
        if (fatalError) {
          testResult.failure = fatalError instanceof Error ? _helper.default.getFailureFromError(fatalError) : fatalError;
        } else {
          testResult.failure = this._getFirstFailure(testResult);
        }
      }
      if (finalResult && finalResult.capabilities) {
        testResult.capabilities = finalResult.capabilities;
      }
      if (testResult.failure) {
        if (testResult.failure.type && testResult.failure.location) {
          console.log(`Error: ${testResult.failure.type} at ${testResult.failure.location}.`);
        } else if (testResult.failure.type) {
          console.log(`Error: ${testResult.failure.type}`);
        } else if (typeof testResult.failure === 'string') {
          console.log(`Error: ${testResult.failure}`);
        }
      }
      console.log(`Test ${rid} has finished with status: ${testResult.status.toUpperCase()}.`);
    }
    const eventArgs = {
      rid,
      result: testResult
    };
    this.emit('runner:end', eventArgs);
    await this._invokeReportersHook('onRunnerEnd', eventArgs);
    if (this.runnerEndPromises[rid]) {
      process.nextTick(() => {
        this.runnerEndPromises[rid].resolve(testResult);
      });
    }
  }
  _determineTestStatusBySuites(testResult, fatalError) {
    if (fatalError) {
      return _status.default.FAILED;
    }
    const hasFailedSuites = testResult.suites.some(x => x.status === _status.default.FAILED);
    if (hasFailedSuites) {
      return _status.default.FAILED;
    }
    const hasWarningSuites = testResult.suites.some(x => x.status === _status.default.WARNING);
    if (hasWarningSuites) {
      return _status.default.WARNING;
    }
    const allSuitesSkipped = testResult.suites.every(x => x.status === _status.default.SKIPPED);
    if (allSuitesSkipped) {
      return _status.default.SKIPPED;
    }
    return _status.default.PASSED;
  }
  async onIterationStart(rid, iteration, start) {
    if (iteration) {
      const msg = `${start} Iteration #${iteration} started...`;
      console.log(msg);
      await this.onLogEntry(null, 'INFO', msg, 'user');
    }
  }
  async onIterationEnd(rid, result, start) {
    var _result$status;
    if (result !== null && result !== void 0 && result.iterationNum && (_result$status = result.status) !== null && _result$status !== void 0 && _result$status.toUpperCase) {
      const msg = `${start} Iteration #${result.iterationNum} ended with status: ${result.status.toUpperCase()}.`;
      console.log(msg);
      await this.onLogEntry(null, 'INFO', msg, 'user');
    }
  }
  async onSuiteStart(rid, suiteId, suite) {
    console.log(`Suite "${suite.name}" has started...`);
    let eventArgs = {
      rid,
      suiteId: suiteId,
      suite: suite
    };
    this.emit('suite:start', eventArgs);
    await this._invokeReportersHook('onSuiteStart', eventArgs);
  }
  async onSuiteEnd(rid, suiteId, suiteResult) {
    const testResult = this.results.find(x => x.rid === rid);
    if (!testResult) {
      return;
    }
    testResult.suites.push(suiteResult);
    console.log(`Suite "${suiteResult.name}" has ended with status: ${suiteResult.status.toUpperCase()}.`);
    const eventArgs = {
      rid,
      suiteId,
      result: suiteResult
    };
    this.emit('suite:end', eventArgs);
    await this._invokeReportersHook('onSuiteEnd', eventArgs);
  }
  async onCaseStart(rid, suiteId, suiteRefId, caseId, caseDef) {
    console.log(`- Case "${caseDef.name}" has started...`);
    const eventArgs = {
      rid,
      suiteId,
      suiteRefId,
      caseId,
      case: caseDef
    };
    this.emit('case:start', eventArgs);
    await this._invokeReportersHook('onCaseStart', eventArgs);
  }
  async onCaseEnd(rid, suiteId, suiteRefId, caseId, caseResult) {
    console.log(`- Case "${caseResult.name}" has ended with status: ${caseResult.status.toUpperCase()}.`);
    await this._saveTestCaseVideoAttachment(caseResult);
    const eventArgs = {
      rid,
      suiteId,
      suiteRefId,
      caseId,
      result: caseResult
    };
    this.emit('case:end', eventArgs);
    await this._invokeReportersHook('onCaseEnd', eventArgs);
  }
  async onStepStart(rid, suiteId, caseId, step) {
    if (this.options && this.options.rootPath && this.options.framework && this.options.framework === 'cucumber') {
      const fullPath = _path.default.resolve(this.options.rootPath, step.location);
      step.location = fullPath + ':1';
    }
    const eventArgs = {
      rid,
      suiteId,
      caseId,
      step: step
    };
    this.emit('step:start', eventArgs);
    await this._invokeReportersHook('onStepStart', eventArgs);
  }
  async onStepEnd(rid, suiteId, caseId, stepResult) {
    const eventArgs = {
      rid,
      suiteId,
      caseId,
      step: stepResult
    };
    this.emit('step:end', eventArgs);
    await this._invokeReportersHook('onStepEnd', eventArgs);
  }
  async onLogEntry(time, level, msg, src = null, {
    suiteId,
    caseId,
    stepId
  } = {}) {
    const eventArgs = {
      suiteId,
      caseId,
      stepId,
      level,
      msg,
      time,
      src
    };
    this.emit('log', eventArgs);
    if (!time) {
      eventArgs.time = Date.now();
    }
    await this._invokeReportersHook('onLog', eventArgs);
  }
  _getFirstFailure(testResult) {
    if (testResult.status === _status.default.FAILED) {
      if (testResult.failure) {
        return testResult.failure;
      }
      for (let suiteResult of testResult.suites) {
        if (suiteResult.status !== _status.default.FAILED) {
          continue;
        }
        if (suiteResult.failure) {
          return suiteResult.failure;
        }
        for (let caseResult of suiteResult.cases) {
          if (suiteResult.status !== _status.default.FAILED) {
            continue;
          }
          if (caseResult.failure) {
            return caseResult.failure;
          }
          for (let stepResult of caseResult.steps) {
            if (stepResult.status !== _status.default.FAILED) {
              continue;
            }
            return stepResult.failure || null;
          }
        }
      }
    }
    return null;
  }
  groupResults() {
    const groupedResults = {};
    const ungroupedResults = [];
    if (!Array.isArray(this.results) || this.results.length == 0) {
      return false;
    }
    for (let result of this.results) {
      if (!result.options._groupResult || !result.options._groupResult.resultKey) {
        ungroupedResults.push(result);
        continue;
      }
      const resultKey = result.options._groupResult.resultKey;
      const suiteKey = result.options._groupResult.suiteKey;
      let groupedResult = groupedResults[resultKey];
      if (!groupedResult) {
        groupedResult = groupedResults[resultKey] = {
          ...result,
          suites: []
        };
        if (suiteKey) {
          groupedResult['_suitesHash'] = {};
        }
      } else {
        groupedResult.startTime = Math.min(groupedResult.startTime, result.startTime);
        groupedResult.endTime = Math.max(groupedResult.endTime, result.endTime);
        groupedResult.duration = groupedResult.endTime - groupedResult.startTime;
      }
      if (!suiteKey) {
        groupedResult.suites = [...groupedResult.suites, ...result.suites];
      } else {
        for (const currentSuiteResult of result.suites) {
          const groupKey = `${suiteKey}-${currentSuiteResult.iterationNum}`;
          const groupedSuiteResult = groupedResult._suitesHash[groupKey];
          if (!groupedSuiteResult) {
            groupedResult._suitesHash[groupKey] = {
              ...currentSuiteResult
            };
          } else {
            groupedSuiteResult.startTime = Math.min(groupedSuiteResult.startTime, currentSuiteResult.startTime);
            groupedSuiteResult.endTime = Math.max(groupedSuiteResult.endTime, currentSuiteResult.endTime);
            groupedSuiteResult.duration = groupedSuiteResult.endTime - groupedSuiteResult.startTime;
            groupedSuiteResult.cases = [...groupedSuiteResult.cases, ...currentSuiteResult.cases];
            if (currentSuiteResult.cases.some(c => c.status === _status.default.FAILED)) {
              groupedSuiteResult.status = _status.default.FAILED;
            } else if (currentSuiteResult.cases.some(c => c.status === _status.default.WARNING)) {
              groupedSuiteResult.status = _status.default.WARNING;
            }
          }
        }
      }
    }
    const groupedResultsList = Object.keys(groupedResults).map(groupKey => {
      const groupedResult = groupedResults[groupKey];
      if (!groupedResult._suitesHash) {
        return groupedResult;
      }
      groupedResult.suites = [...groupedResult.suites, ...Object.keys(groupedResult._suitesHash).map(suiteGroupKey => groupedResult._suitesHash[suiteGroupKey])];
      const firstFailedSuite = groupedResult.suites.find(s => s.status === _status.default.FAILED);
      if (firstFailedSuite) {
        groupedResult.status = _status.default.FAILED;
        groupedResult.failure = firstFailedSuite.failure || groupedResult.failure;
      } else {
        const firstSuiteWithWarning = groupedResult.suites.find(s => s.status === _status.default.WARNING);
        if (firstSuiteWithWarning) {
          groupedResult.status = _status.default.WARNING;
          groupedResult.failure = firstSuiteWithWarning.failure || groupedResult.failure;
        }
      }
      delete groupedResult['_suitesHash'];
      return groupedResult;
    });
    let results = [...groupedResultsList, ...ungroupedResults];
    results = this.recalculateResultForStatus(results);
    this.validateResult(results);
    return results;
  }
  recalculateResultForStatus(results) {
    return results.map(result => {
      if (result && result.suites && Array.isArray(result.suites) && result.suites.length > 0) {
        const failed = result.suites.find(suite => suite.status === _status.default.FAILED);
        if (failed) {
          result.status = _status.default.FAILED;
        } else if (result.suites.find(suite => suite.status === _status.default.WARNING)) {
          result.status = _status.default.WARNING;
        }
      }
      return result;
    });
  }
  validateResult(results) {
    const uniqueSuitesIterationIds = [];
    results.map(result => {
      result.suites.map(suite => {
        if (uniqueSuitesIterationIds.includes(suite.iterationNum)) {
          console.warn('suite.iterationNum', suite.iterationNum, ' not unique');
        } else {
          uniqueSuitesIterationIds.push(suite.iterationNum);
        }
      });
    });
  }
  async _invokeReportersHook(hookName, eventArgs) {
    if (!Array.isArray(this.reporters) || this.reporters.length == 0) {
      return false;
    }
    for (let reporter of this.reporters) {
      if (reporter[hookName]) {
        try {
          await reporter[hookName](eventArgs);
        } catch (e) {
          console.warn(`Failed to invoke reporter hook "${hookName}": ${e.message}`);
        }
      }
    }
  }
  async _saveTestCaseVideoAttachment(caseResult) {
    if (!caseResult || !caseResult.attachments || !caseResult.attachments.length) {
      return;
    }
    const attachmentsToRemove = [];
    for (let i = 0; i < caseResult.attachments.length; i++) {
      const attachment = caseResult.attachments[i];
      if (!attachment._url) {
        continue;
      }
      try {
        const videoFilePath = await _util.default.downloadVideo(attachment.fileName, attachment._url, this.options);
        if (!videoFilePath) {
          attachmentsToRemove.push(attachment);
        } else {
          attachment.filePath = videoFilePath;
          delete attachment['_url'];
        }
      } catch (e) {
        console.warn('Failed to download video file: ', e.message);
        attachmentsToRemove.push(attachment);
      }
    }
    for (let i = 0; i < attachmentsToRemove.length; i++) {
      const elmIndexToRemove = caseResult.attachments.indexOf(attachmentsToRemove[i]);
      if (!elmIndexToRemove > -1) {
        caseResult.attachments.splice(elmIndexToRemove, 1);
      }
    }
  }
}
exports.default = ReportAggregator;
