// oxygen-cli/build/reporter/WebSocketReporter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _socket = require("socket.io");
const io = new _socket.Server({
  serveClient: false,
  pingInterval: 10000,
  pingTimeout: 60000,
  maxHttpBufferSize: 256 * 1024 * 1024,
  cookie: true,
  cors: {
    origin: '*'
  },
  transports: ['websocket']
});
class WebSocketReporter {
  constructor(aggregator) {
    this._aggregator = aggregator;
    this._hookAggregatorEvents();
  }
  startAndWaitForClient(serverPort) {
    io.listen(serverPort);
    return new Promise((resolve, reject) => {
      this._handleServerEvents(resolve);
    });
  }
  stop() {
    if (this._socket) {
      io.close();
      this._socket = undefined;
    }
  }
  _handleServerEvents(resolveWait) {
    const _this = this;
    io.on('connection', socket => {
      _this._handleSocketEvents(socket);
      resolveWait();
    });
  }
  _handleSocketEvents(socket) {
    this._socket = socket;
    socket.on('disconnect', reason => {});
    socket.on('error', err => {
      console.error('Error in Socket:', err);
    });
  }
  _hookAggregatorEvents() {
    this._aggregator.on('runner:start', e => this._handleRunnerStart(e));
    this._aggregator.on('runner:end', e => this._handleRunnerEnd(e));
    this._aggregator.on('suite:start', e => this._handleSuiteStart(e));
    this._aggregator.on('suite:end', e => this._handleSuiteEnd(e));
    this._aggregator.on('case:start', e => this._handleCaseStart(e));
    this._aggregator.on('case:end', e => this._handleCaseEnd(e));
    this._aggregator.on('log', e => this._handleLog(e));
  }
  _handleRunnerStart(event) {
    var _this$_socket;
    (_this$_socket = this._socket) === null || _this$_socket === void 0 ? void 0 : _this$_socket.emit('runner:start', event);
  }
  _handleRunnerEnd(event) {
    var _this$_socket2;
    (_this$_socket2 = this._socket) === null || _this$_socket2 === void 0 ? void 0 : _this$_socket2.emit('runner:end', event);
  }
  _handleSuiteStart(event) {
    var _this$_socket3;
    (_this$_socket3 = this._socket) === null || _this$_socket3 === void 0 ? void 0 : _this$_socket3.emit('suite:start', event);
  }
  _handleSuiteEnd(event) {
    var _this$_socket4;
    (_this$_socket4 = this._socket) === null || _this$_socket4 === void 0 ? void 0 : _this$_socket4.emit('suite:end', event);
  }
  _handleCaseStart(event) {
    var _this$_socket5;
    (_this$_socket5 = this._socket) === null || _this$_socket5 === void 0 ? void 0 : _this$_socket5.emit('case:start', event);
  }
  _handleCaseEnd(event) {
    var _this$_socket6;
    (_this$_socket6 = this._socket) === null || _this$_socket6 === void 0 ? void 0 : _this$_socket6.emit('case:end', event);
  }
  _handleStepStart(event) {
    var _this$_socket7;
    (_this$_socket7 = this._socket) === null || _this$_socket7 === void 0 ? void 0 : _this$_socket7.emit('step:start', event);
  }
  _handleStepEnd(event) {
    var _this$_socket8;
    (_this$_socket8 = this._socket) === null || _this$_socket8 === void 0 ? void 0 : _this$_socket8.emit('step:end', event);
  }
  _handleLog(event) {
    var _this$_socket9;
    (_this$_socket9 = this._socket) === null || _this$_socket9 === void 0 ? void 0 : _this$_socket9.emit('log', event);
  }
}
exports.default = WebSocketReporter;
