// oxygen-cli/build/ox_reporters/reporter-excel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _path = _interopRequireDefault(require("path"));
var _lodash = _interopRequireDefault(require("lodash"));
var _xlsx = _interopRequireDefault(require("xlsx"));
var _FileReporterBase = _interopRequireDefault(require("../reporter/FileReporterBase"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const DEFAULT_TEMPLATE_PATH = './excel/template.json';
const DEFAULT_SHEET_NAME = 'Sheet1';
const DEFAULT_EXTENSION = '.xlsx';
const EXCEL_COLUMNS = generateExcelColumns();
class ExcelReporter extends _FileReporterBase.default {
  constructor(options) {
    super(options);
  }
  generate(results) {
    var resultFilePath = this.createFolderStructureAndFilePath(DEFAULT_EXTENSION);
    var resultFolderPath = _path.default.dirname(resultFilePath);
    var rows = [];
    this.replaceScreenshotsWithFiles(results, resultFolderPath);
    for (let result of results) {
      for (let suiteResult of result.suites) {
        generateRowsForSuite(suiteResult, rows);
      }
    }
    var templatePath = this.options.template || DEFAULT_TEMPLATE_PATH;
    var workbook = generateWorksheetFromTemplate(templatePath, rows, this.options);
    _xlsx.default.writeFile(workbook, resultFilePath, {
      bookSST: false
    });
    return resultFilePath;
  }
}
exports.default = ExcelReporter;
function generateRowsForSuite(suiteResult, rows) {
  for (let caseResult of suiteResult.cases) {
    let lastFailedStep = null;
    for (let step of caseResult.steps) {
      if (step.status === 'failed') {
        lastFailedStep = step;
      } else {
        lastFailedStep = null;
      }
    }
    rows.push(addValues(suiteResult, caseResult, lastFailedStep));
  }
}
function generateWorksheetFromTemplate(templatePath, rows, options) {
  var template = null;
  try {
    template = require(templatePath);
  } catch (e) {
    throw new Error('Excel reporter template file is not found or is in invalid format: ' + templatePath + '. Error: ' + e.message);
  }
  var wb = new Workbook();
  var ws = {};
  var wsName = options.sheetName || DEFAULT_SHEET_NAME;
  wb.SheetNames.push(wsName);
  wb.Sheets[wsName] = ws;
  var column = 0;
  var row;
  var header;
  var cellAddress;
  for (var h = 0; h < template.length; h++) {
    header = template[h];
    if (header.value === '${param.all}' && rows && rows.length && rows.length > 0) {
      row = rows[0];
      _lodash.default.each(row.param, function (value, name) {
        cellAddress = EXCEL_COLUMNS[column] + '1';
        ws[cellAddress] = {};
        ws[cellAddress].v = name;
        column++;
      });
    } else if (header.value === '${var.all}' && rows && rows.length && rows.length > 0) {
      row = rows[0];
      _lodash.default.each(row.var, function (value, name) {
        cellAddress = EXCEL_COLUMNS[column] + '1';
        ws[cellAddress] = {};
        ws[cellAddress].v = name;
        column++;
      });
    } else {
      cellAddress = EXCEL_COLUMNS[column] + '1';
      ws[cellAddress] = {};
      ws[cellAddress].v = header.header;
      column++;
    }
  }
  var totalColumns = column;
  for (var r = 0; r < rows.length; r++) {
    column = 0;
    for (var c = 0; c < template.length; c++) {
      row = rows[r];
      header = template[c];
      if (header.value === '${param.all}') {
        if (!row.param || !row.param) {
          continue;
        }
        _lodash.default.each(row.param, function (value, name) {
          cellAddress = EXCEL_COLUMNS[column] + (r + 2);
          ws[cellAddress] = {};
          ws[cellAddress].v = value;
          column++;
        });
      } else if (header.value === '${var.all}') {
        if (!row.var || !row.var) {
          continue;
        }
        _lodash.default.each(row.var, function (value, name) {
          cellAddress = EXCEL_COLUMNS[column] + (r + 2);
          ws[cellAddress] = {};
          ws[cellAddress].v = value;
          column++;
        });
      } else {
        cellAddress = EXCEL_COLUMNS[column] + (r + 2);
        ws[cellAddress] = {};
        ws[cellAddress].v = substractParameter(header.value, row) || undefined;
        column++;
      }
    }
  }
  if (template.length > 0 && rows.length > 0) {
    ws['!ref'] = 'A1:' + EXCEL_COLUMNS[template.length - 1 + totalColumns] + (rows.length + 1);
  }
  return wb;
}
function substractParameter(valueDef, data) {
  if (!valueDef || !valueDef.length || valueDef.length == 0) return valueDef;
  if (valueDef.indexOf('${') == 0 && valueDef.indexOf('}') == valueDef.length - 1) {
    var paramName = valueDef.substring(2, valueDef.length - 1);
    if (paramName.length == 0) {
      return valueDef;
    }
    if (paramName.indexOf('param.') == 0) {
      return data.param[paramName.substring('param.'.length)];
    } else if (paramName.indexOf('var.') == 0) {
      return data.var[paramName.substring('var.'.length)];
    } else if (paramName.indexOf('cap.') == 0) {
      return data.cap[paramName.substring('cap.'.length)];
    }
    return data[paramName];
  }
  return valueDef;
}
function addValues(suiteResult, caseResult, lastFailedStep) {
  var values = {};
  values['suite.name'] = suiteResult.name;
  values['suite.iteration'] = suiteResult.iterationNum;
  values['suite.status'] = suiteResult.status;
  values['case.name'] = caseResult.name;
  values['case.iteration'] = caseResult.iterationNum;
  values['case.status'] = caseResult.status;
  if (lastFailedStep) {
    values['failure.step'] = lastFailedStep.name;
    if (lastFailedStep._screenshotFile) {
      values['failure.screenshot'] = lastFailedStep.screenshotFile;
    }
    if (lastFailedStep.failure) {
      values['failure.message'] = lastFailedStep.failure.message;
      values['failure.type'] = lastFailedStep.failure.type;
      if (lastFailedStep.failure.location) {
        values['failure.location'] = lastFailedStep.failure.location;
      }
    }
  }
  values.param = {};
  _lodash.default.each(caseResult.context.params, function (value, key) {
    values.param[key] = value;
  });
  values.var = {};
  _lodash.default.each(caseResult.context.vars, function (value, key) {
    values.var[key] = value;
  });
  values.cap = {};
  _lodash.default.each(caseResult.context.caps, function (value, key) {
    values.cap[key] = value;
  });
  return values;
}
function Workbook() {
  this.SheetNames = [];
  this.Sheets = {};
}
function generateExcelColumns() {
  var output = [];
  for (var i = 'A'.charCodeAt(0); i <= 'Z'.charCodeAt(0); i++) output.push(String.fromCharCode(i));
  return output;
}
