// oxygen-cli/build/ox_reporters/reporter-xml.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _path = _interopRequireDefault(require("path"));
var _fs = _interopRequireDefault(require("fs"));
var _FileReporterBase = _interopRequireDefault(require("../reporter/FileReporterBase"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var EasyXml = require('easyxml');
class XmlReporter extends _FileReporterBase.default {
  constructor(options) {
    super(options);
  }
  generate(results) {
    var resultFilePath = this.createFolderStructureAndFilePath('.xml');
    var resultFolderPath = _path.default.dirname(resultFilePath);
    var serializer = new EasyXml({
      singularize: true,
      rootElement: 'test-results',
      rootArray: 'test-results',
      dateFormat: 'ISO',
      manifest: true,
      unwrapArrays: false,
      filterNulls: true
    });
    this.replaceScreenshotsWithFiles(results, resultFolderPath);
    const forXmlRender = [];
    if (results && Array.isArray(results) && results.length > 0) {
      results.map(result => {
        forXmlRender.push({
          name: result.name,
          status: result.status,
          startTime: result.startTime,
          endTime: result.endTime,
          duration: result.duration,
          failure: result.failure,
          environment: result.environment,
          capabilities: result.capabilities,
          options: {
            cwd: result.options.cwd,
            target: result.options.target,
            browserName: result.options.browserName,
            seleniumUrl: result.options.seleniumUrl,
            appiumUrl: result.options.appiumUrl,
            reopenSession: result.options.reopenSession,
            reRunOnFailed: result.options.reRunOnFailed,
            iterations: result.options.iterations,
            debugPort: result.options.debugPort,
            delay: result.options.delay,
            collectDeviceLogs: result.options.collectDeviceLogs,
            collectAppiumLogs: result.options.collectAppiumLogs,
            collectBrowserLogs: result.options.collectBrowserLogs,
            reporting: result.options.reporting,
            parameters: result.options.parameters,
            parallel: result.options.parallel,
            capabilities: result.options.capabilities,
            services: result.options.services,
            modules: result.options.modules,
            framework: result.options.framework,
            applitoolsOpts: result.options.applitoolsOpts,
            envs: result.options.envs,
            name: result.options.name,
            env: result.options.env,
            po: result.options.po,
            scriptContentLineOffset: result.options.scriptContentLineOffset
          },
          suites: result.suites
        });
      });
    }
    var xml = serializer.render(forXmlRender);
    _fs.default.writeFileSync(resultFilePath, xml);
    return resultFilePath;
  }
}
exports.default = XmlReporter;
