// oxygen-cli/build/ox_reporters/reporter-rp.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _ReporterBase = _interopRequireDefault(require("../reporter/ReporterBase"));
var _status = _interopRequireDefault(require("../model/status"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const RPClient = require('@oxygenhq/rp-client-javascript');
const LAUNCH_MODES = {
  DEFAULT: 'DEFAULT',
  DEBUG: 'DEBUG'
};
const TEST_ITEM_TYPES = {
  SUITE: 'SUITE',
  TEST: 'TEST',
  STEP: 'STEP'
};
const MAX_NAME_LENGTH = 128;
class ReportPortalReporter extends _ReporterBase.default {
  constructor(options, reporterOpts, aggregator) {
    super(options);
    this.reporterOpts = reporterOpts || options.rp;
    this.reportSteps = reporterOpts.reportSteps || true;
    this.reportLogs = reporterOpts.reportLogs || false;
    this.cbSuiteToRpIdHash = {};
    this.cbCaseToRpIdHash = {};
    this.cbStepToRpIdHash = {};
    this.cbSuiteResultListByRefId = {};
    this.cbCaseToCurrentTransactionId = {};
    this.cbCaseToCurrentSubStepId = {};
    this.promises = [];
    if (!this.reporterOpts || !this.reporterOpts.apiKey || !this.reporterOpts.endpoint || !this.reporterOpts.project) {
      throw new Error('ReportPortal options are missing in oxygen.conf file.');
    }
    this.aggregator = aggregator;
    this.rpClient = new RPClient({
      launch: this.reporterOpts.launch || this.options.name,
      ...this.reporterOpts
    });
  }
  async init() {
    try {
      await this.rpClient.checkConnect();
    } catch (e) {
      throw new Error(`RP - Error connecting to the server: ${e.message}`);
    }
  }
  async generate(results) {}
  async onLaunchStart({
    options
  }) {
    try {
      const {
        tempId,
        promise
      } = this.rpClient.startLaunch({
        mode: this.reporterOpts.mode || LAUNCH_MODES.DEFAULT,
        debug: false
      });
      this.tempLaunchId = tempId;
      this.handlePromise(promise);
    } catch (e) {
      console.dir(`RP - Failed to start launch: ${e}`);
    }
  }
  async onLaunchEnd({
    results
  }) {
    await this._reportEndOfStartedSuites();
    const hasFailed = results.some(x => x.status === _status.default.FAILED);
    try {
      await Promise.all(this.promises);
      const {
        promise
      } = await this.rpClient.finishLaunch(this.tempLaunchId, {
        status: hasFailed ? 'FAILED' : 'PASSED'
      });
      await promise;
    } catch (e) {
      console.dir(`RP - Failed to end launch: ${e}`);
    }
  }
  async onRunnerStart({
    rid,
    opts,
    caps
  }) {}
  async onRunnerEnd({
    rid,
    result
  }) {}
  async onSuiteStart({
    rid,
    suiteId,
    suite: suiteDef
  }) {
    const suiteRefId = suiteDef.refId;
    if (this.cbSuiteToRpIdHash[suiteRefId]) {
      return;
    }
    this.cbSuiteToRpIdHash[suiteRefId] = {};
    this.cbSuiteResultListByRefId[suiteRefId] = [];
    const startTestItemReq = {
      name: suiteDef.name || this.options.name,
      type: TEST_ITEM_TYPES.SUITE
    };
    try {
      const {
        tempId,
        promise
      } = this.rpClient.startTestItem(startTestItemReq, this.tempLaunchId);
      this.cbSuiteToRpIdHash[suiteRefId] = tempId;
      this.handlePromise(promise);
    } catch (e) {
      console.dir(`RP - Failed to start suite item: ${e}`);
    }
  }
  async onSuiteEnd({
    rid,
    suiteId,
    result
  }) {
    const suiteRefId = result.refId;
    this.cbSuiteResultListByRefId[suiteRefId] && this.cbSuiteResultListByRefId[suiteRefId].push(result);
  }
  async onCaseStart({
    rid,
    suiteId,
    suiteRefId,
    caseId,
    'case': caseDef
  }) {
    const startTestItemReq = {
      name: caseDef.name,
      type: TEST_ITEM_TYPES.TEST,
      codeRef: caseDef.path
    };
    const rpSuiteId = this.cbSuiteToRpIdHash[suiteRefId];
    if (!rpSuiteId) {
      return;
    }
    try {
      const {
        tempId,
        promise
      } = this.rpClient.startTestItem(startTestItemReq, this.tempLaunchId, rpSuiteId);
      this.cbCaseToRpIdHash[caseId] = tempId;
      this.handlePromise(promise);
    } catch (e) {
      console.dir(`RP - Failed to start case item: ${e}`);
    }
  }
  async onCaseEnd({
    rid,
    suiteId,
    suiteRefId,
    caseId,
    result
  }) {
    const rpTestId = this.cbCaseToRpIdHash[caseId];
    if (!rpTestId) {
      return;
    }
    const status = result.status.toLowerCase();
    if (status === 'failed' && result.failure) {
      const rpFile = result.screenshot ? {
        name: 'screenshot',
        type: 'image/png',
        content: result.screenshot
      } : undefined;
      const logMsg = this._stringify(result.failure.message);
      if (logMsg) {
        const logReq = {
          message: logMsg,
          level: 'error'
        };
        try {
          const {
            promise
          } = this.rpClient.sendLog(rpTestId, logReq, rpFile);
          this.handlePromise(promise);
        } catch (e) {
          console.dir(`RP - Failed to send log for finished case item: ${e}`);
        }
      }
    }
    const transId = this.cbCaseToCurrentTransactionId[caseId];
    if (transId) {
      const finishTransactionReq = {
        status: 'passed'
      };
      try {
        const transRpId = this.cbStepToRpIdHash[transId];
        const {
          promise
        } = this.rpClient.finishTestItem(transRpId, finishTransactionReq);
        this.handlePromise(promise);
      } catch (e) {
        console.dir(`RP - Failed to finish transaction item on case end: ${e}`);
      }
      delete this.cbCaseToCurrentTransactionId[caseId];
    }
    const finishTestItemReq = {
      status: result.status.toLowerCase()
    };
    try {
      const {
        promise
      } = this.rpClient.finishTestItem(rpTestId, finishTestItemReq);
      this.handlePromise(promise);
    } catch (e) {
      console.dir(`RP - Failed to finish case item: ${e}`);
    }
  }
  async onStepStart({
    rid,
    caseId,
    step
  }) {
    if (!this.reportSteps) {
      return;
    }
    if (step.module && step.module === 'log') {
      await this.reportLogStep(caseId, step);
      return;
    }
    const rpCaseId = this.cbCaseToRpIdHash[caseId];
    let rpParentId;
    const transId = this.cbCaseToCurrentTransactionId[caseId];
    if (step.name === 'transaction') {
      if (transId) {
        const finishTransactionReq = {
          status: 'passed'
        };
        try {
          const transRpId = this.cbStepToRpIdHash[transId];
          const {
            promise
          } = this.rpClient.finishTestItem(transRpId, finishTransactionReq);
          this.handlePromise(promise);
        } catch (e) {
          console.dir(`RP - Failed to finish transaction item on case end: ${e}`);
        }
      }
      this.cbCaseToCurrentTransactionId[caseId] = step.id;
      delete this.cbCaseToCurrentSubStepId[caseId];
      rpParentId = rpCaseId;
    } else {
      this.cbCaseToCurrentSubStepId[caseId] = step.id;
      rpParentId = transId ? this.cbStepToRpIdHash[transId] : rpCaseId;
    }
    if (!rpParentId || !rpCaseId) {
      return;
    }
    const stepName = this._getStepName(step);
    const startTestItemReq = {
      name: stepName,
      type: TEST_ITEM_TYPES.STEP,
      codeRef: step.location,
      testCaseId: rpCaseId,
      hasStats: false
    };
    try {
      const {
        tempId,
        promise
      } = this.rpClient.startTestItem(startTestItemReq, this.tempLaunchId, rpParentId);
      this.cbStepToRpIdHash[step.id] = tempId;
      this.handlePromise(promise);
    } catch (e) {
      console.dir(`RP - Failed to start step item: ${e}`);
    }
  }
  async onStepEnd({
    rid,
    caseId,
    step: result
  }) {
    if (!this.reportSteps) {
      return;
    }
    if (result.name && (result.name.startsWith('log.') || result.name.startsWith('web.transaction') || result.name.startsWith('mob.transaction'))) {
      return;
    }
    const rpStepId = this.cbStepToRpIdHash[result.id];
    if (!rpStepId) {
      return;
    }
    delete this.cbCaseToCurrentSubStepId[caseId];
    const status = result.status.toLowerCase();
    if (status === 'failed' && result.failure) {
      const rpFile = result.screenshot ? {
        name: 'screenshot',
        type: 'image/png',
        content: result.screenshot
      } : undefined;
      const logMsg = this._stringify(result.failure.message);
      if (logMsg) {
        const logReq = {
          message: logMsg,
          level: 'error',
          file: rpFile
        };
        try {
          const {
            promise
          } = this.rpClient.sendLog(rpStepId, logReq, rpFile);
          this.handlePromise(promise);
        } catch (e) {
          console.dir(`RP - Failed to send log for finished step item: ${e}`);
        }
      }
    }
    const finishTestItemReq = {
      status: result.status.toLowerCase()
    };
    try {
      const {
        promise
      } = this.rpClient.finishTestItem(rpStepId, finishTestItemReq);
      this.handlePromise(promise);
    } catch (e) {
      console.dir(`RP - Failed to finish step item: ${e}`);
    }
  }
  async onLog({
    suiteId,
    caseId,
    stepId,
    level,
    msg,
    time,
    src
  }) {
    if (!this.reportLogs) {
      return;
    }
    if (!this.tempLaunchId) {
      return;
    }
    const rpParentId = stepId ? this.cbStepToRpIdHash[stepId] : caseId ? this.cbCaseToRpIdHash[caseId] : suiteId ? this.cbSuiteToRpIdHash[suiteId] : undefined;
    const rpLevel = this._getRpLevel(level);
    const logMsg = this._stringify(msg);
    if (logMsg) {
      const logReq = {
        message: logMsg,
        level: rpLevel,
        time: time
      };
      try {
        const {
          promise
        } = this.rpClient.sendLog(rpParentId || this.tempLaunchId, logReq);
        this.handlePromise(promise);
      } catch (e) {
        console.dir(`RP - Failed to create log item: ${e}`);
      }
    }
  }
  async reportLogStep(caseId, step) {
    if (!step.args || !step.args.length) {
      return;
    }
    const level = step.name;
    const msg = step.args[0];
    const time = step.time;
    let stepId = undefined;
    if (this.cbCaseToCurrentSubStepId[caseId]) {
      stepId = this.cbCaseToCurrentSubStepId[caseId];
    } else if (this.cbCaseToCurrentTransactionId[caseId]) {
      stepId = this.cbCaseToCurrentTransactionId[caseId];
    }
    await this.sendRpLog({
      suiteId: undefined,
      caseId,
      stepId,
      level,
      msg,
      time
    });
  }
  async sendRpLog({
    suiteId,
    caseId,
    stepId,
    level,
    msg,
    time
  }) {
    if (!this.tempLaunchId) {
      return;
    }
    const rpParentId = stepId ? this.cbStepToRpIdHash[stepId] : caseId ? this.cbCaseToRpIdHash[caseId] : suiteId ? this.cbSuiteToRpIdHash[suiteId] : undefined;
    const rpLevel = this._getRpLevel(level);
    const logMsg = this._stringify(msg);
    if (logMsg) {
      const logReq = {
        message: logMsg,
        level: rpLevel,
        time: time
      };
      try {
        const {
          promise
        } = this.rpClient.sendLog(rpParentId || this.tempLaunchId, logReq);
        this.handlePromise(promise);
      } catch (e) {
        console.dir(`RP - Failed to create log item: ${e}`);
      }
    }
  }
  async _reportEndOfStartedSuites() {
    for (const suiteRefId of Object.keys(this.cbSuiteToRpIdHash)) {
      const rpSuiteId = this.cbSuiteToRpIdHash[suiteRefId];
      const results = this.cbSuiteResultListByRefId[suiteRefId];
      delete this.cbSuiteToRpIdHash[suiteRefId];
      const hasFailedSuite = results.some(x => x.status === _status.default.FAILED);
      const finishTestItemReq = {
        status: hasFailedSuite ? 'failed' : 'passed'
      };
      try {
        const {
          promise
        } = this.rpClient.finishTestItem(rpSuiteId, finishTestItemReq);
        this.handlePromise(promise);
      } catch (e) {
        console.dir(`RP - Failed to finish suite item: ${e}`);
      }
    }
  }
  _getRpLevel(oxLevel) {
    if (oxLevel) {
      return oxLevel.toLowerCase();
    }
    return 'info';
  }
  _getStepName(step) {
    let name;
    if (step.name === 'transaction' && step.args.length > 0) {
      name = step.args[0];
    } else if (step.signature) {
      name = step.signature;
    } else if (step.module) {
      name = `${step.module}.${step.name}`;
    } else {
      name = step.name;
    }
    if (name.length > MAX_NAME_LENGTH) {
      name = name.substring(0, MAX_NAME_LENGTH - 3) + '...';
    }
    return name;
  }
  _stringify(msg) {
    if (typeof msg === 'undefined' || msg === null) {
      return null;
    }
    if (typeof msg === 'string' || msg instanceof String) {
      return msg;
    }
    try {
      return JSON.stringify(msg);
    } catch (e) {
      console.error('RP - error sending log: msg object cannot be serialized');
    }
  }
  handlePromise(promise) {
    promise.catch(err => {
      console.log('RP error: ', err);
    });
    this.promises.push(promise);
  }
}
exports.default = ReportPortalReporter;
