// oxygen-cli/build/ox_reporters/reporter-html.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _path = _interopRequireDefault(require("path"));
var _fs = _interopRequireDefault(require("fs"));
var _moment = _interopRequireDefault(require("moment"));
var _FileReporterBase = _interopRequireDefault(require("../reporter/FileReporterBase"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const ejs = require('ejs');
class HtmlReporter extends _FileReporterBase.default {
  constructor(options) {
    super(options);
  }
  generateSummary(results) {
    const summary = {
      status: 'passed',
      startDate: null,
      startTime: null,
      duration: null,
      endTime: null,
      totalTests: results.length,
      totalSuites: 0,
      passedSuites: 0,
      failedSuites: 0,
      totalCases: 0,
      passedCases: 0,
      failedCases: 0,
      skippedCases: 0,
      totalSteps: 0,
      passedSteps: 0,
      failedSteps: 0,
      skippedSteps: 0
    };
    summary.startTime = Math.min.apply(null, results.map(res => res.startTime));
    summary.endTime = Math.max.apply(null, results.map(res => res.endTime));
    summary.duration = summary.endTime - summary.startTime;
    summary.duration = this.formatDuration(summary.duration);
    for (let result of results) {
      summary.status = result.status === 'failed' ? 'failed' : summary.status;
      if (!Array.isArray(result.suites)) {
        continue;
      }
      summary.totalSuites += result.suites.length;
      for (let suiteResult of result.suites) {
        if (!Array.isArray(suiteResult.cases)) {
          continue;
        }
        summary.totalCases += suiteResult.cases.length;
        summary.passedSuites += suiteResult.status === 'passed' || suiteResult.status === 'warning' ? 1 : 0;
        summary.failedSuites += suiteResult.status === 'failed' ? 1 : 0;
        for (let caseResult of suiteResult.cases) {
          summary.passedCases += caseResult.status === 'passed' || caseResult.status === 'warning' ? 1 : 0;
          summary.failedCases += caseResult.status === 'failed' ? 1 : 0;
          summary.skippedCases += caseResult.status === 'skipped' ? 1 : 0;
          if (!Array.isArray(caseResult.steps)) {
            continue;
          }
          summary.totalSteps += caseResult.steps.length;
          for (let stepResult of caseResult.steps) {
            summary.passedSteps += stepResult.status === 'passed' || stepResult.status === 'warning' ? 1 : 0;
            summary.failedSteps += stepResult.status === 'failed' ? 1 : 0;
            summary.skippedSteps += stepResult.status === 'skipped' ? 1 : 0;
          }
        }
      }
    }
    summary.startDate = (0, _moment.default)(summary.startTime).format('MMM DD');
    summary.startTime = (0, _moment.default)(summary.startTime).format('HH:mm');
    summary.endTime = (0, _moment.default)(summary.endTime).format('HH:mm');
    return summary;
  }
  generate(results) {
    const resultFilePath = this.createFolderStructureAndFilePath('.html');
    const resultFolderPath = _path.default.dirname(resultFilePath);
    this.replaceScreenshotsWithFiles(results, resultFolderPath);
    this.formatTotalDuration(results);
    const templatePath = _path.default.join(__dirname, '../ox_reporters/html/index.ejs');
    const summary = this.generateSummary(results);
    ejs.renderFile(templatePath, {
      summary: summary,
      results: results
    }, null, function (err, html) {
      if (err) {
        throw err;
      }
      _fs.default.writeFileSync(resultFilePath, html);
    });
    return resultFilePath;
  }
}
exports.default = HtmlReporter;
